---
title: "EMA Alignment Depth and Coverage"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r echo = FALSE, message = FALSE, warning = FALSE}
library(flexdashboard)
library(ggplot2)
library(plotly)
library(dplyr)
library(tidyr)
library(circlize)
library(scales)
```

```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.7.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r fileinput, echo = FALSE, message = FALSE, warning = FALSE}
#covfile <- "~/Documents/harpy/test/Align/ema/stats/coverage/data/sample1.cov.gz"
covfile <- snakemake@input[[1]]
gencov <- read.table(covfile, header = F)
gencov <- gencov[, c(1, 2, 3, 6, 7)]
colnames(gencov) <- c("contig", "start", "end", "Barcode", "noBarcode")
gencov$noBarcode <- gencov$noBarcode * -1
sname <- gsub("\\.\\w?\\w?\\w?.?cov.gz", "", basename(covfile))
```

```{r averages, echo = FALSE, message = FALSE, warning = FALSE}
bc_avg <- mean(gencov$Barcode)
bc_sd <- sd(gencov$Barcode)
nobc_avg <- mean(abs(gencov$noBarcode))
nobc_sd <- sd(abs(gencov$noBarcode))
all_depth <- gencov$Barcode + abs(gencov$noBarcode)
q99 <- quantile(all_depth, 0.99)
gencov$all <- all_depth
global_avg <- mean(all_depth)
global_sd <- sd(all_depth)

zscores <- (all_depth - global_avg) / bc_sd
gencov$outlier <- zscores >= 3
outliers <- gencov[gencov$outlier, -6]
nonoutliers <- gencov[!(gencov$outlier), -6]
```

```{r filter_intervals, echo = FALSE, message = FALSE, warning = FALSE}
nonoutliers$all <- nonoutliers$Barcode + abs(nonoutliers$noBarcode)
bc_avg_filt <- mean(nonoutliers$Barcode)
bc_sd_filt <- sd(nonoutliers$Barcode)
nobc_avg_filt <- mean(abs(nonoutliers$noBarcode))
nobc_sd_filt <- sd(abs(nonoutliers$noBarcode))
all_depth_filt <- nonoutliers$all
global_avg_filt <- mean(all_depth_filt)
global_sd_filt <- sd(all_depth_filt)

filt_df <- data.frame(
  "contig" = "global",
  "average_all" = round(global_avg,2),
  "sdv_all" = round(global_sd,2),
  "average_Barcode" = round(bc_avg,2),
  "sdv_Barcode" = round(bc_sd,2),
  "average_noBarcode" = round(nobc_avg,2),
  "sdv_noBarcode" = round(nobc_sd,2)
)
```

```{r pivot_table, echo = FALSE, message = FALSE, warning = FALSE}
# pivot longer for generalized plotting
Lgencov <- pivot_longer(gencov[,-7], -(1:3), names_to = "alignment", values_to = "depth")
```

<h1> `r sname` </h1>
This report contains information regarding the sequence alignment coverage
and depth for the file `r paste0(sname, ".bam")`. In the colorful summary
boxes below, `BX` refers to valid haplotag barcodes, `no BX` refers to missing or
invalid haplotag barcodes, and `Stdev` refers to the standard deviation. The top
row corresponds to raw summary statistics ("unfiltered") and the bottom row ("filtered")
refers to summary statistics after removing putative outliers. The
term "filtered" here and elsewhere in this report refers to removing intervals
whose depth is greater than 3 standard deviations above the mean depth. The
filtering described is shown for diagnostic purposes and no filtering has been
performed on the original alignment file.

## Unfiltered Information {data-height=100}
### which-col
```{r}
valueBox("Unfiltered", caption = "10kbp Intervals")
```

### ncontigs
```{r}
valueBox(scales::comma(length(unique(gencov$contig))), caption = "Contigs", color = "info")
```

### bx-avg
```{r}
valueBox(scales::comma(bc_avg, accuracy = 0.01), caption = "Average depth (BX)", color = "info")
```

### nobx-avg
```{r}
valueBox(scales::comma(nobc_avg, accuracy = 0.01), caption = "Average depth (no BX)", color = "info")
```

### bx-sd
```{r}
valueBox(scales::comma(bc_sd, accuracy = 0.01), caption = "Stdev depth (BX)", color = "info")
```

### nobx-sd
```{r}
valueBox(scales::comma(nobc_sd, accuracy = 0.01), caption = "Stdev depth (no BX)", color = "info")
```

### n-outliers
```{r}
valueBox(scales::comma(nrow(outliers)), caption = "Possible outlier regions", color = "warning")
```

## Filtered Information {data-height=100}
### which-col
```{r}
valueBox("Filtered", caption = "10kbp Intervals", icon = "fa-filter")
```
### ncontigs
```{r}
valueBox(scales::comma(length(unique(gencov$contig))), caption = "Contigs", color = "info")
```

### bx-avg
```{r}
valueBox(scales::comma(bc_avg_filt, accuracy = 0.01), caption = "Average depth (BX)", color = "info")
```

### nobx-avg
```{r}
valueBox(scales::comma(nobc_avg_filt, accuracy = 0.01), caption = "Average depth (no BX)", color = "info")
```

### bx-sd
```{r}
valueBox(scales::comma(bc_sd_filt, accuracy = 0.01), caption = "Stdev depth (BX)", color = "info")
```

### nobx-sd
```{r}
valueBox(scales::comma(nobc_sd, accuracy = 0.01), caption = "Stdev depth (no BX)", color = "info")
```

### n-outliers
```{r}
valueBox(0, caption = "Possible outlier regions", color = "success")
```

## Distdesc header
### distdesc {.no-title}
<h2> Alignment Depth Distribution </h2>
Below are the frequencies of interval coverage across all 10kbp intervals for all contigs.
The distribution is given for bins of width 5, where `depth = 0` is equivalent to 
intervals of depth between `0` and `5`. 
For visual clarity, the X-axis of this plot is truncated at the 99% quantile of depth
values, which is **`r q99`** for these data.

## disrow
### distcol {.no-title}
```{r echo=F, warning=F, message=F}
distplot <- ggplot(Lgencov[Lgencov$alignment != "all",], aes(x = depth, fill = alignment)) +
  geom_histogram(binwidth = 5, linewidth = 2) +
  theme_minimal() +
  scale_y_continuous(labels = scales::comma) +
  scale_x_continuous(labels = scales::comma) +
  scale_fill_manual(values = c("#3f3f3f", "#cb6843")) +
  xlim(NA, q99) +
  labs(title = "Distribution of Alignment Depths") +
  xlab("Depth") +
  ylab("Count")
ggplotly(distplot)
```

## Sumheader
### sumhead {.no-title}
<h2> Coverage Summary Information </h2>
Below are tables that summarize coverage information for 10kbp intervals.

## Tableheaders
### Sumdesc {.no-title}
<h4> Alignment metrics with all intervals </h4>
The table below shows the global and per-contig average depth and standard 
deviation per 10kbp intervals **including** intervals whose depth is flagged
an outlier in the data. 

### filtdesc {.no-title}
<h4> Alignment metrics without outlier intervals </h4>
The table below shows the global and per-contig average depth and standard 
deviation per 10kbp intervals, **excluding** intervals whose depth is flagged
an outlier in the data, as determined by being greater than 3 standard deviations
above the mean depth. This should be a more accurate representation of read coverage.

### outlierdesc {.no-title}
<h4> Putative outliers </h4>
Below is a table of the intervals with a coverage being greater than 3 standard 
deviations above the mean depth. `Barcode Depth` refers to the alignment depth of sequences with a valid
haplotag barcode, `noBarcode Depth` refers to the alignment depth of sequences without a valid haplotag
barcode, and `Total Depth` is the sum of `Barcode Depth` and `noBarcode Depth`.

## Summary information
### Averages {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE}
contig_avg <- group_by(Lgencov, contig, alignment) %>%
  summarize(average = round(mean(abs(depth)),2), sdv = round(sd(abs(depth)),2))

contig_avg <- pivot_wider(data = contig_avg, names_from = alignment,values_from = c(average,sdv))[, c(1,3,6,2,5,4,7)]

DT::datatable(
  contig_avg,
  rownames = F,
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = c('Average All', 'StDev All', 'Average Barcode', 'StDev Barcode', 'Average noBarcode', 'StDev noBarcode'),
  autoHideNavigation = T,
  fillContainer = T
)
```
### Filtered Averages {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE}
Lnonoutliers <- pivot_longer(nonoutliers[,-6], -(1:3), names_to = "alignment", values_to = "depth")

contig_avg_filt <- group_by(Lnonoutliers, contig, alignment) %>%
  summarize(average = round(mean(abs(depth)),2), sdv = round(sd(abs(depth)),2))

contig_avg_filt <- pivot_wider(data = contig_avg_filt, names_from = alignment,values_from = c(average,sdv))[, c(1,3,6,2,5,4,7)]

contig_avg_filt <- rbind(filt_df, contig_avg_filt)

DT::datatable(
  contig_avg_filt,
  rownames = F,
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = c('Average All', 'StDev All', 'Average Barcode', 'StDev Barcode', 'Average noBarcode', 'StDev noBarcode'),
  autoHideNavigation = T,
  fillContainer = T
)
```


### Outliers {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE}
outliers$noBarcode <- abs(outliers$noBarcode)
outliers$all <- outliers$Barcode + abs(outliers$noBarcode)
DT::datatable(
  outliers[,-6],
  rownames = F,
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = c('Contig', 'Position Start', 'Position End', 'Barcode Depth', 'noBarcode Depth', 'Total Depth'),
  autoHideNavigation = T,
  fillContainer = T
)
```

## Plot header
### pltdsc {.no-title}
<h2> Depth and Coverage Across the Genome </h2>
Below are plots of the depth and coverage of alignments for this sample. Clicking
on a plot will expand it to fill your browser window. Clicking it again will exit
out of the zoomed view.

## Alignment Depth plots desc

```{r echo = FALSE, message = FALSE, warning = FALSE, dev = "jpeg"}
# Find the 30 largest contigs
contigs <- group_by(Lnonoutliers, contig) %>%
  summarize(size = max(end)) %>%
  arrange(desc(size))

# limit the data to only the 30 largest contigs
if (nrow(contigs) > 30) {
    pltheight <- 45
    .contigs <- contigs[1:30, ]
    .gencov <- filter(Lnonoutliers, contig %in% .contigs$contig) %>%
      filter(alignment != "all")
} else {
    pltheight <- round(1.5 * (nrow(contigs)), digits = 0)
    .gencov <- Lnonoutliers %>% filter(alignment != "all")
    .contigs <- contigs
}

.l <-  nrow(.contigs)
```

### circos desc {.no-title}
Plotted below are (up to) the 30 largest contigs to show sequence alignment depth across
the genome. Each bar represents the alignment depth at a 10kb genomic interval, that is, the
number of reads that had a proper alignment in the 10kb interval. "Proper" refers to a read
not marked as a duplicate or flagged with the SAM `UNMAP`, `SECONDARY`,  or `QCFAIL` flags.
These values are derived by using `samtools bedcov -n`. The black lines show depth for properly
barcoded alignments and the orange lines show the depth for alignments with missing/invalid barcodes.

### depth desc {.no-title}
Plotted below is the same depth information, but shown in better detail per 
contig for easier visual assessment. Note that regions of extremely high depth
are plotted with `depth = 0` and are flagged as outliers using blue open circles
along the X-axis for visual clarity. Investigate
the table above for more information about these regions. The non-barcoded alignments
appear below the X axis for visual clarity.

## depth viz
### circos {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE, fig.align='center', out.width= "100%", dev = "jpeg"}
nonoutliers$noBarcode <- nonoutliers$noBarcode * -1

circos.clear()
col_text <- "grey40"
circos.par("track.height" = 0.8, gap.degree = 4, cell.padding = c(0, 0, 0, 0))
circos.initialize(
  factors = .contigs$contig,
  xlim = matrix(c(rep(0, .l), .contigs$size), ncol = 2)
)

# contigs
circos.track(
  ylim = c(0, 1),
  panel.fun = function(x, y) {
    chr = CELL_META$sector.index
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    circos.text(
      mean(xlim),
      mean(ylim),
      chr,
      cex = 0.5,
      col = col_text,
      facing = "bending.inside",
      niceFacing = TRUE
      )
    },
  bg.col = "grey90",
  bg.border = F,
  track.height = 0.06
)

# x axis
brk <- seq(0, 10, 0.5) * 10^7
circos.track(
  track.index = get.current.track.index(),
  panel.fun = function(x, y) {
    circos.axis(
      h = "top",
      major.at = brk,
      labels = round(brk / 10^6, 1),
      labels.cex = 0.4,
      col = col_text,
      labels.col = col_text,
      lwd = 0.7,
      labels.facing = "clockwise"
    )
  },
  bg.border = F
)

maxaln <- max(abs(.gencov$depth))
cov.yaxis <- c(0, round(maxaln * 1/3, digits = -1), round(maxaln * 2/3, digits = -1), maxaln)

# depth
circos.genomicTrack(
  data=nonoutliers, 
  panel.fun=function(region, value, ...) {
    circos.genomicLines(region, value, type="l", col="#3f3f3f", lwd=0.8)
    circos.genomicLines(region, value, type="l", col="#cb6843", lwd=0.8, numeric.column = 2)
  }, 
  track.height=0.5,
  bg.border=F
)
circos.yaxis(at=cov.yaxis, labels.cex=0.4, lwd=0.8, labels.col=col_text, col="grey30")

```

### Per-Contig {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE}
gencov_plotdata <- filter(gencov, contig %in% .contigs$contig) %>%
  select(-6) %>% 
  pivot_longer(4:5, names_to = "alignment", values_to = "depth")
gencov_plotdata$depth[gencov_plotdata$outlier] <- 0
outlier_plotdata <- pivot_longer(outliers, 4:5, names_to = "alignment", values_to = "depth")
```

```{r echo = FALSE, message = FALSE, warning = FALSE, fig.height=pltheight, out.width = "100%", dev='jpeg'}
  ggplot(gencov_plotdata, aes(x = start/1000000, y = depth)) +
    geom_line(aes(color = alignment)) +
    #geom_segment(data = outlier_plotdata, color = "#9393d2", size = 0.2, aes(x = start/1000000, y = 0, xend = start/1000000, yend = global_avg)) +
    geom_point(data = outlier_plotdata, color = "dodgerblue", size = 1, shape = 1, aes(x = start/1000000, y = 0)) +
    theme_minimal() +
    facet_wrap(~contig, ncol = 2, scales = "free_x") +
    labs(x = "position (Mbp)", y = "depth") +
    scale_y_continuous(labels = scales::comma) +
    scale_x_continuous(labels = scales::comma) +
    scale_color_manual(values = c("#3f3f3f", "#cb6843")) +
    theme(
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
    )
```