---
title: "BCFtools Stats Report"
date: "`r format(Sys.time(), '%m-%d-%y %X')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
library(flexdashboard)
library(ggplot2)
library(magrittr)
library(tidyr)
library(DT)
library(plotly)
library(RColorBrewer)
library(scales)
#dataL <- readLines("~/checkstats.txt")
#bcf <- "filename.bcf"
dataL <- readLines(snakemake@input[[1]])
bcf <- gsub(".stats$", ".bcf", snakemake@input[[1]])
```
<h1> `r bcf` </h1>

This reflects the general information stored in the records of `r bcf`.

```{r General Stats, echo = FALSE, message = FALSE, warning = FALSE}
.snL <- grepl("^SN", dataL)
sn <- read.table(text=dataL[.snL], sep = "\t")[,3:4]
names(sn) <- c("Metric", "Number")
sn$Metric <- gsub("number of ", "", sn$Metric)
rownames(sn) <- sn$Metric
sn <- as.data.frame(t(sn[2]))
```
## General Information {data-height=100}

### general-samples
```{r}
valueBox(scales::comma(sn[1,1]), caption = "Samples", color = "success")
```
### general-records
```{r}
valueBox(scales::comma(sn[1,2]), caption = "Records", color = "info")
```
### general-snps
```{r}
valueBox(scales::comma(sn[1,4]), caption = "SNPs", color = "info")
```
### general-mnps
```{r}
valueBox(scales::comma(sn[1,5]), caption = "MNPs", color = "info")
```
### general-indels
```{r}
valueBox(scales::comma(sn[1,6]), caption = "Indels", color = "info")
```
### general-multisites
```{r}
valueBox(scales::comma(sn[1,8]), caption = "Multiallelic Sites", color = "info")
```
### general-multisnps
```{r}
valueBox(scales::comma(sn[1,9]), caption = "Multiallelic SNPs", color = "info")
```
### general-noalts
```{r}
valueBox(scales::comma(sn[1,3]), caption = "No-ALTs", color = "info")
```
### general-other
```{r}
valueBox(scales::comma(sn[1,7]), caption = "Others", color = "info")
```

## Ind Statistics Desc
### Ind Desc {.no-title}
<h2> Individual Statistics </h2>
These are per-locus statistics, which correspond to the Per-Sample Counts (`PSC`) calculations of `bcftools stats`. The table reflects the data visualized in the plot tab.

## Individual Statistics  {.tabset}
### Individual Statistics
```{r Individual plot, echo=FALSE, message=FALSE, warning=FALSE, paged.print=TRUE, out.width="100%"}
.pscL <- grepl("^PSC", dataL)
psc <- read.table(text=dataL[.pscL])[ ,3:14]
names(psc) <- c("Sample", "HomozygousRef", "HomozygousAtl", "Heterozygotes", "Transitions", "Transversions", "Indels",	"MeanDepth", "Singletons",	"HapRef", "HapAlt", "Missing")
psc$Homozygotes <- psc$HomozygousRef + psc$HomozygousAtl
tidy_psc <- pivot_longer(psc[,c(1,4,5,6,7,9,12,13)], -Sample , names_to = "Metric", values_to = "Count")
psc <- psc[,c(1,8,12,2,3,13,4,5,6,7,9,10,11)]

pscplot <- ggplot(data = tidy_psc, mapping = aes(x = Count, y = Sample, color = Metric)) +
  geom_point(size = 2) +
  labs(title = "Individual Statistics") +
  theme_bw() +
  scale_x_continuous(n.breaks=9, labels = comma) +
  xlab("Count") +
  scale_color_brewer(palette = "Dark2") +
  theme(panel.grid.minor = element_blank(), panel.grid.major.y = element_blank(), axis.title.y = element_blank())

ggplotly(pscplot)
```

### Individual Data
```{r Individual table, echo=FALSE, message=FALSE, warning=FALSE}
DT::datatable(psc, rownames = F, filter = "top", extensions = 'Buttons', options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE))
```

## Quality Statistics Desc
### Desc {.no-title}
<h2> Quality Statistics </h2>
These are per-locus statistics, which correspond to the Quality (`QUAL`) statistics of `bcftools stats`. 
Each quality score is has a semi-open bin width of 10, where `0` would be considered "a QUAL score
greater than or equal to `0` and less than `10`".

## Quality Statistics
```{r qual data, echo=FALSE, message=FALSE, warning=FALSE}
.qual <- grepl("^QUAL", dataL)
if(sum(.qual) == 0){
  do_qual <- FALSE
  cat(paste0("No QUAL section in", snakemake@input[[1]], " found\n"))
} else {
  do_qual <- TRUE
}
if(do_qual){
  qual <- read.table(text=dataL[.qual])[ , c(3,4,7)]
  names(qual) <- c("QualityScore", "SNPs", "Indels")
  qual[names(qual)] <- sapply(qual[names(qual)],as.numeric)
}
```
```{r resizehist, echo=F}
# function to merge QUAL bins into a width of 10
mergebins <- function(x){
  bins <- seq(min(x[,1]), max(x[,1]), 10)
  lastbin <- max(x[,1])
  bins <- c(bins, lastbin)
  dict <- vector(mode="list", length=length(bins))
  names(dict) <- as.character(bins)
  for(i in 1:nrow(x)){
    databin <- x[i,1]
    key <- max(which(databin >= bins))
    dict[[key]] <- c(dict[[key]], x[i,2])
  }
  dict <- lapply(dict, sum)
  df <- as.data.frame(t(as.data.frame(dict)))
  df$V2 <- as.integer(gsub("X", "", row.names(df)))

  df <- df[,c(2,1)]
  names(df) <- names(x)
  row.names(df) <- NULL
  return(df)
}

```
### SNP Quality {.no-title}
```{r snpqual plot, echo=FALSE, message=FALSE, warning=FALSE}
if(do_qual){
qualsnpplot <- ggplot(data = mergebins(qual[,-3]), mapping = aes(x = QualityScore, y = SNPs)) +
  geom_col(fill = "#8fbbbc") +
  scale_y_continuous(labels = scales::comma) +
  scale_x_continuous(labels = scales::comma) +
  labs(title = "SNP Genotype Quality") +
  ylab("Count") +
  theme_minimal() +
  theme(panel.grid.minor.y = element_blank())

ggplotly(qualsnpplot)
} else {
  cat(paste0("No QUAL section in", snakemake@input[[1]], " found\n"))
}
```

### Indel Quality {.no-title}
```{r indelqual plot, echo=FALSE, message=FALSE, warning=FALSE}
qualindelplot <- ggplot(data = mergebins(qual[,-2]), mapping = aes(x = QualityScore, y = Indels)) +
  geom_col(fill = "#9393d2") +
  scale_y_continuous(labels = scales::comma) +
  scale_x_continuous(labels = scales::comma) +
  labs(title = "Indel Genotype Quality") +
  ylab("Count") +
  theme_minimal() +
  theme(panel.grid.minor.y = element_blank())

ggplotly(qualindelplot)
```

## Depth Desc
### DepthDesc {.no-title}
<h2> Depth Statistics </h2>
These are per-locus statistics, which correspond to the Depth (`DP`) calculations of `bcftools stats`
The table reflects the data visualized in the plot tab.

## Depth Statistics

### Depth Statistics {.no-title}
```{r depth plot, echo=FALSE, message=FALSE, warning=FALSE, paged.print=TRUE, out.width="100%"}
.dpL <- grepl("^DP", dataL)
if(sum(.dpL) == 0){
  cat(paste0("No DP section in", snakemake@input[[1]], " found\n"))
} else {
  dp <- read.table(text=dataL[.dpL])[ , 3:7]
  names(dp) <- c("Bin", "Genotypes", "PercentGenotypes", "NumberSites", "PercentSites")
  dp$Bin <- as.numeric(dp$Bin)
  dp$PercentGenotypes <- round(dp$PercentGenotypes, digits = 3)
  dp$PercentSites <- round(dp$PercentSites, digits = 3)
  dpplot <- ggplot(data = dp, mapping = aes(x = Bin, y = NumberSites)) +
    geom_bar(stat = "identity", color = "#F28500", fill = "#F28500") +
    scale_y_continuous(n.breaks=10) +
    labs(title = "Depth per Site") +
    ylab("Number of Sites") +
    theme_minimal() +
    theme(panel.grid.minor.y = element_blank())

  ggplotly(dpplot)
}
```

### Depth Data  {data-width=350 .no-title}

```{r depth table, echo=FALSE, message=FALSE, warning=FALSE, paged.print=TRUE, out.width="100%"}
if(sum(.dpL) == 0){
  cat(paste0("No DP section in", snakemake@input[[1]], " found\n"))
} else {
  DT::datatable(
    dp,
    rownames = F,
    filter = "top",
    extensions = 'Buttons',
    options = list(dom = 'Brtip', buttons = c('csv')),
    colnames = c("Bin", "Genotypes", "% Genotypes", "Sites", "% Sites")
    )
}
```

## Indels
### indelDesc {.no-title}
<h2> Insertions and Deletions </h2>
Below is the distribution of insertions and deletions based on length and frequency. The deletion
length is presented here as negative values for visual clarity.

## Insertions and Deletions
### Insertions and Deletions {.no-title}
```{r Indel Stats, echo=FALSE, message=FALSE, warning=FALSE, paged.print=TRUE, out.width="100%"}
.iddL <- grepl("^IDD", dataL)
if(sum(.iddL) == 0){
  cat(paste0("No IDD section in", snakemake@input[[1]], " found\n"))
} else {
  idd <- read.table(text=dataL[.iddL])[ ,3:4]
  names(idd) <- c("Length", "Count")
  idd$Type <- idd$Length > 0
  idd$Type <-  gsub(TRUE, "Insertion", idd$Type)
  idd$Type <-  gsub(FALSE, "Deletion", idd$Type)
  iddplot <- ggplot(data = idd, mapping = aes(x = Length, y = Count,  color = Type)) +
    geom_point() +
    scale_color_manual(values=c("indianred","#56B4E9"))+
    labs(title = "Insertion-Deletion Distribution") +
    xlab("indel length") +
    ylab("number of sites") +
    theme_minimal()

  ggplotly(iddplot)
}
```

