---
title: "Haplotag QC Bardcode Summary"
date: "`r format(Sys.time(), '%m-%d-%y %X')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r echo = FALSE, message = FALSE}
library(flexdashboard)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(magrittr)
library(DT)
library(plotly)
```

This report details the counts of valid and invalid haplotag barcodes in the 
sample read headers. The correct haplotag format is `BX:Z:AXXCXXBXXDXX`, where 
there is a whitespace/tab before `BX:Z:` and `XX` are 2-digit numbers between
`00` and `96`. A valid FASTQ read header would then look like:

```         
@A00814:267:HTMH3DRXX:2:1101:4580:1000 BX:Z:A02C01B11D46        RX:Z:GAAACGACCAACA+CGAACACGTTAGC   QX:Z:F,FFFFFFFFFFF+FF,FFF:FFFFFF
```

Notably, only the sequence ID (`@...`) and `BX:Z:` tag are required. 
In the example above, there are additional tags (`RX:Z:` and `QX:Z:`) 
which arent used by Harpy, but they conform to the 
[SAM comment spec (section 1.5)](https://samtools.github.io/hts-specs/SAMv1.pdf) 
of `TAG:TYPE:VALUE`. The takeaway is that the `BX:Z` tag can be anywhere\* in the 
read header after the sequence ID as long as any tags after it conform to the 
SAM spec `TAG:TYPE:VALUE`. 

| \*The structural variant caller Levaithan requires the BX:Z: tag to be the final tag in an alignment record. Harpy automatically moves the `BX:Z:` tag to the end of the alignment record when using `harpy align`. |
| :---:|

## genstats {data-height=100}
```{r instantiate_df, echo = FALSE, message = FALSE, warning = FALSE}
BXstats <- data.frame(
  Sample = character(),
  TotalReads = integer(),
  TotalBarcodes = integer(),
  ValidBarcodes = integer(),
  ValidPercent = numeric(),
  InvalidBarcodes = integer(),
  InvalidPercent = numeric()
)

invalids <- data.frame(
  Sample = character(),
  A = integer(),
  C = integer(),
  B = integer(),
  D = integer()
)
```

```{r fill_df, echo = FALSE, message = FALSE, warning = FALSE}
#for (i in list.files("~/Documents/harpy/test/QC/logs/bxcount/", ".count.log", full.names = TRUE)) {
for (i in snakemake@input[["countlog"]]) {
  samplename <- gsub(".count.log", "", basename(i))
  s_df <- read.table(i, header = F)
  total <- s_df$V2[1]
  bc_total <- s_df$V2[2]
  bc_ok <- s_df$V2[3]
  bc_invalid <- s_df$V2[4]
  if (bc_total == 0) {
    bc_percent <- 0
    ig_percent <- 0
  } else {
    bc_percent <- round((bc_ok / bc_total) * 100, digits = 3)
    ig_percent <- round((bc_invalid / bc_total) * 100, digits = 3)
  }

  BXstats <- rbind(
    BXstats,
    data.frame(
      Sample = samplename,
      TotalReads = total,
      TotalBarcodes = bc_total,
      ValidBarcodes = bc_ok,
      ValidPercent = bc_percent,
      InvalidBarcodes = bc_invalid,
      InvalidPercent = ig_percent
      )
  )
  invalids <- rbind(
    invalids,
    data.frame(
      Sample = samplename,
      TotalBarcodes = bc_total,
      A = s_df$V2[5],
      C = s_df$V2[6],
      B = s_df$V2[7],
      D = s_df$V2[8]
    )
  )
}
```


### filerow {.no-title}
```{r}
valueBox(scales::comma(nrow(BXstats)), caption = "Samples", color = "success")
```
### mininv {.no-title}
```{r}
valueBox(min(BXstats$ValidPercent), caption = "Minimum % Valid", color = "info")
```
### maxinv {.no-title}
```{r}
valueBox(max(BXstats$ValidPercent), caption = "Maximum % Valid", color = "info")
```

### minign {.no-title}
```{r}
absent <- BXstats$TotalReads-BXstats$TotalBarcodes
valueBox(min(absent), caption = "Minimum Absent Barcodes", color = "info")
```
### maxign {.no-title}
```{r}
valueBox(max(absent), caption = "Maximum Absent Barcodes", color = "info")
```

## persamp
### Persamp desc {.no-title}
<h3> General Per-Sample Haplotag Barcode Statistics </h3>

Below is a table listing all the samples Harpy processed and their associated
haplotag barcode statistics as determined by the reads in the **forward read only**.
If for some reason `TotalBarcodes` equals `0`, then there may be an issue with
the format of your FASTQ headers.

## persamp viz
```{r echo = FALSE, message = FALSE, warning = FALSE}
plotdata <- BXstats %>%
  select(Sample, ValidBarcodes, InvalidBarcodes) %>% 
  rename(Valid = ValidBarcodes, Invalid = InvalidBarcodes)

percdata <- BXstats %>%
  select(Sample, ValidPercent, InvalidPercent) %>% 
  rename(Valid= ValidPercent, Invalid = InvalidPercent)
```

### Table {.no-title}
```{r}
DT::datatable(
  BXstats,
  rownames = FALSE,
  filter = "top",
  extensions = "Buttons",
  options = list(
    dom = "Brtip",
    buttons = c("csv"),
    scrollX = TRUE
  ),
  caption = "Overall barcode quality per sample",
  autoHideNavigation = T,
  fillContainer = T
)
```


## table desc
### inv desc {.no-title}
<h3> Percent Beadtag Segment Invalidations </h3>
Below is information describing segment invalidations across all **forward reads**
within samples. The term "segments" refers to the combinatorial `Axx`, `Bxx`, 
`Cxx`, `Dxx` segments that make up a complete `AxxCxxBxxDxx` haplotag barcode.

## segment dist desc
### dist desc {.no-title}
This plot shows the distribution of percent invalid beadtag segments
across all samples. This should help you assess whether some segments have a 
higher incidence of failure than others. The darker horizontal lines within each
violin plot indicate the 25th, 50th (median), 75th percentiles in ascending order
from the bottom of the plot. The grey point in each violin plot is the mean and
the vertical grey lines emerging from it are the standard deviation. 

### dist desc {.no-title}
The table below details the number of invalid beadtag segments as a 
percent of total beadtag barcodes within a sample.
It's likely that percentages will have greater diagnostic value than the total counts
because the number of reads and barcodes will vary between individual samples.
These data are visualized in the scatterplot below this section.

## Segment
```{r echo = FALSE, message = FALSE, warning = FALSE}
invalids_percent <- data.frame(
  Sample = invalids$Sample,
  A = round(invalids$A/invalids$TotalBarcodes * 100, digits = 3),
  C = round(invalids$C/invalids$TotalBarcodes * 100, digits = 3),
  B = round(invalids$B/invalids$TotalBarcodes * 100, digits = 3),
  D = round(invalids$D/invalids$TotalBarcodes * 100, digits = 3)
)
invalids_long <- pivot_longer(invalids_percent, -1, names_to = "Segment", values_to = "Count")
```

### segment dist {.no-title}
```{r}
ggplot(invalids_long, aes(x = Segment, y = Count, fill = Segment, color = Segment)) +
  geom_violin(trim = FALSE, alpha = 0.5, draw_quantiles = c(0.25, 0.5, 0.75)) +
  stat_summary(fun.data=mean_sdl, geom="pointrange", color="grey20", alpha = 0.5) +
  scale_y_continuous(labels = scales::comma) +
  scale_fill_manual(values = c("#4a9fea", "#5a8c84", "#ffd75f", "#df487f")) +
  scale_color_manual(values = c("#4a9fea", "#5a8c84", "#ffd75f", "#df487f")) +
  theme_minimal() +
  theme(legend.position = "none", panel.grid.major.x = element_blank()) +
  labs(title = "Distribution of Percent Invalid Beadtag Segments")
```

### table per thousand {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE}
DT::datatable(
  invalids_percent,
  rownames = FALSE,
  extensions = "Buttons",
  options = list(
    dom = "Brtip",
    buttons = c("csv"),
    scrollX = TRUE
  ),
  colnames = c("Sample","A", "C","B","D"),
  caption = "Invalid segments relative to total barcodes",
  autoHideNavigation = T,
  fillContainer = T
)
```
## Segment viz description
### indiv desc {.no-title}
<h3> Per-Sample Barcode and Segment Invalidations </h3>
Below are two visualizations showing the occurrence of invalid haplotag barcodes.
The plot on the left shows this as the count of valid vs. invalid barcodes 
per individual, whereas the plot on the right shows this in finer detail as
the percent invalid beadtag segments per individual. Together, these two plots
should help you assess if any individuals have an unusually high amount of barcode
invalidations and if any particular beadtag segments appear to be more prone to
failure than others.

```{r}
pltheight <- 150 + (15 * nrow(invalids))
frameheight <- 1.2 + (0.166 * nrow(invalids))
```

## segment plots {data-height=pltheight}
### general {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE, fig.height = frameheight, out.width = '100%'}
p <-  plotdata %>%
  pivot_longer(-Sample, names_to = "Barcode", values_to = "Count") %>%
  ggplot(aes(y = Sample, x = Count, fill = Barcode)) +
    geom_col() +
    theme_minimal() +
    theme(panel.grid.major.y = element_blank(), axis.title.y = element_blank()) +
    scale_fill_manual(values = c(Valid = "#afd5e7", Invalid = "#c02554")) +
    scale_x_continuous(labels = scales::comma) +
    labs(title = "Barcode Count and Validity")

ggplotly(p, height = pltheight)
```
### indiv segments {.no-title}
```{r fig.height=frameheight}
p <- ggplot(invalids_long, aes(x = Count, y = Sample, color = Segment)) +
  geom_point(size = 2) +
  theme_minimal() +
  #theme(panel.grid.major.y = element_blank()) +
  scale_color_manual(values = c("#4a9fea", "#5a8c84", "#ffd75f", "#df487f")) +
  scale_x_continuous(labels = scales::comma) +
  ylab("") +
  labs(title = "Percent Invalid Beadtag Segments")
ggplotly(p, height = pltheight)
```
