---
title: "Haplotag Molecule Report"
date: "`r format(Sys.time(), '%m-%d-%y %X')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r echo = F, results = F, message = F}
library(flexdashboard)
library(dplyr)
library(ggplot2)
library(plotly)
library(magrittr)
library(DT)
library(scales)
```

```{r echo = F, results = F, message = F}
#infile <- "~/file.bx.stats.gz"
#mdist <- scales::comma(200000)
infile <- snakemake@input[[1]]
mdist <- snakemake@params[[1]]
bamfile <- gsub(".bxstats.gz", ".bam", infile)
samplename <- gsub(".bxstats.gz", "", basename(infile))
tb <- read.table(infile, header = T, sep = "\t") %>% select(-start, -end)
tb$valid <- tb$bx
tb[!(tb$valid %in% c("noBX", "invalidBX")), "valid"] <- "validBX"
tb$valid <- gsub("BX", " BX", tb$valid)
```

```{r bxper, echo = F, results = F, message = F}
valids <- tb[!(tb$valid %in% c("no BX", "invalid BX")),]
nBX <- group_by(valids, contig) %>% 
  summarize(nBX = length(bx))

avgBX <- round(mean(nBX$nBX), digits = 2)

totuniqBX <- length(unique(valids$bx))

tots <- tb %>% 
    group_by(valid) %>%
    summarize(total = length(bx)) %>% 
    as.data.frame()
for(i in c("no BX", "invalid BX", "valid BX")){
  if (!(i %in% tots$valid)){
    tots <- rbind(tots, data.frame("valid" = i, total = 0))
  }
}
```

## fileheader
### hdr {.no-title}
<h1> Haplotag Barcode Statistics </h1>
The information presented below were gathered from the alignments within `r basename(bamfile)`.


## General Information {data-height=100}
### ncontigs
```{r}
valueBox(scales::comma(length(unique(tb$contig))), caption = "Contigs")
```

### validBX
```{r}

valueBox(scales::comma(tots[tots$valid == "valid BX", 2]), caption = "Valid Barcodes", color = "success", icon = "fa-vial-circle-check")
```

### invalidBX
```{r}
valueBox(scales::comma(tots[tots$valid == "invalid BX", 2]), caption = "Invalid Barcodes", color = "warning", icon = "fa-x")
```

### noBX
```{r}
valueBox(scales::comma(tots[tots$valid == "no BX", 2]), caption = "Missing Barcodes", color = "danger", icon = "fa-exclamation")
```

### glob-avg
```{r}
valueBox(scales::comma(avgBX), caption = "Average BX per contig", color = "info")
```

### glob-total
```{r}
valueBox(scales::comma(totuniqBX), caption = "Total unique barcodes", color = "info")
```

## Reads per molecule dec
### Reads per mol dec {.no-title}
<h2> Reads per Molecule </h2>
The chart below shows the distribution of the number of reads per haplotag molecule. That is, how many alignments
are associated with a particular haplotag barcode. This excludes the number of reads associated with invalid
or absent haplotag barcodes.

### bases per desc {.no-title}
<h2> Total Bases Aligned </h2>
Below is a histogram showing the distribution of the number of base pairs aligned
per unique barcode per contig. These data are shown in 500 bp bins, _i.e._ the Y-value
at `X = 0` would represent the number of barcodes with 0-500bp aligned.


## Reads per molecule
### Reads per mol {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE, out.width = '100%'}
p <- filter(tb, valid == "valid BX") %>% 
  ggplot(aes(x = reads)) +
    geom_histogram(fill = "#9393d2", color = "#8484bd", binwidth = 2, boundary = 0.5) +
    labs(title = "Alignments Per Barcoded Molecule", subtitle = samplename) +
    xlab("Alignments per BX molecule") +
    ylab("Number of barcodes") +
    scale_y_continuous(labels = scales::comma) +
    theme_light() +
    xlim(c(1, NA)) +
    theme(
        panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank()
    )

ggplotly(p)

```

### Bases Per Barcode {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE, out.width = '100%'}
p <- tb %>% filter(valid == "valid BX", length_inferred > 10) %>%
    ggplot(aes(x = aligned_bp)) +
        geom_histogram(binwidth = 500, fill = "#99cccc", color = "white", boundary = 0.5) +
        labs(title = "Bases Aligned Per Barcode", subtitle = samplename) +
        xlab("Number of aligned bases (500 bp bins)") +
        ylab("Number of occurrences") +
        theme_light() +
        scale_y_continuous(labels = scales::comma) +
        scale_x_continuous(labels = scales::comma) +
        theme(
            panel.grid.minor.y = element_blank(),
        )
ggplotly(p)
```
## inferred-header
### inferred desc {.no-title}
<h2> Inferred Molecule Lengths </h2>
The chart below shows the distribution of molecule lengths (in 2500bp bins) 
inferred from the first and last alignment positions along a contig for all
alignments associated with a single haplotag barcode on a given contig.

## Inferred
### Inferred molecule Lengths
```{r echo = FALSE, message = FALSE, warning = FALSE, out.width = '100%'}
p <- tb %>% filter(valid == "valid BX", length_inferred > 10) %>%
    ggplot(aes(x = length_inferred)) +
        geom_histogram(binwidth = 2500, fill = "#70a7d9", boundary = 0.5) + #, color = "white") +
        labs(title = "Inferred Haplotagged Molecule Lengths", subtitle = samplename) +
        xlab("Inferred Length of Molecule (2,500 bp bins)") +
        ylab("Number of occurrences") +
        theme_light() +
        scale_y_continuous(labels = scales::comma) +
        scale_x_continuous(labels = scales::comma) +
        theme(
            panel.grid.minor.y = element_blank(),
        )
ggplotly(p)
```

### A table of some kind
pretty sure there is a table that belongs here

## Interpreting the supporting data
### interp desc {.no-title}
<h2> Interpreting the Data </h2>
Below are details on how to interpret the information presented in this report, as well as the 
underlying data used to create this report.

## inttable
### interpreting {.no-title}
<h3> Interpreting the Supporting File </h3>
Listed below are the descriptions of the columns in **`r basename(infile)`**, which was created
by Harpy using the included `bxStats.py` script.

- `contig`: name of the contig the barcode occurs on 
- `bx`:  the barcode 
- `reads`:  number of alignments with this barcode 
- `start`:  the start position of the first alignment for that barcode on that contig 
- `end`: the end position of the last alignment for that barcode on that contig 
- `length_inferred`: inferred length of the molecule based on the start/end of the alignments with that barcode 
- `aligned_bp`: number of total bp aligned for that barcode 
- `mindist`: the minimum basepair distance between two alignments sharing a barcode (excluding read pairs)

## Interpreting this report
### Barcode validity {.no-title}
<h3> Interpreting Barcode Validity </h3>
BX barcode validity is classified into one of three categories:

- `valid BX`: a complete BX barcode was present in the read (i.e. no `00` for any segments)
- `invalid BX`: A barcode was present in the read, but it contained `00` for any of the haplotag segments
- `no BX`: no barcode was present in the read

### Molecule splitting {.no-title}
<h3> Molecule Splitting, Explained </h3>

**Split threshold: `r mdist`**

It's common for a barcode shared by reads not originating from the same molecule
to reappear much further along a chromosome or across multiple chromosomes. The 
process that derives the data in this report separates those recurring barcodes
as unique molecules when their distance is greater than `r mdist` base pairs along a single contig.
The "split" molecules are identified by a numerical suffix in their name, such
as the `.1` in `A01C22B96D11.1`. These suffixes were added only for reporting purposes and 
the barcodes in your alignment file **were not** altered. If aligned with BWA, Harpy added
a corresponding `MI:i:` (Molecular Identifier) tag that reflects these splits given
the molecule distance threshold you specified. If aligned with EMA, the EMA software
itself determines the splits and assigns the `MI:i` tag without user specification.
