---
title: "Harpy Alignment Barcode Summary"
date: "`r format(Sys.time(), '%m-%d-%y %X')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r imports, echo = F, results = F, message = F}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}

using("flexdashboard","dplyr","tidyr", "highcharter","DT")
```

```{r nxx_and_process_funs, echo = F, results = F, message = F}
NX <- function(lengths, X=50){
  lengths <- as.numeric(sort(lengths, decreasing=TRUE))
  index <- which(cumsum(lengths) / sum(lengths) >= X/100)[1]
  return(lengths[index])
}

std_error <- function(x){
  sd(x)/sqrt(length((x)))
}

process_input <- function(infile){
  bamfile <- gsub(".bxstats.gz", ".bam", infile)
  samplename <- gsub(".bxstats.gz", "", basename(infile))
  tb <- read.table(infile, header = T, sep = "\t") %>% select(1,2,3,6)
  tb$valid <- tb$molecule
  tb[tb$valid != "invalidBX", "valid"] <- "validBX"
  tb$valid <- gsub("BX", " BX", tb$valid)

  nMolpercont <- group_by(tb[tb$valid == "valid BX",], contig) %>% 
    summarize(nMol = length(molecule))
    
  avg_mol <- round(mean(nMolpercont$nMol), digits = 2)
  sem_mol_per_cont <- round(std_error(nMolpercont$nMol), digits = 3)
  tot_uniq_bx <- system(paste("zcat",infile, "| tail -1"), intern = T)
  tot_uniq_bx <- gsub("#total unique barcodes: ", "", tot_uniq_bx) |> as.integer()

  tot_mol <- sum(tb$valid == "valid BX")
  tot_valid_reads <- sum(tb[tb$valid == "valid BX", "reads"])
  avg_reads_per_mol <- round(tot_valid_reads/tot_mol,1)
  sem_reads_per_mol <- round(std_error(tb[tb$valid == "valid BX", "reads"]), 1)
  tot_invalid_reads <- sum(tb[tb$valid == "invalid BX", "reads"])
  n50 <- NX(tb$length_inferred, 50)
  n75 <- NX(tb$length_inferred, 75)
  n90 <- NX(tb$length_inferred, 90)

  outrow <- data.frame(
    sample = samplename,
    totalreads = tot_valid_reads + tot_invalid_reads,
    totaluniquemol = tot_mol,
    totaluniquebx = tot_uniq_bx,
    molecule2bxratio = round(tot_mol / tot_uniq_bx,2),
    totalvalidreads = tot_valid_reads,
    totalinvalidreads = tot_invalid_reads,
    totalvalidpercent = round(tot_valid_reads/(tot_valid_reads + tot_invalid_reads) * 100,2),
    averagemolpercont = avg_mol,
    semolpercont = sem_mol_per_cont,
    averagereadspermol = avg_reads_per_mol,
    sereadspermol = sem_reads_per_mol,
    N50 = n50,
    N75 = n75,
    N90 = n90
  )
  return(outrow)
}
```

```{r setup_df, echo = F, results = F, message = F}
#infiles <- snakemake@input[[1]]
infiles <- c("~/sample1.bxstats.gz", "~/sample2.bxstats.gz")

aggregate_df <- data.frame(
  sample = character(),
  totalreads = integer(),
  totaluniquemol = integer(),
  totaluniquebx = integer(),
  molecule2bxratio = numeric(),
  totalvalidreads = integer(),
  totalinvalidreads = integer(),
  totalvalidpercent = numeric(),
  averagemolpercont = integer(),
  semolpercont = numeric(),
  averagereadspermol = numeric(),
  sereadspermol = numeric(),
  N50 = integer(),
  N75 = integer(),
  N90 = integer()
)

for(i in infiles){
  aggregate_df <- rbind(aggregate_df, process_input(i))
}
```

## General Barcode Information from Alignments
### Desc {.no-title}
<h2> General Barcode Information from Alignments </h2>
This report aggregates the barcode-specific information from the alignments
that were created using `harpy align`. Detailed information for any one sample
can be found in that sample's individual report. The table below is an aggregation
of data for each sample based on their `*.bxstats.gz` file.

```{r}
DT::datatable(
  aggregate_df,
  rownames = F,
  extensions = 'Buttons', 
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  fillContainer = T,
  colnames = c("sample", "alignments", "unique molecules", "unique barcodes", "molecule:bx ratio", "valid bx alignments", "invalid bx alignments", "% valid bx", "avg molecules/contig", "SEM molecules/contig", "avg reads/molecule", "SEM reads/molecule", "N50", "N75","N90")
)
```

```{r sampleplot_height, echo=FALSE, message=FALSE, warning=FALSE}
n_samples <- nrow(aggregate_df)
plotheight <- 150 + (15 * n_samples)
figheight <- 1 + (0.2 * n_samples)
```

```{r nxxplot, echo = F, results = F, message = F}
highchart() |>
  hc_chart(type = "area") |>
  hc_add_series(data = density(aggregate_df$N50), name = "N50", type = "areaspline") |>
  hc_add_series(data = density(aggregate_df$N75), name = "N75", type = "areaspline") |>
  hc_add_series(data = density(aggregate_df$N90), name = "N90", color = "#d3805f", type = "areaspline") |>
  hc_tooltip(enabled = FALSE) |>
  hc_title(text = "NX Stats Across Samples") |>
  hc_xAxis(title = list(text = "NX value"), min = 0) |>
  hc_yAxis(title = list(text = "density")) |>
  hc_exporting(enabled = T, filename = "NX.stats",
    buttons = list(contextButton = list(menuItems = c("downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```

```{r perc_valid_dist, echo = F, results = F, message = F}
hs <- hist(
  aggregate_df$totalvalidpercent,
  breaks = min(aggregate_df$totalvalidpercent):max(aggregate_df$totalvalidpercent),
  plot = F
)
hs$counts <- round(hs$counts / sum(hs$counts) * 100, 4)
hs <- data.frame(val = hs$breaks[-1], freq = hs$counts)

hchart(hs, "areaspline", hcaes(x = val, y = freq), color = "#8484bd", name = "Percent Alignments", marker = list(enabled = FALSE)) |>
  hc_title(text = "Percent of Alignments with Valid BX tags") |>
  hc_xAxis(min = 0, max = 100, title = list(text = "% alignment with valid BX tag")) |>
  hc_yAxis(max = 100, title = list(text = "% samples")) |>
  hc_tooltip(crosshairs = TRUE) |>
  hc_exporting(enabled = T, filename = "percent.valid.dist",
    buttons = list(contextButton = list(menuItems = c("downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```

```{r perc_valid_per_sample, echo = F, results = F, message = F, fig.height=figheight, out.width="100%"}
hchart(aggregate_df, "xrange", pointPadding = 0.1, groupPadding = 0.0001,
  hcaes(x = 0, x2 = totalreads, y = 0:(n_samples-1), partialFill = totalvalidpercent/100), dataLabels = list(enabled = TRUE)) |> 
  hc_xAxis(title = list(text = "Total Alignments")) |> 
  hc_yAxis(title = FALSE, gridLineWidth = 0, categories = aggregate_df$sample) |>
  hc_caption(text = "Percentage represents the percent of alignments with a valid BX barcode") |>
  hc_tooltip(enabled = FALSE) |>
  hc_colors(color = "#95d8a7") |>
  hc_title(text = "Percent Alignments with Valid BX Barcode") |>
  hc_exporting(enabled = T, filename = "BX.valid.alignments",
    buttons = list(contextButton = list(menuItems = c("downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```
```{r mol_per_contig, echo = F, results = F, message = F}
# Create a scatter plot with horizontal error bars
err_df <- data.frame(x = 0:n_samples-1), y = aggregate_df$averagemolpercont, low = aggregate_df$averagemolpercont - aggregate_df$semolpercont, high = aggregate_df$averagemolpercont + aggregate_df$semolpercont)
highchart() |>
  hc_chart(inverted=TRUE, pointPadding = 0.1, groupPadding = 0.0001) |>
  hc_add_series(data = aggregate_df, type = "scatter", name = "mean", hcaes(x = 0:(nrow(aggregate_df)-1), y = averagemolpercont), marker = list(radius = 8), color = "#c26d9b", zIndex = 1) |>
  hc_add_series(data = err_df, type = "errorbar", name = "standard deviation", linkedTo = ":previous", zIndex = 0, stemColor = "#c26d9b", whiskerColor = "#c98fae") |>
  hc_xAxis(title = FALSE, gridLineWidth = 0, categories = aggregate_df$sample) |>
  hc_title(text = "Average Molecules Per Contig") |>
  hc_caption(text = "Error bars show the standard error of the mean") |>
  hc_tooltip(crosshairs = TRUE, pointFormat= '<b>{point.y}</b>') |>
  hc_exporting(enabled = T, filename = "avg.molecules.per.contig",
    buttons = list(contextButton = list(menuItems = c("downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```

```{r reads_per_mol_sample, echo = F, results = F, message = F, fig.height=figheight, out.width="100%"}
err_df <- data.frame(x = 0:(n_samples-1), y = aggregate_df$averagereadspermol, low = aggregate_df$averagereadspermol - aggregate_df$sereadspermol, high = aggregate_df$averagereadspermol + aggregate_df$sereadspermol)
highchart() |>
  hc_chart(inverted=TRUE, pointPadding = 0.0001, groupPadding = 0.0001) |>
  hc_add_series(data = aggregate_df, type = "scatter", name = "mean", hcaes(x = 0:(nrow(aggregate_df)-1), y = averagereadspermol), marker = list(radius = 8), color = "#6d73c2", zIndex = 1) |>
  hc_add_series(data = err_df, type = "errorbar", name = "standard deviation", linkedTo = ":previous", zIndex = 0, stemColor = "#8186c7", whiskerColor = "#8186c7") |>
  hc_xAxis(title = FALSE, gridLineWidth = 0, categories = aggregate_df$sample) |>
  hc_title(text = "Average Reads Per Molecule") |>
  hc_caption(text = "Error bars show the standard error of the mean") |>
  hc_tooltip(crosshairs = TRUE, pointFormat= '<b>{point.y}</b>') |>
  hc_exporting(enabled = T, filename = "avg.reads.per.molecule",
    buttons = list(contextButton = list(menuItems = c("downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```