---
title: "Haplotype Phasing Report"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r echo= FALSE, message = FALSE, warning = FALSE}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr","ggplot2","DT","scales", "highcharter","ggridges")

#library(flexdashboard)
#library(dplyr)
#library(ggplot2)
#library(DT)
#library(scales)
#library(ggridges)
```

```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.7.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r echo= FALSE, message = FALSE, warning = FALSE}
file <- "/home/pdimens/Documents/harpy/Phase/reports/blocks.summary.gz"
#file <- snakemake@input[[1]]
df <- read.table(
  file, header = T,
  colClasses = c("factor","factor","integer","integer", "integer")
)
if(nrow(df) < 1){
  cat(paste("No phase blocks were observed in the input file", file, "\n"))
  knitr::knit_exit()
}
df$pos_end <- df$pos_start + df$block_length
df <- df[, c(1,2,3,4,6,5)]
levels(df$sample) <- basename(levels(df$sample))
```

```{r echo = FALSE, message = FALSE, warning = FALSE}
# Find the 30 largest contigs
contigs <- group_by(df, contig) %>%
  summarize(size = max(pos_end)) %>%
  arrange(desc(size))
# limit the data to only the 30 largest contigs
if (nrow(contigs) > 30){
    .contigs <- contigs[1:30, ]$contig
} else {
    .contigs <- contigs$contig
}
#pltheight <- round(1.2 * (length(.contigs)), digits = 0)
ridgeheight <- 1 + round(0.7 * (length(.contigs)), digits = 0)
#pltheight.samples <- <- round(1.2 * (length(levels(df$sample))), digits = 0)
ridgeheight.samples <- 1 + round(0.7 * (length(levels(df$sample))), digits = 0)
```

# General Stats
## fileheader 
### details desc {.no-title}
<h1> Haplotype Phasing Results </h1>
The data presented here are the results of phasing genotypes into haplotypes using
[HapCut2](https://github.com/vibansal/HapCUT2). The information is derived from `r file`,
which summarized information across all samples using the `.blocks` files HapCut2
generates. The VCF files HapCut2 also generates were not used as they lack
some of the information in the `.blocks` files that were collated in this report.
This page shows general and per-contig information. You may click the plots to
expand them in your browser. Clicking them again will exit from the zoomed pop-up
window. The `Per-Sample Stats` tab in the navigation bar above will show you
statistics relating to haplotypes on the per-sample level.

## Haplotyping details {data-height=100}
```{r echo= FALSE, message = FALSE, warning = FALSE}
overall <- df %>% summarise(
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
)
```
### avg snps
```{r}
valueBox(scales::comma(overall$mean_snps[1]), caption = "Avg SNPs/Haplotype", color = "info")
```

### median snps
```{r}
valueBox(scales::comma(overall$median_snps[1]), caption = "Median SNPs/Haplotype", color = "info")
```
### avg blocksize
```{r}
valueBox(scales::comma(overall$mean_blocksize[1]), caption = "Avg Haplotype Length", color = "info")
```

### median blocksize
```{r}
valueBox(scales::comma(overall$median_blocksize[1]), caption = "Median Haplotype Length", color = "info")
```

### avg snps
```{r}
valueBox(scales::comma(overall$max_blocksize[1]), caption = "Longest Haplotype (bp)", color = "info")
```

## distribution plots
### haplotype length {.no-title}
<h2> Distribution of Haplotype Lengths </h2>
The plots below show the distribution of haplotype length (in base pairs). 
Phasing typically has a few very large haplotypes, resulting in extreme 
right-tails in these distributions, therefore, the X-axis is **log-scaled lengths**
to collapse the right tail for better readability. 

## dist plot
### the dist plpt {.no-title}
```{r warning=FALSE, message=FALSE, echo= FALSE, out.width="100%"}
hs <- hist(
  df$block_length,
  breaks = 500,
  plot = F
)
hs <- data.frame(val = hs$mids, freq = hs$counts)
hchart(hs, "areaspline", hcaes(x = val, y = freq), color = "#7eb495", name = "count") |>
  hc_xAxis(title = list(text = "haplotype length in base pairs (log scale)"), type = "logarithmic") |>
  hc_title(text = "Haplotype block lengths") |>
  hc_yAxis(title = list(text = "count")) |>
  hc_tooltip(crosshairs = TRUE) |>
  hc_exporting(enabled = T, filename = "haplotype.hist",
    buttons = list(contextButton = list(menuItems = c("viewFullscreen", "separator", "downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```

## Per contig data desc
### table per contig {.no-title}
<h2> Stats Per Contig </h2>
The plot below shows the distribution of haplotype length (in base pairs) for 
each contig, up to 30 of the largest contigs, the size of whom is inferred/assumed
from the end position of the last haplotype. Phasing typically results in extreme
right-tails in these distributions, therefore the plot presented below has **log-scaled lengths**
to collapse the right tails for better readability. 
The table below provides details on the outcome of haplotype phasing on a per-contig basis.

## Per contig data
### per contig plot {.no-title}
```{r warning=FALSE, message=FALSE, echo= FALSE, fig.height=ridgeheight, fig.width=8}
ggplot(df, aes(x = block_length, y = contig, fill = after_stat(x))) +
  geom_density_ridges_gradient(scale = 1.4) +
  scale_fill_viridis_c(option = "C", trans = "log10") +
  theme_minimal() +
  theme(legend.position = "none") +
  scale_x_log10(
     breaks = trans_breaks("log10", function(x) 10^x),
     labels = trans_format("log10", math_format(10^.x))
  ) +
  labs(title = "Distribution of Haplotype Lengths by Contig", fill = "Haplotype Length (bp)") +
  xlab("Haplotype Length (log scale)") +
  ylab("")
```

### stats per contig {.no-title}
```{r echo= FALSE, message = FALSE, warning = FALSE, out.width="100%"}
percontig <- df %>% group_by(contig) %>% summarise(
    n_haplo = round(length(n_snp)),
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
  )

DT::datatable(
  percontig,
  rownames = F,
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE, pageLength = length(levels(df$contig))),
  colnames = c("Contig", "Total Haplotypes", "Mean SNPs", "Median SNPs", "Mean Haplotype Length", "Median Haplotype Length", "Largest Haplotype"),
  fillContainer = F
  )
```

# Per-Sample Stats
## per sample desc
### stats per sample desc {.no-title}
<h1> Per-Sample Stats </h1>
Haplotype phasing occurs per-sample and the table below provides details on the outcome of haplotype phasing for each sample.

The ridgeline plot below shows the distribution of haplotype lengths (in base pairs) for each 
sample. While no Y-axis for these counts are provided, this plot is intended to be more of
a visual reference of the relative distributions of haplotype lengths among samples.
Phasing typically results in extreme right-tails in these distributions, making regular 
visualization difficult to plot meaningfully. The plot presented below has 
**log-scaled lengths** to collapse the right tails for better readability. 

## per sample 
### the ridgeplot {.no-title}
```{r warning=FALSE, message=FALSE, echo= FALSE, fig.height=ridgeheight.samples, fig.width=8}
ggplot(df, aes(x = block_length, y = sample, fill = after_stat(x))) +
  geom_density_ridges_gradient(scale = 1.4) +
  scale_fill_viridis_c(option = "G", trans = "log10") +
  theme_minimal() +
  theme(legend.position = "none") +
  scale_x_log10(
     breaks = trans_breaks("log10", function(x) 10^x),
     labels = trans_format("log10", math_format(10^.x))
  ) +
  labs(title = "Distribution of Haplotype Lengths by Sample", fill = "Haplotype Length (bp)") +
  xlab("Haplotype Length (log scale)") +
  ylab("")
```

### stats per sample {.no-title}
```{r echo= FALSE, message = FALSE, warning = FALSE, out.width="100%"}
persample <- df %>% group_by(sample) %>% summarise(
    n_haplo = sum(length(n_snp)),
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
)

DT::datatable(
  persample,
  rownames = F,
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE, pageLength = length(levels(df$sample))),
  colnames = c("Sample", "Haplotypes", "Mean SNPs", "Median SNPs", "Mean Haplotype Length", "Median Haplotype Length", "Largest Haplotype"),
  fillContainer = F
)
```