---
title: "Haplotype Phasing Report"
author:
    - "Created using HARPY"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    highlight: tango
    number_sections: no
    theme: cosmo
    df_print: kable
    toc: no
---

```{r echo = F, message = F, warning = F}
library(ggplot2)
library(plotly)
library(dplyr)
library(scales)
library(GenomicRanges)
library(DT)
```
```{r echo = F, message = F, warning = F}
#infile <- snakemake@input[[1]]
infile <- "~/ZS10.bx.haplosummary"
bx <- read.table(infile, header = T)
bx$size <- bx$end - bx$start
bx$capsize <- ifelse(bx$size > 50000, 50000, bx$size)
```
```{r echo = FALSE, warnings = FALSE, message = FALSE}
#fai <- snakemake@input[[2]]
fai <- "~/assembly.fai"
fa.sizes <- read.table(fai, header = F) |> arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")
# limit the data to only the 30 largest contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]
```
## Haplotype lengths {.tabset}
### Size frequency
Below is a distribution of the binned lengths of haplotypes in base pairs (bp).

```{r echo = FALSE, warnings = FALSE, message = FALSE}
bxhist <- hist(x = bx$size, breaks = 200, plot = F)
hists <- data.frame("sizebin" = bxhist$breaks[-1], "count" = bxhist$counts)
hists <- hists[hists$count > 0, ]
DT::datatable(
  hists, 
  rownames = F, 
  extensions = 'Buttons', 
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  colnames = c("size bin", "count")
)
```
### Plot
Below is a plot of the distribution of the lengths of haplotypes in base pairs (bp). If you have many
larger haplotypes with only a few representatives, this plot may be somewhat challenging to see
the right tail of the distribution. If that's the case, the truncated distribution may be more
helpful to understand the distribution of haplotype sizes. Notice that the size of the bins, which
is configured to a fixed number of bins (200) will vary from the truncated plot (100 bins and smaller
numeric range). The plots are interactive and you can zoom in to areas of interest by dragging the cursor.
```{r echo = FALSE, warnings = FALSE, message = FALSE, out.width = "100%"}
d_ <- ggplot(hists, aes(x = sizebin, y = count)) +
  geom_bar(stat = "identity", fill = "#893A70", color = "#893A70") +
  theme_minimal() +
  theme(
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  xlab("Haplotype block size (bp)") +
  ylab("Count") +
  labs(title = "Frequency of haplotype lengths") +
  scale_x_continuous(labels = scales::comma) +
  scale_y_continuous(labels = scales::comma)
ggplotly(d_, height = 350) |> layout(hovermode = 'x')
```

### Plot (truncated)
Below is a **truncated** distribution of the lengths of haplotypes in base pairs (bp). Haplotypes
that are 50,000 base pairs or greater are all consolidated into `50000`, which should hopefully
make it easier to assess the size bins below 50,000.The plots are interactive and you can zoom in
to areas of interest by dragging the cursor. 
```{r echo = FALSE, warnings = FALSE, message = FALSE, fig.height = 2.5, out.width = "100%"}
capbxhist <- hist(x = bx$capsize, breaks = 100, plot = F)
caphists <- data.frame("sizebin" = capbxhist$breaks[-1], "count" = capbxhist$counts)
caphists <- caphists[caphists$count > 0, ]


d_t <- ggplot(caphists, aes(x = sizebin, y = count)) +
  geom_bar(stat = "identity", fill = "#893A70", color = "#893A70") +
  theme_minimal() +
  theme(
    panel.grid.minor.y = element_blank(),
    legend.position = "none"
  ) +
  xlab("Haplotype block size (bp)") +
  ylab("Count") +
  labs(title = "Frequency of haplotype lengths (truncated)") +
  scale_x_continuous(labels = scales::comma) +
  scale_y_continuous(labels = scales::comma)

ggplotly(d_t, height = 350) |> layout(hovermode = 'x')
```

```{r echo = FALSE, warnings = FALSE, message = FALSE}
gr <- GRanges(
  seqnames = bx$contig,
  ranges = IRanges(start = bx$start, end = bx$end)
)
bx <- bx[(bx$contig %in% fa.sizes$contig) & (bx$n_snps > 1), ]
bx$offset <- disjointBins(gr)
pltheight <- round(1.3 * (nrow(fa.sizes)), digits = 0)
```

## Phased haplotypes
Below is a series of plots showing the positions and lengths of the phased haplotypes in the data.
The haplotypes are colored according to the number of SNPs that comprise them. Each plot represents
a contig in the assembly used for alignment, up to the 30 longest contigs in the assembly. Haplotypes with fewer than 2 supporting SNPs have been removed from the plots. Haplotypes with overlaps have been vertically offset for visual clarity.
```{r echo = FALSE, warnings = FALSE, message = FALSE, fig.height = pltheight, out.width = "100%"}
ggplot(bx) +
geom_segment(
  linewidth = 1,
  lineend = "round",
  aes(x = start, xend = end, y = offset, yend = offset, color = n_snps)
) +
theme_light() +
facet_wrap(~contig, ncol = 1, scales = "free") +
theme(
  axis.line.x = element_blank(),
  axis.ticks.x = element_blank(),
  axis.ticks.y = element_blank(),
  axis.text.y = element_blank(),
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  strip.background = element_rect(fill = "grey80"),
  strip.text = element_text(color = "grey20"),
  legend.position = "right"
) +
scale_colour_gradient(labels = comma,low = "dodgerblue", high = "indianred") +
scale_x_continuous(labels = comma) +
xlab("Position (bp)") +
ylab("") +
labs(color = "Supporting SNPs")
```