---
title: "Leviathan SV Calling Summary"
date: "`r format(Sys.time(), '%d %b, %Y at %H:%M')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    mathjax: NULL
    logo: https://raw.githubusercontent.com/pdimens/harpy/docs/static/logo_report.png
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_report.png"
    navbar:
      - { title : "Docs", icon: "fa-book", href: "https://pdimens.github.io/harpy/", align: right }
      - { title : "Source", icon: "fa-github", href: "https://www.github.com/pdimens/harpy/", align: right }
---

```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr","ggplot2","DT","BioCircos")
```

## Intro
### introtext {.no-title}
```{r}
fai <- snakemake@input[["faidx"]]
statsfile <- snakemake@input[["statsfile"]]
samplename <- gsub(".stats", "", basename(statsfile))
```

<h1> Structural Variants: `r samplename`</h1>
Below is a series of tables and plots detailing the structural variants
identified by [LEVIATHAN](https://github.com/morispi/LEVIATHAN) [(preprint)](https://doi.org/10.1101/2021.03.25.437002). The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE}
knitr::kable(
  data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
)
```

This tab (`General Stats`) shows overview information. Clicking `Per-Contig Plots`
in the navigation bar at the top of this page will show you interactive
plots detailing all variants identified.

```{r}
sv <- read.table(statsfile, header = T,
  colClasses = c("factor", "factor", "integer", "integer", "character", "factor", "integer", "integer")
)
if (nrow(sv) == 0) {
  cat(paste0("There are no variants in the file ", "`", bcf, "`"))
  knitr::knit_exit()
}
```

## General Information {data-height=100}
```{r summaryinfo, echo=FALSE, message=FALSE, warning=FALSE}
summinfo <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n()) %>% 
  group_by(type) %>%
  summarise(
    total = sum(count),
    avg_per_contig = round(mean(count), 2),
    sd = round(sd(count), 2)
)

nvar = sum(summinfo$total)
```
### nvariants
```{r}
valueBox(scales::comma(nvar), caption = "Variants", color = ifelse(nvar > 0, "success", "warning"))
```

### nbreakends
```{r}
nbnd <- ifelse("BND" %in% summinfo$type, summinfo$type[which(summinfo$type == "BND")], 0)
valueBox(scales::comma(nbnd), caption = "Breakends", color = "info")
```
### ndeletions
```{r}
ndel <- ifelse("DEL" %in% summinfo$type, summinfo$type[which(summinfo$type == "DEL")], 0)
valueBox(scales::comma(ndel), caption = "Deletions", color = "info")
```

### nduplications
```{r}
ndup <- ifelse("DUP" %in% summinfo$type, summinfo$type[which(summinfo$type == "DUP")], 0)
valueBox(scales::comma(ndup), caption = "Duplications", color = "info")
```

### ninversions
```{r}
ninv <- ifelse("INV" %in% summinfo$type, summinfo$type[which(summinfo$type == "INV")], 0)
valueBox(scales::comma(ninv), caption = "Inversions", color = "info")
```

## Various Stats Headers
### summary info {.no-title}
<h2> Variant Information </h2>
Below are two tables with information on the variants. The first (left) table
details all the variants LEVIATHAN detected and the quantity/quality of evidence
for those detections. The second (right) table details each type of structural variant for
every contig. 

## Various Stats Tables
### All variants found
```{r various stats, echo = FALSE, warnings = FALSE, message = FALSE}
DT::datatable(
  sv,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = paste0(samplename , "_sv_leviathan`"))),  
    scrollX = TRUE
  ),
  autoHideNavigation = T,
  fillContainer = T
)
```

### Per-contig info
```{r echo = FALSE, message = FALSE}
sv$length <- gsub("\\.", "1", sv$length) %>%  as.numeric()
```

```{r percontig, echo = FALSE, message = FALSE, warning = FALSE}
grpstats <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n(), total_bp  = sum(length))

DT::datatable(
  grpstats, 
  rownames = F, 
  filter = "top", 
  extensions = 'Buttons', 
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = paste0(samplename ,"_sv_leviathan_per_contig"))),
    scrollX = TRUE
  ),
  autoHideNavigation = T,
  fillContainer = T
)
```

## SV size and location
### legend {.no-title data-width=200}
To the right is a circular plot to visualize the distribution of structural variants
across (up to) 30 of the largest contigs. This should help you assess
the presence/absence of expected variants. The legend below shows which
colors correspond to which type of variant.
You may click on the plot to expand it in your browser window. Clicking it again will exit
out of the zoomed view.

```{r plot legend, echo = FALSE, warning = FALSE, message = FALSE}
color.palette <- c("deletion" = "#5a8c84", "duplication" = "#99278a", "inversion" = "#4a9fea", "breakend" = "#c56e34")

data.frame(x = 1, y = 1:4, colour = c("breakend","deletion","duplication", "inversion")) %>% 
ggplot(aes(x, y, fill = colour))+
  geom_point(alpha=0, shape = 22, color = "white")+ # completely transparent rectangular point 
  scale_fill_manual(values=color.palette, drop=FALSE) +
  guides(fill = guide_legend(override.aes = list(alpha=1, size = 25)))+ # showing the point in the legend
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "inside",
        legend.position.inside = c(0.3, 0.5),
        legend.text = element_text(size = 25),
        legend.title=element_text(size=25),
        legend.key = element_rect(fill='NA'),
        panel.grid = element_blank(),
        panel.background = element_rect(fill='transparent'), 
        plot.background = element_rect(fill='transparent', color=NA),
        legend.background = element_rect(fill='transparent'),
        legend.box.background = element_rect(fill='transparent', color = "transparent"), 
  ) +
  labs(fill="Variant Type")
```

### Structural Variants: `r samplename` (save by taking a screenshot)
```{r filter_contigs}
fa.sizes <- read.table(fai, header = F) %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")

# make sure that the contigs with SV are the ones being plotted, not the others
fa.sizes <- fa.sizes[fa.sizes$contig %in% unique(sv$contig), 1:2]

# limit the data to only the 30 of the largest present contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]

# filter out variants that aren't on one of these contigs
sv <- sv[sv$contig %in% fa.sizes$contig, ]
sv$contig <- factor(sv$contig)
```


```{r interactive_circos, out.width= "80%", out.height="900px"}
genomeChr <- fa.sizes$size
names(genomeChr) <- fa.sizes$contig
genomeChr <- as.list(genomeChr)
tracks <- BioCircosTracklist()
radius <- 0.3

# Add one track for each SV type
# inversions
inv <- sv[sv$type == "INV",]
if(nrow(inv) > 0){
  tracks <- tracks + BioCircosArcTrack(
    'Inversions',
    as.character(inv$contig),
    inv$position_start,
    inv$position_end,
    opacities = 0.7,
    values = inv$n_barcodes,
    minRadius = 0.78,
    maxRadius = 0.87,
    colors = color.palette["inversion"],
    labels = "Type: Inversion",
  )
}

# deletions
del <- sv[sv$type == "DEL",]
if(nrow(del) > 0){
  tracks <- tracks + BioCircosSNPTrack(
    'Deletions',
    as.character(del$contig),
    del$position_start,
    values = abs(del$position_end - del$position_start),
    minRadius = 0.63,
    maxRadius = 0.71,
    size = 1,
    shape = "rect",
    opacities = 0.7,
    colors = color.palette["deletion"],
    labels = "Type: Deletion"
  )
}


# duplications
dup <- sv[sv$type == "DUP",]
if(nrow(dup) > 0){
  tracks <- tracks + BioCircosArcTrack(
    'Duplications',
    as.character(dup$contig),
    dup$position_start,
    dup$position_end,
    values = dup$n_barcodes,
    opacities = 0.7,
    minRadius = 0.48,
    maxRadius = 0.57,
    colors = color.palette["duplication"],
    labels = "Type: Duplication",
  )
}

# breakends
bnd <- sv[sv$type == "BND",]
if(nrow(bnd) > 0){
  tracks <- tracks + BioCircosSNPTrack(
    'Breakends',
    as.character(bnd$contig),
    bnd$position_start,
    values = 0,
    minRadius = 0.33,
    maxRadius = 0.42,
    size = 1,
    shape = "rect",
    opacities = 0.7,
    colors = color.palette["breakend"],
    labels = "Type: Breakend"
  )
}

# loop to create backgrounds
for(i in c("inversion", "deletion", "duplication")){
  tracks <- tracks + BioCircosBackgroundTrack(
    paste0(i,"_background"),
    minRadius = radius, maxRadius = radius + 0.15,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius + 0.15
}

BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  SNPMouseOverCircleSize = 2,
  SNPMouseOverTooltipsHtml03 = "<br/>Length: ",
  width = "100%",
  height = "1000px",
  elementId = "circosplot" 
)
```