---
title: "Leviathan Structural Variant Calling Summary"
date: "`r format(Sys.time(), '%d %b, %Y at %H:%M')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    logo: https://raw.githubusercontent.com/pdimens/harpy/docs/static/logo_report.png
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_report.png"
    navbar:
      - { title : "Docs", icon: "fa-book", href: "https://pdimens.github.io/harpy/", align: right }
---

```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr", "tidyr", "ggplot2","DT","plotly","circlize", "viridisLite", "tools")
```

```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.7.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r}
#statsfiles <- c("/home/pdimens/1.sv.stats", "/home/pdimens/2.sv.stats")
statsfiles <- snakemake@input[["statsfiles"]]
#faidx <- "/home/pdimens/test.fai"
faidx <- snakemake@input[["faidx"]]
```

# General Stats
## Intro
### introtext {.no-title}
<h1> Structural Variants Summary </h1>
Below is a series of tables and plots detailing the structural variants
identified by [LEVIATHAN](https://github.com/morispi/LEVIATHAN) [(preprint)](https://doi.org/10.1101/2021.03.25.437002). The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE}
knitr::kable(
  data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
)
```
This tab (`General Stats`) shows overview information. Clicking `Per-Contig Plots`
in the navigation bar at the top of this page will show you interactive
plots detailing all variants identified.

In the colorful summary boxes are the per-population average number of
variants detected by type.
```{r}
readvariants <- function(x){
  read.table(
    x,
    header = T,
    colClasses = c("factor", "factor", "integer", "integer", "character", "factor", "integer", "integer")
  )
}
```

```{r}
emptyfiles <- character()
sv <- data.frame(matrix(ncol=8,nrow=0, dimnames=list(NULL, c("population",	"contig",	"position_start",	"position_end",	"length",	"type",	"n_barcodes",	"n_pairs"))))
for(i in statsfiles){
  .df <- tryCatch(readvariants(i), error = function(e){return(0)})
  if(length(.df)==1){
      emptyfiles <- c(emptyfiles, i)
  } else {
    sv <- rbind(sv, .df)
  }
}
if(length(emptyfiles) == length(statsfiles)){
  cat("There are no variants detected in any of the populations:\n")
  cat(paste0("- ", statsfiles), sep = "\n")
  knitr::knit_exit()
}
if(length(emptyfiles > 0)){
  cat("Some of the populations did not have any variants detected and are omitted from this report:\n")
  cat(paste0("- ", emptyfiles), sep = "\n")
}
```

```{r}
fa.sizes <- read.table(faidx, header = F)[,1:2] %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")

# make sure that the contigs with SV are the ones being plotted, not the others
fa.sizes <- fa.sizes[fa.sizes$contig %in% unique(sv$contig), ]

# limit the data to only the 30 of the largest present contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]
```

```{r echo = FALSE, warnings= FALSE, message = FALSE}
summinfo <- sv %>%
  group_by(population, type) %>%
  summarise(tot = n()) %>%
  group_by(type) %>%
  summarise(count = sum(tot), avg = round(mean(tot, na.rm = T),2 ), sd = round(sd(tot, na.rm = T), 2)) #%>% 
```

## header info
### npops
```{r}
valueBox(scales::comma(length(levels(sv$population))), caption = "Populations", color = "success")
```

### nbreakends
```{r}
nbnd <- ifelse("BND" %in% summinfo$type, summinfo$type[which(summinfo$type == "BND")], 3)
valueBox(scales::comma(nbnd, accuracy = 0.01), caption = "Breakends (average)", color = "info")
```
### ndeletions
```{r}
ndel <- ifelse("DEL" %in% summinfo$type, summinfo$type[which(summinfo$type == "DEL")], 3)
valueBox(scales::comma(ndel, accuracy = 0.01), caption = "Deletions (average)", color = "info")
```

### nduplications
```{r}
ndup <- ifelse("DUP" %in% summinfo$type, summinfo$type[which(summinfo$type == "DUP")], 3)
valueBox(scales::comma(ndup, accuracy = 0.01), caption = "Duplications (average)", color = "info")
```

### ninversions
```{r}
ninv <- ifelse("INV" %in% summinfo$type, summinfo$type[which(summinfo$type == "INV")], 3)
valueBox(scales::comma(ninv, accuracy = 0.01), caption = "Inversions (average)", color = "info")
```


## Various information {.tabset data-height=450}
### all variants
This table details all the variants LEVIATHAN detected and the quantity/quality of evidence for those detections.
```{r all variants, message = FALSE, echo = FALSE}
DT::datatable(
  sv,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  fillContainer = T
  )
```

```{r echo = FALSE, warnings= FALSE, message = FALSE}
sv$length <- as.numeric(gsub("\\.", "1", sv$length))
```
### counts per contig
This table details counts each type of structural variant for every contig for every population.
```{r percontig counts, echo = FALSE, warnings= FALSE, message = FALSE}
grpcounts <- sv %>%
  group_by(population, contig, type) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = population, values_from = count)

DT::datatable(
  grpcounts,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  fillContainer = T
  )
```

### bp per contig
This table details the lengths (in bp) of structural variants for every contig for every population. Breakends are not counted because they do not have a size.
```{r bp per contig, echo = FALSE, warnings= FALSE, message = FALSE}
grplens <- sv %>% 
  filter(type != "BND") %>%
  group_by(population, contig, type) %>%
  summarise(bp = sum(length)) %>%
  pivot_wider(names_from = population, values_from = bp)

DT::datatable(
  grplens,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  fillContainer = T
  )
```

---

## SV size and location
### legend {.no-title data-height=700 data-width=200}
The plots to the right are a visual representation of each structural variant where it was detected.
Contigs that are absent did not have any variants detected on them. Within each 
contig, each population is plotted as a row (ring) to help visualize shared
variants between populations. Below is the legend to this plot, where
colors represent populations. Plotting only the first 30 largest contigs.
You may click on the plot to expand it in your browser window. Clicking it again will exit
out of the zoomed view.

```{r circoslegend, fig.width = 6}
populations <- unique(sv$population)
col.ramp <- viridis(length(populations))

pops <- data.frame(x = 1, population = levels(populations))
ggplot(pops, aes(y = population, x = 1, fill = population)) +
  geom_tile(color = "white", lwd = 1.5,linetype = 1) +
  scale_fill_manual(values=col.ramp, drop=FALSE) +
  theme_minimal() +
  theme(
    panel.grid = element_blank(),
    legend.position = "none",
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    axis.title = element_blank()
  ) +
  labs(title = "Population")
```

### circosplot {.no-title}
```{r echo = FALSE, message = FALSE, warning = FALSE, dev = "jpeg"}
circosplot <- function(data, contigdata, svtype, trackheight, poplist, colormap){
  circos.clear()
  col_text <- "grey40"
  circos.par("track.height" = 0.8, gap.degree = 2, cell.padding = c(0, 0, 0, 0))
  circos.initialize(
    factors = contigdata$contig,
    xlim = matrix(c(rep(0, nrow(contigdata)), contigdata$size), ncol = 2)
  )

  # contigs
  circos.track(
    ylim = c(0, 1),
    bg.col = "grey90",
    bg.border = F,
    track.height = 0.06,
    panel.fun = function(x, y) {
      chr = CELL_META$sector.index
      xlim = CELL_META$xlim
      ylim = CELL_META$ylim
      circos.text(
        mean(xlim),
        mean(ylim),
        chr,
        cex = 0.6,
        col = col_text,
        facing="bending.inside",
        niceFacing = T
      )
    }
  )

  # x axis
  brk <- seq(0, 10, 0.5) * 10^7
  circos.track(
    ylim = c(0,1),
    track.index = get.current.track.index(),
    bg.border = F,
    panel.fun = function(x, y) {
      circos.axis(
        h = "top",
        major.at = brk,
        labels = round(brk/10^6, 1),
        labels.cex = 0.6,
        col = col_text,
        labels.col = col_text,
        lwd = 0.7,
        labels.facing = "clockwise"
      )
    }
  )

  # plot the track
  for(i in 1:length(poplist)){
    svquery = data$population == populations[i] & data$type == svtype 
    circos.genomicTrack(
      ylim = c(0,1),
      data = sv[svquery, -1],
      track.height = trackheight,
      bg.border = F,
      panel.fun = function(region, value, ...) {
      circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = colormap[i], border = colormap[i])
      }
    )
  }
  text(0, 0, svtype, cex = 2)
}
```

```{r circosplots, echo = FALSE, message = FALSE, warning = FALSE, dev="jpeg", fig.align='center', fig.width=12, fig.height=12, out.width= "100%"}
par(mfrow = c(2, 2))
popheight <- 5 / length(populations) / 10

for (SV in c("INV", "DEL", "DUP", "BND")){
  circosplot(sv, fa.sizes, SV, popheight, populations, col.ramp)
}
```

# Per-Contig Plots
## Per-contig
### plots per contig {.no-title data-height=10000}
<h1> Structural Variants Per Contig Per Population </h1>
Below is a plot to help you assess what structural variants were
detected by LEVIATHAN. These plots are interactive, allowing you
to hover over a variant to provide additional information, including
the genomic interval in which it occurs and the number of haplotag
barcodes supporting the variant. Populations are stacked vertically
for each contig, denoted by the grey strip on the right side of each row.

```{r echo = FALSE, message = FALSE, warning = FALSE}
sv$ystart <- case_when(
  sv$type == "INV" ~ 0.1,
  sv$type == "DUP" ~ 1.1,
  sv$type == "DEL" ~ 2.1,
  sv$type == "BND" ~ 3.1,
)
sv$ystop <- sv$ystart + 0.8
```

```{r percontig plotly, echo = FALSE, out.width = '100%', warning=FALSE}
color.palette <- c("DEL"="#5a8c84", "DUP"="#ffd75f", "INV"="#4a9fea","BND"="#df487f")
l <- list()
nplotrows <- 0

for (i in 1:nrow(fa.sizes)) {
  sv.filt <- sv %>% filter(contig == fa.sizes$contig[i])
  sv_stats <- group_by(sv.filt, type) %>% summarise(n = length(type))
  if (nrow(sv_stats) == 0) {
    next
  }
  nplotrows <- nplotrows + 1
  plt <- sv.filt %>%
    ggplot() +
    geom_rect(
      alpha = 0.7,
      aes(
        ymin = ystart,
        ymax = ystop,
        xmin = position_start,
        xmax = position_end,
        fill = type,
        color = type,
        text = sprintf("Population: %s<br>Type: %s<br>Position: %s-%s<br>barcodes: %s", population, type, position_start, position_end, n_barcodes)
        )
      ) +
    geom_hline(yintercept = 1:3, color = "grey96") +
    scale_color_manual(values = color.palette) +
    scale_fill_manual(values = color.palette) +
    facet_grid(rows = vars(population)) +
    theme_light() +
    theme(
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey80"),
      strip.text = element_text(color = "grey20"),
      legend.position = "none",
    ) +
    scale_x_continuous(labels = scales::comma) +
    scale_y_continuous(
      breaks = c(0.5, 1.5, 2.5, 3.5),
      labels = c("INV", "DUP", "DEL", "BND"),
      limits = c(0,4)
    ) +
    coord_cartesian(xlim = c(0, fa.sizes$size[i] + 1), expand = F) +
    xlab("Position (bp)") +
    labs(fill = "SV type", color = "SV type")
    annotations <- list( 
      x = 0.5,
      y = 1.05,  
      text = as.character(fa.sizes$contig[i]),  
      xref = "paper",  
      yref = "paper",  
      xanchor = "center",  
      yanchor = "bottom",  
      showarrow = FALSE 
    )
  pp <- subplot(ggplotly(plt, height = 850 * length(populations), tooltip = "text")) %>% 
    layout(annotations = annotations)
    l[[i]] <- pp
}
subplot(l, nrows = nplotrows, shareX = FALSE, titleX = TRUE)
```