---
title: "Leviathan Variant Calling Summary"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
library(flexdashboard)
library(dplyr)
library(DT)
library(BioCircos)
#library(ComplexHeatmap)
#library(ggplot2)
#library(magrittr)
#library(plotly)
#library(circlize)
```
# General Stats
## Intro
### introtext {.no-title}
```{r}
statsfile <- snakemake@input[["statsfile"]]
samplename <- gsub(".sv.stats", "", basename(statsfile))
```

<h1> Structural Variants: `r samplename`</h1>
Below is a series of tables and plots detailing the structural variants
identified by [LEVIATHAN](https://github.com/morispi/LEVIATHAN) [(preprint)](https://doi.org/10.1101/2021.03.25.437002). The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE}
knitr::kable(
  data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
)
```

This tab (`General Stats`) shows overview information. Clicking `Per-Contig Plots`
in the navigation bar at the top of this page will show you interactive
plots detailing all variants identified.

```{r}
sv <- read.table(statsfile, header = T,
  colClasses = c("factor", "factor", "integer", "integer", "character", "factor", "integer", "integer")
)
if (nrow(sv) == 0) {
  cat(paste0("There are no variants in the file ", "`", bcf, "`"))
  knitr::knit_exit()
}
```
```{r contigSizes, echo = FALSE}
faidx <- snakemake@input[["faidx"]]
fa.sizes <- read.table(faidx, header = F)[,1:2] %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")

# DEPRECATED
# make sure that the contigs with SV are the ones being plotted, not the others
#fa.sizes <- fa.sizes[fa.sizes$contig %in% unique(sv$contig), ]

# limit the data to up to 30 of the largest contigs
#fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]
fa.sizes <- arrange(fa.sizes, size)[1:min(nrow(fa.sizes), 30), ]
```

## General Information {data-height=100}
### nvariants
```{r summaryinfo, echo=FALSE, message=FALSE, warning=FALSE}
summinfo <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n()) %>% 
  group_by(type) %>%
  summarise(
    total = sum(count),
    avg_per_contig = round(mean(count), 2),
    sd = round(sd(count), 2)
)

nvar = sum(summinfo$total)
valueBox(scales::comma(nvar), caption = "Variants", color = ifelse(nvar > 0, "success", "warning"))
```

### nbreakends
```{r}
nbnd <- ifelse("BND" %in% summinfo$type, summinfo$type[which(summinfo$type == "BND")], 0)
valueBox(scales::comma(nbnd), caption = "Breakends", color = "info")
```
### ndeletions
```{r}
ndel <- ifelse("DEL" %in% summinfo$type, summinfo$type[which(summinfo$type == "DEL")], 0)
valueBox(scales::comma(ndel), caption = "Deletions", color = "info")
```

### nduplications
```{r}
ndup <- ifelse("DUP" %in% summinfo$type, summinfo$type[which(summinfo$type == "DUP")], 0)
valueBox(scales::comma(ndup), caption = "Duplications", color = "info")
```

### ninversions
```{r}
ninv <- ifelse("INV" %in% summinfo$type, summinfo$type[which(summinfo$type == "INV")], 0)
valueBox(scales::comma(ninv), caption = "Inversions", color = "info")
```

## Various Stats Headers
### summary info {.no-title}
<h2> Variant Information </h2>
Below are two tables with information on the variants. The first (left) table
details all the variants LEVIATHAN detected and the quantity/quality of evidence
for those detections. The second (right) table details each type of structural variant for
every contig. 

## Various Stats Tables
### All variants found
```{r various stats, echo = FALSE, warnings = FALSE, message = FALSE}
DT::datatable(
  sv,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  autoHideNavigation = T,
  fillContainer = T
)
```

### Per-contig info
```{r echo = FALSE, message = FALSE}
# breakends are written as ".", so make them length = 1 for plotting
sv$length <- gsub("\\.", "1", sv$length) %>%  as.numeric()
```

```{r percontig, echo = FALSE, message = FALSE, warning = FALSE}
grpstats <- sv %>%
  group_by(contig, type) %>%
  summarise(count = n(), total_bp  = sum(length))

DT::datatable(
  grpstats, 
  rownames = F, 
  filter = "top", 
  extensions = 'Buttons', 
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  fillContainer = T
)
```

## Circleplot {.no-title data-height=900}
### pltdsc {.no-title data-width=150}
<h2> Variant size and position </h2>
To the right is a circular plot to visualize the distribution of structural variants
across (up to) 30 of the largest contigs.
This should help you assess the presence/absence of expected variants. There are 4 labelled "rings" in this plot, each corresponding
to a type of structural variant (inversions in the outer ring, then deletions, etc.). All four rings
are going to be shown, even if there are not variants of that type.

This plot allows you to hover your mouse over variants to display specific information for that variant,
pan by clicking and dragging, and zoom. In case you become unable to scroll up from the plot due to these interactive 
features, place your cursor over this left column or the navigation bar at the top of this report and you will
be able to scroll the report instead of zooming on the plot.

### circosSV {.no-title}
```{r fig.align='center', out.width= "80%", out.height="900px"}
genomeChr <- fa.sizes$size
names(genomeChr) <- fa.sizes$contig
genomeChr <- as.list(genomeChr)
tracks = BioCircosTracklist()
fullnames = list(BND = c("Breakend","#df487f") , DEL = c("Deletion", "#5a8c84"), DUP = c("Duplication", "#ffd75f"), INV = c("Inversion", "#4a9fea"))
existing_sv <- unique(sv3$type)
plot_position <- 0.15

#for (i in existing_sv){
for (i in c("BND","DEL","DUP","INV")){
  .svdata <- sv3[sv3$type == i, ]
  if(i %in% existing_sv){
    tracks = tracks + BioCircosArcTrack(
      paste0(i,"arcs"),
      as.character(.svdata$contig),
      starts = .svdata$position_start,
      ends = .svdata$position_end,
      opacities = 0.5,
      minRadius = plot_position,
      maxRadius = plot_position + 0.14,
      labels = fullnames[[i]][1],
      colors = fullnames[[i]][2]
    )
  }
  tracks = tracks + BioCircosTextTrack(paste0(i,"Text"), fullnames[[i]][1], weight = "lighter", size = "0.8em",  y = plot_position + 0.19)

  tracks = tracks + BioCircosBackgroundTrack(
    paste0(i, "background"),
    fillColors = "#fafafa",
    borderColors = "#C9C9C9",
    minRadius = plot_position,
    maxRadius = plot_position + 0.14,
  )
  plot_position <- plot_position + 0.21
}

BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.02,
  genomeTicksLen = 0,
  genomeTicksTextSize = 0,
  ARCMouseOverColor = "#91ff76", ARCMouseOverArcOpacity = 1.0,
  genomeLabelDy = 7,
  width = "100%"
)
```