---
title: "NAIBR Structural Variant Summary"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---


```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr", "ggplot2", "tidyr","DT","plotly","circlize", "viridisLite", "tools")

#library(flexdashboard)
#library(dplyr)
#library(ggplot2)
#library(magrittr)
#library(tidyr)
#library(DT)
#library(plotly)
#library(circlize)
#library(viridisLite)
#library(tools)
#library(ComplexHeatmap)
```
```{css zoom-lib-src, echo = FALSE, message = FALSE, warning = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.7.1/jquery.min.js"
```
```{js zoom-jquery, echo = FALSE, message = FALSE, warning = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r echo = FALSE, warnings = FALSE, message = FALSE}
infiles <- snakemake@input[["bedpe"]]
fai <- snakemake@input[["fai"]]
#infiles <- paste0("/home/pdimens/", c(1,2,3), ".bedpe")
#infiles <- list.files("~/downsampled/", full.names = T)
#fai <- "/home/pdimens/test.fai"
```

# General Stats
## Intro
### introtext {.no-title}
<h1> Structural Variants Summary </h1>
Below is a series of tables and plots detailing the structural variants
identified by [NAIBR](https://github.com/pontushojer/NAIBR) [(publication)](https://doi.org/10.1093/bioinformatics/btx712).
The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
knitr::kable(
  data.frame(
    "Name" = c("Inversion", "Duplication", "Deletion"),
    "Description" = c(
      "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
      "a type of mutation involving the production of one or more copies of a gene or region of a chromosome",
      "a type of mutation involving the loss of one or more nucleotides from a segment of DNA"
      )
    )
)
```
This tab (`General Stats`) shows overview information. Clicking `Per-Contig Plots`
in the navigation bar at the top of this page will show you interactive
plots detailing all variants identified.

In the colorful summary boxes are the per-population average number of
variants detected by type.

```{r echo = FALSE, warnings = FALSE, message = FALSE}
readvariants <- function(x){
  y <- read.table(
    x,
    header = T,
    colClasses = c("factor", "integer", "factor", "integer", "integer", "integer", "character", "character", "numeric", "factor", "factor")
  )
  y$Population <- gsub(".bedpe", "", basename(x))
  y$Population <- as.factor(y$Population)
  y <- y[, c(12,1:11)]
  return(y)
}

cleanbedpe <- function(x){
  y <- x[x$Chr1 == x$Chr2, c(-4, -11)]
  names(y)[2] <- "Chr1"
  y$Score <- round(y$Score, digits = 3)
  y$Length <- abs(y$Break2 - y$Break1)
  return(y)
}


chimeric <- function(x){
  y <- x[x$Chr1!=x$Chr2, -11]
  y$Score <- round(y$Score, digits = 3)
  return(y)
}
```

```{r echo = FALSE}
emptyfiles <- character()
sv <- data.frame(matrix(ncol=12,nrow=0, dimnames=list(NULL, c("Population", "Chr1", "Break1", "Chr2", "Break2", "SplitMolecules", "DiscordantReads", "Orientation", "Haplotype", "Score", "PassFilter", "SV"  ))))
for(i in infiles){
  .df <- tryCatch(readvariants(i), error = function(e){return(0)})
  if(length(.df)==1){
      emptyfiles <- c(emptyfiles, i)
  } else {
    sv <- rbind(sv, .df)
  }
}
if(length(emptyfiles) == length(infiles)){
  cat("There are no variants detected in any of the populations:\n")
  cat(paste0("- ", infiles), sep = "\n")
  knitr::knit_exit()
}
if(length(emptyfiles > 0)){
  cat("Some of the populations did not have any variants detected and are omitted from this report:\n")
  cat(paste0("- ", emptyfiles), sep = "\n")
}
```

```{r echo = FALSE, warnings= FALSE, message = FALSE}
sv_clean <- cleanbedpe(sv)
```

```{r echo = FALSE, warnings = FALSE, message = FALSE}
fa.sizes <- read.table(fai, header = F) %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")
# make sure that the contigs with SV are the ones being plotted, not the others
fa.sizes <- fa.sizes[fa.sizes$contig %in% unique(sv_clean$Chr1), ]

# limit the data to only the 30 of the largest present contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]
```

```{r echo = FALSE, warnings= FALSE, message = FALSE}
summinfo <- sv %>%
  group_by(Population, SV) %>%
  summarise(tot = n()) %>%
  group_by(SV) %>%
  summarise(count = sum(tot), avg = round(mean(tot, na.rm = T),2 ), sd = round(sd(tot, na.rm = T), 2)) #%>% 
```

## header info
### npops
```{r}
valueBox(scales::comma(length(levels(sv$Population))), caption = "Populations", color = "success")
```

### ndeletions
```{r}
variant <- "deletion"
ndel <- ifelse(variant %in% summinfo$SV, summinfo$avg[which(summinfo$SV == variant)], 0)
valueBox(scales::comma(ndel, accuracy = 0.1), caption = "Deletions (average)", color = "info")
```

### nduplications
```{r}
variant <- "duplication"
ndup <- ifelse(variant %in% summinfo$SV, summinfo$avg[which(summinfo$SV == variant)], 0)
valueBox(scales::comma(ndup, accuracy = 0.1), caption = "Duplications (average)", color = "info")
```

### ninversions
```{r}
variant <- "inversion"
ninv <- ifelse(variant %in% summinfo$SV, summinfo$avg[which(summinfo$SV == variant)], 0)
valueBox(scales::comma(ninv, accuracy = 0.1), caption = "Inversions (average)", color = "info")
```

## per contig descs
### by count {.no-title}
<h3> Variants by count </h3>
The table below details counts each type of structural variant for every contig for every population.

### by length {.no-title}
<h3> Variants by length </h3>
The table below details the lengths (in bp) of structural variants for every contig for every population.

## Per contig counts
### Variants by count {.no-title}
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grpcounts <- sv_clean %>%
  group_by(Population, Chr1, SV) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = Population, values_from = count)

DT::datatable(
  grpcounts,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  autoHideNavigation = T,
  fillContainer = T
)
```

### Variants by base-pair span {.no-title}
```{r echo = FALSE, warnings= FALSE, message = FALSE}
grplens <- sv_clean %>% 
  group_by(Population, Chr1, SV) %>%
  summarise(bp = sum(Length)) %>%
  pivot_wider(names_from = Population, values_from = bp)

DT::datatable(
  grplens,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  autoHideNavigation = T,
  fillContainer = T
)
```

## Various Stats
### interpreting output {.no-title}
<h2> Interpreting Output </h2>
NAIBR outputs a tab-delimited file with named columns (along with a VCF
and a reformatted bedpe file with extra information). The columns of the
bedpe file are deciphered as such:

```{r}
knitr::kable(
  data.frame(
    "Column" = names(sv),
    "Descrption" = c(
      "Name of the population",
      "Name of the contig where the variant starts",
      "Base-pair position in `Chr1` of the start of the variant",
      "Name of the contig where the variant ends",
      "Base-pair position in `Chr2` of the end of the variant",
      "Number of split molecules supporting variant",
      "Number of discordant reads supporting variant",
      "Orientation of variant relative to reference",
      "The haplotype of the variant",
      "log-likelihood score for variant",
      "`PASS` if passes internal NAIBR filter threshold, otherwise `FAIL`",
      "The type of variant (inversion, deletion, or duplication)"
    )
  ), col.names = c("Column Name", "Description")
)
```
## variants anc chimeras desc
### all variants desc {.no-title}
<h3> All non-chimeric variants </h3>
The table below details the variants detected by NAIBR that appear on a single contig/chromosome and passed the programs internal filtering.

### chimeras desc {.no-title}
<h3> Chimeric variants </h3>
The table below shows any structural variants whose breakpoints span multiple contigs and may require further assessment. These variants are not shown in the plots below.

## Variants and Chimeras
### all variants {.no-title}
```{r message = FALSE, echo = FALSE}
DT::datatable(
  sv_clean,
  rownames = F,
  filter = "top",
  extensions = 'Buttons', 
  options = list(dom = 'Brtip', buttons = 'csv', scrollX = TRUE),
  autoHideNavigation = T,
  fillContainer = T
)
```
### chimeric variants {.no-title}
```{r echo = FALSE, warnings = FALSE, message = FALSE}
DT::datatable(
  chimeric(sv),
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE)
)
```

## circos plots
### circos plots desc {.no-title}
<h2> Variant Summary Plots </h2>
For an overview of the size and position of detected variants across
populations, three "circos" plots are shown below. These plots are for
summarization purposes and are not interactive. If you wish to view these
data in better detail, click the `Per-Contig Plots` tab in the navigation
bar above.

The plots below are a visual representation of each structural variant where it was detected.
Contigs that are absent did not have any variants detected on them. Within each 
contig, each population is plotted as a row (ring) to help visualize shared
variants between populations. Below is the legend to these plots, where
colors represent populations. Plotting only the first 30 largest contigs.
You may click on the plot to expand it in your browser window. Clicking it again will exit
out of the zoomed view.

## legend {data-height=125}
### Populations legend {.no-title}
```{r fig.width=20, fig.height=1.5}
populations <- unique(sv$Population)
popheight <- .5 / length(populations)
col.ramp <- viridis(length(populations))

pops <- data.frame(x = 1, population = levels(populations))
ggplot(pops, aes(x = population, y = 1, fill = population)) +
  geom_tile(color = "white", lwd = 1.5,linetype = 1) +
  scale_fill_manual(values=col.ramp, drop=FALSE) +
  theme_minimal() +
  scale_x_discrete(position = "top", guide = guide_axis(n.dodge = 2)) +
  theme(
    panel.grid = element_blank(),
    legend.position = "none",
    axis.text.y=element_blank(),
    axis.ticks.y=element_blank(),
    axis.title = element_blank()
  ) +
  labs(title = "Population")
```

## SV size and location {data-height=600}
### Inversions
```{r echo = FALSE, message = FALSE, warning = FALSE}
circosplot <- function(data, contigdata, svtype, trackheight, poplist, colormap){
  circos.clear()
  col_text <- "grey40"
  circos.par("track.height" = 0.8, gap.degree = 2, cell.padding = c(0, 0, 0, 0))
  circos.initialize(
    factors = contigdata$contig,
    xlim = matrix(c(rep(0, nrow(contigdata)), contigdata$size), ncol = 2)
  )

  # contigs
  circos.track(
    ylim = c(0, 1),
    bg.col = "grey90",
    bg.border = F,
    track.height = 0.06,
    panel.fun = function(x, y) {
      chr = CELL_META$sector.index
      xlim = CELL_META$xlim
      ylim = CELL_META$ylim
      circos.text(
        mean(xlim),
        mean(ylim),
        chr,
        cex = 0.6,
        col = col_text,
        facing="bending.inside",
        niceFacing = T
      )
    }
  )

  # x axis
  brk <- seq(0, 10, 0.5) * 10^7
  circos.track(
    ylim = c(0,1),
    track.index = get.current.track.index(),
    bg.border = F,
    panel.fun = function(x, y) {
      circos.axis(
        h = "top",
        major.at = brk,
        labels = round(brk/10^6, 1),
        labels.cex = 0.6,
        col = col_text,
        labels.col = col_text,
        lwd = 0.7,
        labels.facing = "clockwise"
      )
    }
  )

  # plot the track
  for(i in 1:length(poplist)){
    svquery = data$Population == populations[i] & data$SV == svtype
    circos.genomicTrack(
      ylim = c(0,1),
      data = data[svquery, -1],
      track.height = trackheight,
      bg.border = F,
      panel.fun = function(region, value, ...) {
      circos.genomicRect(region, value, ybottom = 0, ytop = 1, col = colormap[i], border = colormap[i])
      }
    )
  }
  text(0, 0, svtype, cex = 1)
}
```
```{r echo = FALSE, message = FALSE, warning = FALSE, dev = "jpeg", fig.align='center'}
circosplot(sv_clean, fa.sizes, "inversion", popheight, populations, col.ramp)
```

### Deletions
```{r echo = FALSE, message = FALSE, warning = FALSE, dev = "jpeg", fig.align='center'}
circosplot(sv_clean, fa.sizes, "deletion", popheight, populations, col.ramp)

```

### Duplications
```{r echo = FALSE, message = FALSE, warning = FALSE, dev = "jpeg", fig.align='center'}
circosplot(sv_clean, fa.sizes, "duplication", popheight, populations, col.ramp)
```

# Per-Contig Plots
## percontig desc
### desc {.no-title}
<h1> Structural Variants Per Contig Per Population </h1>
Below is a plot to help you assess what structural variants were
detected by NAIBR These plots are interactive, allowing you
to hover over a variant to provide additional information, including
the genomic interval in which it occurs and the number of haplotag
barcodes supporting the variant. Populations are stacked vertically
for each contig, denoted by the grey strip on the right side of each row.

## per contig
### plots per contig {.no-title data-height=10000}
```{r echo = FALSE, message = FALSE, warning = FALSE}
color.palette <- c(
  "deletion" = "#5a8c84",
  "duplication" = "#ffd75f",
  "inversion" = "#4a9fea"
)

sv_clean$ystart <- case_when(
  sv_clean$SV == "inversion" ~ 0.1,
  sv_clean$SV == "duplication" ~ 1.1,
  sv_clean$SV == "deletion" ~ 2.1
)
sv_clean$ystop <- sv_clean$ystart + 0.8

```


```{r echo = FALSE, out.width = '100%', warning=FALSE}
l <- list()
nplotrows <- 0

for (i in 1:nrow(fa.sizes)) {
  sv.filt <- sv_clean %>% filter(Chr1 == fa.sizes$contig[i])
  sv_stats <- group_by(sv.filt, SV) %>% summarise(n = length(SV))
  if (nrow(sv_stats) == 0) {
    next
  }
  plt <- sv.filt %>%
    ggplot() +
    geom_rect(
      alpha = 0.7,
      aes(
        xmin = Break1,
        xmax = Break2,
        ymin = ystart,
        ymax = ystop,
        fill = SV,
        color = SV,
        text = sprintf("Type: %s<br>Position: %s-%s<br>barcodes: %s", SV, Break1, Break2, SplitMolecules)        )
      ) +
    geom_hline(yintercept = 1:3, color = "grey96") +
    scale_color_manual(values = color.palette) +
    scale_fill_manual(values = color.palette) +
    facet_grid(rows = vars(Population)) +
    theme_light() +
    theme(
      axis.line.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_rect(fill = "grey80"),
      strip.text = element_text(color = "grey20"),
      legend.position = "none",
    ) +
    scale_x_continuous(labels = scales::comma) +
    scale_y_continuous(
      breaks = c(0.5, 1.5, 2.5),
      labels = c("INV", "DUP", "DEL"),
      limits = c(0,3)
    ) +
    coord_cartesian(xlim = c(0, fa.sizes$size[i] + 1), expand = F) +
    xlab("Position (bp)")
    annotations <- list( 
      x = 0.5,
      y = 1.05,  
      text = as.character(fa.sizes$contig[i]),  
      xref = "paper",  
      yref = "paper",  
      xanchor = "center",  
      yanchor = "bottom",  
      showarrow = FALSE 
    )
  pp <- subplot(ggplotly(plt, height = 800 * length(populations), tooltip = "text")) %>% 
    layout(annotations = annotations)
    l[[i]] <- pp
}
# remove null
l <- Filter(Negate(is.null), l)
subplot(l, nrows = length(l) , shareX = FALSE, titleX = TRUE)
```