---
title: "NAIBR SV Calling Summary"
date: "`r format(Sys.time(), '%d %b, %Y at %H:%M')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    mathjax: NULL
    logo: https://raw.githubusercontent.com/pdimens/harpy/docs/static/logo_report.png
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_report.png"
    navbar:
      - { title : "Docs", icon: "fa-book", href: "https://pdimens.github.io/harpy/", align: right }
      - { title : "Source", icon: "fa-github", href: "https://www.github.com/pdimens/harpy/", align: right }
---



```{r echo = FALSE, warnings = FALSE, message = FALSE}
infile <- snakemake@input[["bedpe"]]
fai <- snakemake@input[["fai"]]
samplename <- gsub(".bedpe", "", basename(infile))
```

## file top
### file header {.no-title}
<h1> NAIBR Variant Calling: `r samplename`  </h1>

```{r echo = FALSE, warnings = FALSE, message = FALSE, results = 'asis'}
tryCatch(
  expr = {
    variants <- read.table(infile, header = T) 
  },
  error = function(e){ 
    cat(paste0("There are no variants in the file ", "`", infile, "`"))
    knitr::knit_exit()
  }
)

if (nrow(variants) == 0) {
  cat(paste0("There are no variants in the file ", "`", infile, "`"))
  knitr::knit_exit()
}
```
```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr","tidyr", "ggplot2","DT", "BioCircos")
```

Below is a series of tables and plots detailing the structural variants
identified by [NAIBR](https://github.com/pontushojer/NAIBR) [(publication)](https://doi.org/10.1093/bioinformatics/btx712).
The variants are given by:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE}
knitr::kable(
  data.frame(
    "Name" = c("Inversion", "Duplication", "Deletion"),
    "Description" = c(
      "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
      "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
      "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA"
      )
    )
)
```
This tab (`General Stats`) shows overview information. Clicking `Per-Contig Plots`
in the navigation bar at the top of this page will show you interactive
plots detailing all variants identified.

```{r echo = FALSE, warnings = FALSE, message = FALSE}
cleanbedpe <- function(x){
  y <- x[x$Chr1==x$Chr2, c(-3,-10)]
  y$Score <- round(y$Score, digits = 3)
  y$Length <- abs(y$Break2 - y$Break1)
  names(y)[1] <- "Chr1"
  return(y)
}

chimeric <- function(x){
  y <- x[x$Chr1!=x$Chr2, -10]
  y$Score <- round(y$Score, digits = 3)
  names(y)[1] <- "Chr1"
  return(y)
}

sv <- cleanbedpe(variants)
```

## overivew desc
### overvew desc {.no-title}
<h3> Summary of Identified Variants </h3>
The tables below detail how many of each type of structural variant appear in each contig (left) and how many base-pairs these variants span (right).

## overview datatable
### summary info
```{r echo = FALSE, message = FALSE}
summstats <- sv %>%
  group_by(Chr1, SV) %>%
  summarise(n = n())

sumtable <- summstats %>%
  group_by(SV) %>%
  summarise(count = sum(n))

nvar <- sum(sumtable$count)
valueBox(scales::comma(nvar), caption = "Variants", color = ifelse(nvar > 0, "success", "warning"))
```

### ndeletions
```{r}
ndel <- ifelse("deletion" %in% sumtable$SV, sumtable$count[which(sumtable$SV == "deletion")], 0)
valueBox(scales::comma(ndel), caption = "Deletions", color = "info")
```

### nduplications
```{r}
ndup <- ifelse("duplication" %in% sumtable$SV, sumtable$count[which(sumtable$SV == "duplication")], 0)
valueBox(scales::comma(ndup), caption = "Duplications", color = "info")
```

### ninversions
```{r}
ninv <- ifelse("inversion" %in% sumtable$SV, sumtable$count[which(sumtable$SV == "inversion")], 0)
valueBox(scales::comma(ninv), caption = "Inversions", color = "info")
```

## data tables
### Variants by count
```{r echo = FALSE, message = FALSE, warning = FALSE}
grpstats <- sv %>%
  group_by(Chr1, SV) %>%
  summarise(count = n(), total_bp  = sum(Length))

DT::datatable(
  grpstats %>% pivot_wider(id_cols = 1, names_from = SV, values_from = count),
  rownames = F,
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = paste0(samplename , "_sv_naibr_count"))),
    scrollX = TRUE
  ),
  autoHideNavigation = T,
  fillContainer = T
)
```
### Variants by base-pair span
```{r}
DT::datatable(
  grpstats %>% pivot_wider(id_cols = 1, names_from = SV, values_from = total_bp),
  rownames = F,
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = paste0(samplename , "_sv_naibr_count_bp"))),
    scrollX = TRUE
  ),  autoHideNavigation = T,
  fillContainer = T,
)
```
## Various Stats
### interpreting output {.no-title}
<h2> Interpreting Output </h2>
NAIBR outputs a tab-delimited file with named columns (along with a VCF
and a reformatted bedpe file with extra information). The columns of the
bedpe file are deciphered as such:

```{r}
knitr::kable(
  data.frame(
    "Column" = names(variants),
    "Descrption" = c(
      "Name of the contig where the variant starts",
      "Base-pair position in `Chr1` of the start of the variant",
      "Name of the contig where the variant ends",
      "Base-pair position in `Chr2` of the end of the variant",
      "Number of split molecules supporting variant",
      "Number of discordant reads supporting variant",
      "Orientation of variant relative to reference",
      "The haplotype of the variant",
      "log-likelihood score for variant",
      "`PASS` if passes internal NAIBR filter threshold, otherwise `FAIL`",
      "The type of variant (inversion, deletion, or duplication)"
    )
  ), col.names = c("Column Name", "Description")
)
```
## Variants and Chimeras desc
### variants desc {.no-title}
<h3> Non-Chimeric Variants </h3>
The table below details the variants detected by NAIBR that appear on a single
contig/chromosome and passed the programs internal filtering. In other words, the
detected non-chimeric variants.

### chimeras desc {.no-title}
<h3> Chimeric Variants </h3>
The table below shows any structural variants whose breakpoints span multiple contigs
("chimeric") and may require further assessment. These variants are omitted from the plots below.

## Variants and Chimeras
### variants {.no-title}
```{r echo = FALSE, warnings = FALSE, message = FALSE}
DT::datatable(
  sv[,-10],
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = paste0(samplename ,"_sv_naibr"))),
    scrollX = TRUE
  ),
  autoHideNavigation = T,
  fillContainer = T
)
```
### chimeric variants {.no-title}
```{r echo = FALSE, warnings = FALSE, message = FALSE}
DT::datatable(
  chimeric(variants),
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = paste0(samplename ,"_sv_naibr_chimeric"))),
    scrollX = TRUE
  ),
  autoHideNavigation = T,
  fillContainer = T
)
```


## variant plots {data-height=1000}
### Circos Plot Legend {data-width=200}
To the right is a circular plot to visualize the distribution of structural variants
across (up to) 30 of the largest contigs. If you are unfamiliar with this kind of visualization,
it's a circular version of a linear genome. Each "wedge" is a different contig, from position 0 to 
the end of the contig, and is labelled by the contig name. Each internal (grey) ring is a plot
of observed structural variants for that contig, where the innermost is duplications, middle is deletions,
and outermost is inversions.

Very small variants may be difficult to see or will not appear on the plot. Deletions are displayed as points as they tend to be small
and clump together. You may hover your cursor over variants to view their positions, pan by clicking and dragging,
and zoom using scroll (mouse or touchpad). In case you become unable to scroll up from the plot due to these interactive 
features, place your cursor over this left column or the navigation bar of this report and you will
be able to scroll the report instead of zooming on the plot.

```{r dev="png"}
color.palette <- c("deletion" = "#5a8c84", "duplication" = "#99278a", "inversion" = "#4a9fea")
data.frame(x = 1, y = 1:3, colour = c("inversion", "deletion","duplication")) %>% 
ggplot(aes(x, y, fill = colour))+
  geom_point(alpha=0, shape = 22, color = "white")+ # completely transparent rectangular point 
  scale_fill_manual(values=color.palette, drop=FALSE) +
  guides(fill = guide_legend(override.aes = list(alpha=1, size = 25)))+ # showing the point in the legend
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "inside",
        legend.position.inside = c(0.3, 0.5),
        legend.text = element_text(size = 25),
        legend.title=element_text(size=25),
        legend.key = element_rect(fill='NA'),
        panel.grid = element_blank(),
        panel.background = element_rect(fill='transparent'), 
        plot.background = element_rect(fill='transparent', color=NA),
        legend.background = element_rect(fill='transparent'),
        legend.box.background = element_rect(fill='transparent', color = "transparent"), 
  ) +
  labs(fill="Variant Type")
```

### Structural Variants: `r samplename` (save by taking a screenshot)
```{r filter_contigs}
fa.sizes <- read.table(fai, header = F) %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")

# make sure that the contigs with SV are the ones being plotted, not the others
fa.sizes <- fa.sizes[fa.sizes$contig %in% unique(sv$Chr1), 1:2]

# limit the data to only the 30 of the largest present contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]

# filter out variants that aren't on one of these contigs
sv <- sv[sv$Chr1 %in% fa.sizes$contig, ]
sv$Chr1 <- factor(sv$Chr1)
```

```{r interactive_circos, out.width= "80%", out.height="900px"}
genomeChr <- fa.sizes$size
names(genomeChr) <- fa.sizes$contig
genomeChr <- as.list(genomeChr)
tracks <- BioCircosTracklist()
radius <- 0.3

# Add one track for each SV type
# inversions
inv <- sv[sv$SV == "inversion",]
if(nrow(inv) > 0){
  tracks <- tracks + BioCircosArcTrack(
    'Inversions',
    inv$Chr1,
    inv$Break1,
    inv$Break2,
    opacities = 0.7,
    values = inv$SplitMolecules,
    minRadius = 0.73,
    maxRadius = 0.87,
    colors = color.palette["inversion"],
    labels = "Type: Inversion",
  )
}

# deletions
del <- sv[sv$SV == "deletion",]
if(nrow(del) > 0){
  tracks <- tracks + BioCircosSNPTrack(
    'Deletions',
    del$Chr1,
    del$Break1,
    values = abs(del$Break2 - del$Break1),
    minRadius = 0.53,
    maxRadius = 0.67,
    size = 1,
    shape = "rect",
    opacities = 0.7,
    colors = color.palette["deletion"],
    labels = "Type: Deletion"
  )
}

# duplications
dup <- sv[sv$SV == "duplication",]
if(nrow(dup) > 0){
  tracks <- tracks + BioCircosArcTrack(
    'Duplications',
    dup$Chr1,
    dup$Break1,
    dup$Break2,
    values = dup$SplitMolecules,
    opacities = 0.7,
    minRadius = 0.33,
    maxRadius = 0.47,
    colors = color.palette["duplication"],
    labels = "Type: Duplication",
  )
}

# loop to create backgrounds
for(i in c("inversion", "deletion", "duplication")){
  tracks = tracks + BioCircosBackgroundTrack(
    paste0(i,"_background"),
    minRadius = radius, maxRadius = radius + 0.2,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius + 0.2
}

BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  SNPMouseOverCircleSize = 2,
  SNPMouseOverTooltipsHtml03 = "<br/>Length: ",
  width = "100%",
  height = "1000px",
  elementId = "circosplot" 
)
```