---
title: "FASTQ File Format Check"
date: "`r format(Sys.time(), '%m-%d-%y %X')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows  
    vertical_layout: scroll
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r load environment, echo=FALSE, message=FALSE, warning=FALSE}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need)
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr","tidyr","DT","highcharter","scales")

library(flexdashboard)
library(dplyr)
library(tidyr)
library(DT)
library(highcharter)
library(scales)
#library(ggplot2)

#infile <- "~/Documents/harpy/test/fastq/Preflight/filecheck.fastq.tsv"
infile <- snakemake@input[[1]]
#dirname <- "filler text"
data <- read.table(infile, header = T)
attention_df <- data %>% 
  mutate(allmissing = (reads == noBX)) %>% 
  select(4,5,7) %>% 
  rowwise() %>% 
  summarise(count = sum(c(badBX,badSamSpec, allmissing)))
attention <- sum(attention_df$count > 0)
```

## filetop
### fltop {.no-title}
<h1> Preflight checks for FASTQ files </h1>

This report reflects the FASTQ files provided. Harpy has processed
the files to identify formatting issues that may require your attention.

## General Information {data-height=100}
### nfiles
```{r}
valueBox(scales::comma(nrow(data)), caption = "Files", color = "info")
```
### fileformat
```{r}
valueBox("FASTQ", caption = "File Format", color = "info")
```
### issues
```{r}
valueBox(scales::comma(attention), caption = "Files With Issues", color = ifelse(attention > 0, "warning", "success"))
```

### BX not last
```{r}
bxnotlast <- sum(data$bxNotLast > 0)
valueBox(bxnotlast, caption = "BX:Z: tag not last", color = ifelse(bxnotlast > 0, "#feda75", "success"))
```

## interpret
### interpcol {.no-title}
<h2> Interpreting the Data </h2>

The `harpy preflight fastq -d ...` created a `Preflight` folder within
the supplied directory and within it you will find this report, along with the
`filecheck.fastq.tsv` file summing the results that are included in this report.
This file contains a tab-delimited table with the columns: `file`,	`reads`, `noBX`,	`badBX`, and	`badSamSpec`. These columns are defined by:

- `file`: the name of the FASTQ file
- `reads`: the total number of reads in the file
- `noBX`: the number of reads that do not have a `BX:Z:` tag in the read header
  - severity: <strong style='color:#c0c0c0;'>minimal</strong> and likely won't cause issues unless no reads have `BX:Z:` tags
  - if you expect all of your reads should have `BX:Z:` tags, then further investigation is necessary
- `badBX`: the haplotag barcode in the `BX:Z:` comment does not adhere to the proper `AxxCxxBxxDxx` format
  - severity: <strong style='color:red;'>critical</strong> and will cause downstream issues and errors
- `badSamSpec`: the comments in the read header after the read ID do not conform to the `TAG:TYPE:VALUE` [SAM specification](https://samtools.github.io/hts-specs/SAMv1.pdf)
  - severity: <strong style='color:red;'>critical</strong> and will cause downstream issues and errors
- `bxNotLast`: the `BX:Z:` tag in the FASTQ header is not the last comment
  - severity: <strong style='color:orange;'>conditional</strong> and only relevant for LEVIATHAN variant calling 
  - can be ignored if not intending to call structural variants with LEVIATHAN
  - if intending to use LEVIATHAN, the `BX:Z:` tag must be the last comment in a read header
  - `harpy align -m bwa` automatically moves the `BX:Z:` tag to the end of the alignment record

### Proper header {.no-title}
<h2> Proper Read Headers </h2>

An example of a proper FASTQ read header is like the one below, where all comments
following the initial read ID (`@A00814...`) have the proper [SAM spec](https://samtools.github.io/hts-specs/SAMv1.pdf)
`TAG:TYPE:VALUE` format and among them is a `BX:Z:` tag followed by a `AxxCxxBxxDxx`
formatted haplotag barcode. The comments must be **TAB separated** because the `:Z:` tag type allows a 
whitespace character for its value. The read header below contains the read ID `@A00814...`,
the `BX:Z:` haplotag barcode tag, and two more comments `RX:Z:` and `QX:Z:` that both
adhere to the SAM specification. If using LEVIATHAN to call structural variants, the `BX:Z:` tag
must be the last comment in the read header.

```
@A00814:267:HTMH3DRXX:2:1101:4580:1000	BX:Z:A02C01B11D46	RX:Z:GAAACGACCAACA+CGAACACGTTAGC	QX:Z:F,FFFFFFFFFFF+FF,FFF:FFFFFF
```

## viz desc
### desc {.no-title}
<h2> FASTQ file format Check </h2>
Below are a plot and corresponding table of the results of the preflight checks.
The plot renames the parameters for clarity and they are:

```{r}
knitr::kable(
  data.frame(
    "data column" = c("`badBX`", "`badSamSpec`", "`bxNotLast`", "`noBX`"),
    "name in plot" = c("AxxCxxBxxDxx format", "follows SAM spec", "BX:Z: last comment", "BX:Z: tag"),
    "pass condition" = c(
      "**all** reads with BX:Z: tag have properly formatted `AxxCxxBxxDxx` barcodes",
      "**all** reads have proper `TAG:TYPE:VALUE` comments",
      "**all** reads have `BX:Z`: as final comment",
      "any `BX:Z:` tags present"
      ),
    "fail condition" = c(
        "**any** BX:Z: barcodes have incorrect format",
        "**any** reads have incorrectly formatted comments",
        "**at least 1 read** doesn't have `BX:Z:` tag as final comment",
        "**all** reads lack `BX:Z:` tag"
      )
  ), col.names = c("data column", "name in plot", "pass condition", "fail condition")
)
```

If the heatmap appears to be completely green, then all of your samples have passed all the preflight checks.
If any sample fails at least one check, fails will appear in yellow and passes will appear in blue.

## datarow 
### plot {.no-title}
```{r status_df}
data2 <- data %>% 
  mutate(
    noBX = ifelse(noBX == reads, 0, 1),
    badBX = ifelse(badBX > 0, 0, 1),
    badSamSpec = ifelse(badSamSpec > 0, 0, 1),
    bxNotLast = ifelse(bxNotLast > 0, 0, 1)
    ) %>% 
  select(-2)
  m <- as.matrix(data2[,-1]) * 1
  colnames(m) <- c("AxxCxxBxxDxx format", "follows SAM Spec", "BX:Z: last comment", "BX:Z: tag")
  rownames(m) <- data$file
```


```{r readsper, echo = FALSE, message = FALSE, warning = FALSE, out.width = '100%'}
hchart(m) |>
  hc_colorAxis(stops = list(c(0, "#feda75"), c(1, "#4faad1")), tickAmount = 2) |>
  hc_title(text = "FASTQ file preflight QC") |>
  hc_xAxis(labels = list(style = list(fontSize = '16px')) )|>
  hc_yAxis(labels = list(style = list(fontSize = '16px')) )|>
  hc_plotOptions(heatmap = list(borderWidth = 1, borderColor = "white", animated = F)) |>
  hc_legend(enabled = F) |>
  hc_tooltip(
    animation = FALSE,
    formatter = JS("function () {return '<b>' + this.series.xAxis.categories[this.point.x] + '</b><br>' + this.series.yAxis.categories[this.point.y] + '<br/>QC: <b>' + (this.point.value<1 ? 'FAIL' : 'PASS') + '</b>';}") 
  ) |>
  hc_exporting(enabled = T, filename = "fastq.preflight",
    buttons = list(contextButton = list(menuItems = c("viewFullscreen", "separator", "downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```

### table {.no-title}
```{r dtable, echo=FALSE, message=FALSE, warning=FALSE}
DT::datatable(
  data,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = T),
  fillContainer = T
)
```
