---
title: "STITCH Imputation Details"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_dark.png"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
options(knitr.graphics.rel_path = FALSE)
```
```{css zoom-lib-src}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.7.1/jquery.min.js"
```
```{js zoom-jquery}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

```{r setup environment}
library(flexdashboard)
library(ggplot2)
library(magrittr)
library(tidyr)
library(DT)
library(plotly)
library(RColorBrewer)
```

```{r load data}
#dataL <- readLines("~/contig1.stats")
dataL <- readLines(snakemake@input[[1]])
#bcf <- "filename.bcf"
bcf <- gsub(".stats$", ".bcf", snakemake@input[[1]])
basedir <- dirname(normalizePath(snakemake@input[[1]]))
#basedir <- "/home/pdimens/Documents/harpy/test/Impute/modeldiploid_usebxTrue_bxlimit50000_k3_s1_ngen5/contigs/contig1/"
chrom <- basename(basedir)
plotdir <- paste(basedir, "plots/", sep = "/")
```
## file header
### headertext {.no-title}
<h1> `r bcf` </h1>
For convenience, this report consolidates the various outputs from `STITCH` and `bcftools` into a single document.

This reflects the general information stored in the records of `r bcf`. Since STITCH
requires only bi-allelic SNPs as input, you should not expect to see any other types
of variants (MNPs, indels, _etc._).

```{r General Stats}
.snL <- grepl("^SN", dataL)
sn <- read.table(text=dataL[.snL], sep = "\t")[, 3:4]
names(sn) <- c("Metric", "Number")
sn$Metric <- gsub("number of ", "", sn$Metric)
sn$Metric <- gsub(":", "", sn$Metric)
rownames(sn) <- sn$Metric
sn <- as.data.frame(t(sn[2]))
rownames(sn) <- NULL
```

<h2> General Information </h2>
Below is a plot and a table of variant statistics per individual, corresponding 
to the `r sn$samples[1]` samples in `r bcf`. The table to the right of it is the
underlying data being visualized. Clicking the plot will expand it to
fill your browser window. Clicking it again will exit out of the zoomed view.

## General Information {data-height=100}
### nsamples
```{r}
valueBox(scales::comma(sn$samples[1]), caption = "Samples", color = "success")
```

### records
```{r}
valueBox(scales::comma(sn$records[1]), caption = "Total Records", color = "info")
```
### no alts
```{r}
valueBox(sn[3,1], caption = "no-ALTs", color = "info")
```

### snps
```{r}
valueBox(scales::comma(sn$SNPs[1]), caption = "SNPs", color = "info")
```

### mnps
```{r}
valueBox(scales::comma(sn$MNPs[1]), caption = "MNPs", color = "info")
```

## Individual Statistics
### plot {.no-title}
```{r}
# scale plot height to sample count
n <- as.numeric(sn$samples[1])
pltheight <- 330 + (15 * n)

pheight <- 1.2 + (0.166 * n)
```

```{r basic stats, fig.width = 7.5, dev ='jpeg'}
.pscL <- grepl("^PSC", dataL)
psc <- read.table(text=dataL[.pscL])[ ,3:14]
names(psc) <- c("Sample", "HomozygousRef", "HomozygousAtl", "Heterozygotes", "Transitions", "Transversions", "Indels",	"MeanDepth", "Singletons",	"HapRef", "HapAlt", "Missing")
psc$Homozygotes <- psc$HomozygousRef + psc$HomozygousAtl
tidy_psc <- pivot_longer(psc[,c(1,2,3,4,5,6,9,12,13)], -Sample , names_to = "Category", values_to = "Count")

ggplot(data = tidy_psc, mapping = aes(x = Count, y = Sample, color = Category)) +
  geom_point(size = 2) +
  labs(title = "Statistics by Sample") +
  theme_bw() +
  scale_x_continuous(n.breaks=9, labels = scales::comma) +
  xlab("Count") +
  ylab("") +
  scale_color_brewer(palette = "Dark2") +
  theme(panel.grid.minor = element_blank(), panel.grid.major.y = element_blank())
```

### table {.no-title}
```{r stats table}
DT::datatable(
  psc[,c(1,12,2,3,13,4,5,6,9)],
  rownames = F, 
  filter = "top", 
  extensions = 'Buttons', 
  options = list(dom = 'Brtip', buttons = c('csv'), scrollX = TRUE),
  autoHideNavigation = T,
  fillContainer = T
)
```
----
## STITCH Peformance Plots
### stitch desc {.no-title}
<h2> Plots Generated by STITCH </h2>
Below is a series of plots generated by `STITCH` along with descriptions of how to read and interpret the information.
The descriptions appear on the left and their corresponding plots appear to their right in labelled columns.
The images may appear small and can be clicked on to expand/focus them. Click on the image again to remove focus from the image. For the wider images, you may right-click them and open them in a new tab/window to view them at their original
sizes, which may prove to be more useful in some cases.

## QC
### Chromosome-wide Imputation QC
To the right is a series of informative plots, aligned vertically.

<h3> top plot </h3>
Gives -log10 (HWE p-value)

<h3> middle plot </h3>
Shows `INFO` score, which is a measure of the confidence the imputation has in its performance

<h3> bottom plot </h3>
Shows allele frequency, which can be useful to get a sense of the LD structure with the distribution of allele frequencies, as well as its relationship to imputation performance. Can also be used to test how various parameter choices affect imputation. For example, for species that have recently been through bottlenecks, allele frequencies are often highly correlated locally, visible as multiple nearby SNPs that have the same allele frequency. Choices of imputation parameters that increase the tightness in the spread of these allele frequencies often correlate with increased imputation performance.

_- adapted from the STITCH [documentation](https://github.com/rwdavies/STITCH/blob/5a5ba98442a105548587ed8cafbf00aece212c94/plots.md)_.

### Chromosome-wide Imputation QC Plot
```{r QC_chromwide, dev = "jpeg", out.width="100%"}
QC <- list.files(path = plotdir, pattern = "QC", full.names = TRUE)
knitr::include_graphics(QC[2])
```

## impute
### Impuation QC
Similar to the above plots, but plotting each of the three against each other. Useful for getting an overall sense of the distribution of the metrics, particularly `INFO`, which can be useful towards thinking about a threshold for filtering out variants after imputation. For example, in the middle plot with estimated allele frequency vs info, if you expect your data to cluster into a small range of allele frequencies, and your data seems concentrated in those frequencies for high `INFO`, you can think about an `INFO` score cutoff that allows you to capture most of these variants.

_- excerpt from the STITCH [documentation](https://github.com/rwdavies/STITCH/blob/5a5ba98442a105548587ed8cafbf00aece212c94/plots.md)_.


### Imputation QC Plot
```{r QC, dev = "jpeg", out.width="100%"}
knitr::include_graphics(QC[1])
```

## real vs
### Real vs Est. Allele Frequency
Scatter plot of real allele frequency as estimated from the pileup of sequencing reads (x-axis) and estimated allele frequency from the posterior of the model (y-axis). This is per-SNP, the sum of the average usage of each ancestral haplotype times the probability that ancestral haplotype emits an alternate base, divided by the number of samples. One should generally see good agreement between these. Note that these are at "good SNPs" with `INFO` score > `0.4` and HWE p-value > `1x10^-6` (the later criterion in particular might not be appropriate for all settings).

_- excerpt from the STITCH [documentation](https://github.com/rwdavies/STITCH/blob/5a5ba98442a105548587ed8cafbf00aece212c94/plots.md)_.

### Real vs Est. Allele Frequency Plot
```{r goodonly, dev = "jpeg", out.width="100%"}
goodonly <- list.files(path = plotdir, pattern = "goodonly", full.names = TRUE)
knitr::include_graphics(goodonly[1])
```

## exp hap
### Expected Haplotypes
Across the region being imputed (x-axis, physical position), either the fraction (normal version) or log10 of the expected number of haplotypes each sample carries from each ancestral haplotype. The most important thing to see in this plot is that there is not drastic, but rather gradual, movement between the relative proportions of the ancestral haplotypes. Large swings in ancestral haplotype sums indicate the heuristics are working sub-optimally (though this could also be interesting biology!). It is also useful to check here how many ancestral haplotypes are being used, to see if some could be trimmed, for example if some are very infrequently used, though this is rare, as STITCH will try to fill them up. 

_- excerpt from the STITCH [documentation](https://github.com/rwdavies/STITCH/blob/5a5ba98442a105548587ed8cafbf00aece212c94/plots.md)_.

### Expected Haplotypes Plots
```{r hapnorm, dev='jpeg', out.width="100%"}
hapsum <- list.files(path = plotdir, pattern = "hapSum", full.names = TRUE)
knitr::include_graphics(hapsum)
```

## alpha mat
### Alpha Matrix
Likely not informative for general use. `alphaMat` is the name of the internal `STITCH` variable that stores the probability of jumping into an ancestral haplotype conditional on a jump. Un-normalized is with respect to movement of all samples, while normalized has sum 1. 

_- excerpt from the STITCH [documentation](https://github.com/rwdavies/STITCH/blob/5a5ba98442a105548587ed8cafbf00aece212c94/plots.md)_.

### Alpha Matrix Plots
```{r alphamat, dev='jpeg', out.width="100%"}
alphamat <- list.files(path = plotdir, pattern = "alphaMat", full.names = TRUE)
knitr::include_graphics(alphamat)
```