<!DOCTYPE html>
<html lang="en" class="h-full">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="turbo-root" content="/harpy">
    <meta name="turbo-cache-control" content="no-cache" data-turbo-track="reload" data-track-token="3.11.0.807373459261">

    <!-- See retype.com -->
    <meta name="generator" content="Retype 3.11.0">

    <!-- Primary Meta Tags -->
    <title> Intro to linked-read data | Harpy haplotag</title>
    <meta name="title" content=" Intro to linked-read data | Harpy haplotag">
    <meta name="description" content="What you ought to know about linked reads">

    <!-- Canonical -->
    <link rel="canonical" href="https://pdimens.github.io/harpy/getting_started/linked_read_data/">

    <!-- Open Graph / Facebook -->
    <meta property="og:type" content="website">
    <meta property="og:url" content="https://pdimens.github.io/harpy/getting_started/linked_read_data/">
    <meta property="og:title" content=" Intro to linked-read data | Harpy haplotag">
    <meta property="og:description" content="What you ought to know about linked reads">
    <meta property="og:image" content="https://pdimens.github.io/harpy/static/linked_reads.svg">

    <!-- Twitter -->
    <meta property="twitter:card" content="summary_large_image">
    <meta property="twitter:url" content="https://pdimens.github.io/harpy/getting_started/linked_read_data/">
    <meta property="twitter:title" content=" Intro to linked-read data | Harpy haplotag">
    <meta property="twitter:description" content="What you ought to know about linked reads">
    <meta property="twitter:image" content="https://pdimens.github.io/harpy/static/linked_reads.svg">

    <script data-cfasync="false">(function(){var cl=document.documentElement.classList,ls=localStorage.getItem("retype_scheme"),hd=cl.contains("dark"),hl=cl.contains("light"),wm=window.matchMedia&&window.matchMedia("(prefers-color-scheme: dark)").matches;if(ls==="dark"||(!ls&&wm&&!hd&&!hl)){cl.remove("light");cl.add("dark")}else if(ls==="light"||(!ls&&!wm&&!hd&&!hl)){cl.remove("dark");cl.add("light")}})();</script>

    <link href="../../static/favicon.png" rel="icon">
    <link href="../../resources/css/retype.css?v=3.11.0.807373459261" rel="stylesheet">

    <script data-cfasync="false" src="../../resources/js/config.js?v=3.11.0.807373459261" data-turbo-eval="false" defer></script>
    <script data-cfasync="false" src="../../resources/js/retype.js?v=3.11.0" data-turbo-eval="false" defer></script>
    <script id="lunr-js" data-cfasync="false" src="../../resources/js/lunr.js?v=3.11.0.807373459261" data-turbo-eval="false" defer></script>
    <link href="../../resources/css/katex.css?v=3.11.0" rel="stylesheet">
    <script id="katex-js" data-cfasync="false" src="../../resources/js/katex.js?v=3.11.0" defer></script>
</head>
<body>
    <div id="retype-app" class="relative text-base antialiased text-base-text bg-base-bg font-body">
        <div class="absolute bottom-0 left-0" style="top: 5rem; right: 50%"></div>
    
        <header id="retype-header" class="sticky top-0 z-30 flex w-full h-16 bg-header-bg border-b border-header-border md:h-20">
            <div class="container relative flex items-center justify-between pr-6 grow md:justify-start">
                <!-- Mobile menu button skeleton -->
                <button v-cloak class="skeleton retype-mobile-menu-button flex items-center justify-center shrink-0 overflow-hidden dark:text-white focus:outline-none rounded-full w-10 h-10 ml-3.5 md:hidden"><svg xmlns="http://www.w3.org/2000/svg" class="mb-px shrink-0" width="24" height="24" viewBox="0 0 24 24" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor"><path d="M2 4h20v2H2zM2 11h20v2H2zM2 18h20v2H2z"></path></g></svg></button>
                <div v-cloak id="retype-sidebar-left-toggle-button"></div>
        
                <!-- Logo -->
                <div class="flex items-center justify-between h-full py-2 md:w-75">
                    <div class="flex items-center px-2 md:px-6">
                        <a id="retype-branding-logo" href="../../" class="flex items-center leading-snug text-2xl">
                            <span class="w-10 mr-2 grow-0 shrink-0 overflow-hidden">
                                <img class="max-h-10 dark:hidden md:inline-block" src="../../static/logo_navbar.png">
                                <img class="max-h-10 hidden dark:inline-block" src="../../static/logo_navbar.png">
                            </span>
                            <span class="dark:text-white font-bold line-clamp-1 md:line-clamp-2">Harpy</span>
                        </a><span id="retype-branding-label" class="inline-flex mt-1 px-2 py-1 ml-4 text-xs font-medium leading-none items-center rounded-md bg-branding-label-bg text-branding-label-text ring-1 ring-branding-label-border ring-inset md:inline-block">v2.5</span>
                    </div>
        
                    <span class="hidden h-8 border-r md:inline-block border-base-border"></span>
                </div>
        
                <div class="flex justify-between md:grow">
                    <!-- Top Nav -->
                    <nav id="retype-header-nav" class="hidden md:flex">
                        <ul class="flex flex-col mb-4 md:pl-16 md:mb-0 md:flex-row md:items-center">
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-header-tex font-header-text-weight hover:text-header-text-hover" href="https://www.therkildsenlab.org/" target="_blank">
                                    <img class="mb-px mr-1" width="18" alt="" src="https://www.therkildsenlab.org/uploads/1/4/5/5/145548022/published/new-logo-smaller.png">
                                    <span>Therkildsen Lab</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-header-text font-header-text-weight hover:text-header-text-hover" href="https://github.com/pdimens/harpy" target="_blank">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M12 1C5.923 1 1 5.923 1 12c0 4.867 3.149 8.979 7.521 10.436.55.096.756-.233.756-.522 0-.262-.013-1.128-.013-2.049-2.764.509-3.479-.674-3.699-1.292-.124-.317-.66-1.293-1.127-1.554-.385-.207-.936-.715-.014-.729.866-.014 1.485.797 1.691 1.128.99 1.663 2.571 1.196 3.204.907.096-.715.385-1.196.701-1.471-2.448-.275-5.005-1.224-5.005-5.432 0-1.196.426-2.186 1.128-2.956-.111-.275-.496-1.402.11-2.915 0 0 .921-.288 3.024 1.128a10.193 10.193 0 0 1 2.75-.371c.936 0 1.871.123 2.75.371 2.104-1.43 3.025-1.128 3.025-1.128.605 1.513.221 2.64.111 2.915.701.77 1.127 1.747 1.127 2.956 0 4.222-2.571 5.157-5.019 5.432.399.344.743 1.004.743 2.035 0 1.471-.014 2.654-.014 3.025 0 .289.206.632.756.522C19.851 20.979 23 16.854 23 12c0-6.077-4.922-11-11-11Z"/>
                                        </g>
                                    </svg>
                                    <span>Source</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-header-text font-header-text-weight hover:text-header-text-hover" href="https://github.com/pdimens/harpy/issues/new/choose" target="_blank">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M7.72.22a.75.75 0 0 1 1.06 0l1.204 1.203A4.98 4.98 0 0 1 12 1c.717 0 1.4.151 2.016.423L15.22.22a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-.971.972A4.991 4.991 0 0 1 17 6v1.104a2.755 2.755 0 0 1 1.917 1.974l1.998-.999a.75.75 0 0 1 .67 1.342L19 10.714V13.5l3.25.003a.75.75 0 0 1 0 1.5L19 15.001V16c0 .568-.068 1.134-.204 1.686l.04.018 2.75 1.375a.75.75 0 1 1-.671 1.342l-2.638-1.319A6.998 6.998 0 0 1 12 23a6.998 6.998 0 0 1-6.197-3.742l-2.758 1.181a.752.752 0 0 1-1.064-.776.752.752 0 0 1 .474-.602l2.795-1.199A6.976 6.976 0 0 1 5 16v-.996H1.75a.75.75 0 0 1 0-1.5H5v-2.79L2.415 9.42a.75.75 0 0 1 .67-1.342l1.998.999A2.756 2.756 0 0 1 7 7.104V6a4.99 4.99 0 0 1 1.69-3.748l-.97-.972a.75.75 0 0 1 0-1.06ZM6.5 9.75V16a5.5 5.5 0 1 0 11 0V9.75c0-.69-.56-1.25-1.25-1.25h-8.5c-.69 0-1.25.56-1.25 1.25ZM8.5 7h7V6a3.5 3.5 0 1 0-7 0Z"/>
                                        </g>
                                    </svg>
                                    <span>Submit Issue</span>
                                </a>
                            </li>
        
                        </ul>
                    </nav>
        
                    <!-- Header Right Skeleton -->
                    <div v-cloak class="flex justify-end grow skeleton">
        
                        <!-- Search input mock -->
                        <div class="relative hidden w-40 lg:block lg:max-w-sm lg:ml-auto">
                            <div class="absolute flex items-center justify-center h-full pl-3 dark:text-dark-300">
                                <svg xmlns="http://www.w3.org/2000/svg" class="icon-base" width="16" height="16" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 1px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                            </div>
                            <input class="w-full h-10 placeholder-search-placeholder transition-colors duration-200 ease-in bg-search-bg border border-transparent rounded md:text-sm hover:border-search-border-hover focus:outline-none focus:border-search-border-focus" style="padding: 0.625rem 0.75rem 0.625rem 2rem" type="text" placeholder="Search">
                        </div>
        
                        <!-- Mobile search button -->
                        <div class="flex items-center justify-center w-10 h-10 lg:hidden">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="20" height="20" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                        </div>
        
                        <!-- Dark mode switch placeholder -->
                        <div class="w-10 h-10 lg:ml-2"></div>
        
                        <!-- History button -->
                        <div class="flex items-center justify-center w-10 h-10" style="margin-right: -0.625rem;">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="22" height="22" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><g ><path d="M12.01 6.01c-.55 0-1 .45-1 1V12a1 1 0 00.4.8l3 2.22a.985.985 0 001.39-.2.996.996 0 00-.21-1.4l-2.6-1.92V7.01c.02-.55-.43-1-.98-1z"></path><path d="M12.01 1.91c-5.33 0-9.69 4.16-10.05 9.4l-.29-.26a.997.997 0 10-1.34 1.48l1.97 1.79c.19.17.43.26.67.26s.48-.09.67-.26l1.97-1.79a.997.997 0 10-1.34-1.48l-.31.28c.34-4.14 3.82-7.41 8.05-7.41 4.46 0 8.08 3.63 8.08 8.09s-3.63 8.08-8.08 8.08c-2.18 0-4.22-.85-5.75-2.4a.996.996 0 10-1.42 1.4 10.02 10.02 0 007.17 2.99c5.56 0 10.08-4.52 10.08-10.08.01-5.56-4.52-10.09-10.08-10.09z"></path></g></g></svg>
                        </div>
                    </div>
        
                    <div v-cloak class="flex justify-end grow">
                        <div id="retype-mobile-search-button"></div>
                        <doc-search-desktop></doc-search-desktop>
        
                        <doc-theme-switch class="lg:ml-2"></doc-theme-switch>
                        <doc-history></doc-history>
                    </div>
                </div>
            </div>
        </header>
    
    
        <div id="retype-container" class="container relative flex bg-white">
            <!-- Sidebar Skeleton -->
            <div v-cloak class="fixed flex flex-col shrink-0 duration-300 ease-in-out bg-sidebar-left-bg border-sidebar-left-border sidebar top-20 w-75 border-r h-screen md:sticky transition-transform skeleton">
            
            
                <div class="pl-6 mt-1 mb-4">
                    <div class="w-32 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                    <div class="w-48 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                    <div class="w-40 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                    <div class="w-32 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                    <div class="w-48 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                    <div class="w-40 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                </div>
            
                <div class="shrink-0 mt-auto bg-transparent dark:border-base-border">
                    <a class="flex items-center justify-center flex-nowrap h-16 text-gray-350 dark:text-dark-400 hover:text-gray-600 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                        <span class="text-xs whitespace-nowrap">Powered by</span>
                        <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                    </a>
                </div>
            </div>
            
            <!-- Sidebar component -->
            <doc-sidebar v-cloak>
                <template #sidebar-footer>
                    <div class="shrink-0 mt-auto border-t md:bg-transparent md:border-none dark:border-base-border">
            
                        <div class="py-3 px-6 md:hidden border-b dark:border-base-border">
                            <nav>
                                <ul class="flex flex-wrap justify-center items-center">
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-header-tex font-header-text-weight hover:text-header-text-hover" href="https://www.therkildsenlab.org/" target="_blank">
                                            <img class="mb-px mr-1" width="18" alt="" src="https://www.therkildsenlab.org/uploads/1/4/5/5/145548022/published/new-logo-smaller.png">
                                            <span>Therkildsen Lab</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-header-text font-header-text-weight hover:text-header-text-hover" href="https://github.com/pdimens/harpy" target="_blank">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M12 1C5.923 1 1 5.923 1 12c0 4.867 3.149 8.979 7.521 10.436.55.096.756-.233.756-.522 0-.262-.013-1.128-.013-2.049-2.764.509-3.479-.674-3.699-1.292-.124-.317-.66-1.293-1.127-1.554-.385-.207-.936-.715-.014-.729.866-.014 1.485.797 1.691 1.128.99 1.663 2.571 1.196 3.204.907.096-.715.385-1.196.701-1.471-2.448-.275-5.005-1.224-5.005-5.432 0-1.196.426-2.186 1.128-2.956-.111-.275-.496-1.402.11-2.915 0 0 .921-.288 3.024 1.128a10.193 10.193 0 0 1 2.75-.371c.936 0 1.871.123 2.75.371 2.104-1.43 3.025-1.128 3.025-1.128.605 1.513.221 2.64.111 2.915.701.77 1.127 1.747 1.127 2.956 0 4.222-2.571 5.157-5.019 5.432.399.344.743 1.004.743 2.035 0 1.471-.014 2.654-.014 3.025 0 .289.206.632.756.522C19.851 20.979 23 16.854 23 12c0-6.077-4.922-11-11-11Z"/>
                                                </g>
                                            </svg>
                                            <span>Source</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-header-text font-header-text-weight hover:text-header-text-hover" href="https://github.com/pdimens/harpy/issues/new/choose" target="_blank">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M7.72.22a.75.75 0 0 1 1.06 0l1.204 1.203A4.98 4.98 0 0 1 12 1c.717 0 1.4.151 2.016.423L15.22.22a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-.971.972A4.991 4.991 0 0 1 17 6v1.104a2.755 2.755 0 0 1 1.917 1.974l1.998-.999a.75.75 0 0 1 .67 1.342L19 10.714V13.5l3.25.003a.75.75 0 0 1 0 1.5L19 15.001V16c0 .568-.068 1.134-.204 1.686l.04.018 2.75 1.375a.75.75 0 1 1-.671 1.342l-2.638-1.319A6.998 6.998 0 0 1 12 23a6.998 6.998 0 0 1-6.197-3.742l-2.758 1.181a.752.752 0 0 1-1.064-.776.752.752 0 0 1 .474-.602l2.795-1.199A6.976 6.976 0 0 1 5 16v-.996H1.75a.75.75 0 0 1 0-1.5H5v-2.79L2.415 9.42a.75.75 0 0 1 .67-1.342l1.998.999A2.756 2.756 0 0 1 7 7.104V6a4.99 4.99 0 0 1 1.69-3.748l-.97-.972a.75.75 0 0 1 0-1.06ZM6.5 9.75V16a5.5 5.5 0 1 0 11 0V9.75c0-.69-.56-1.25-1.25-1.25h-8.5c-.69 0-1.25.56-1.25 1.25ZM8.5 7h7V6a3.5 3.5 0 1 0-7 0Z"/>
                                                </g>
                                            </svg>
                                            <span>Submit Issue</span>
                                        </a>
                                    </li>
            
                                </ul>
                            </nav>
                        </div>
            
                        <a class="flex items-center justify-center flex-nowrap h-16 text-gray-350 dark:text-dark-400 hover:text-gray-600 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                            <span class="text-xs whitespace-nowrap">Powered by</span>
                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                        </a>
                    </div>
                </template>
            </doc-sidebar>
    
            <div class="grow min-w-0 bg-body-bg">
                <!-- Render "toolbar" template here on api pages --><!-- Render page content -->
                <div class="flex">
                    <div id="retype-main" class="min-w-0 p-4 grow md:px-16">
                        <main class="relative pb-12 lg:pt-2">
                            <div class="retype-markdown" id="retype-content">
                                <!-- Rendered if sidebar right is enabled -->
                                <div id="retype-sidebar-right-toggle"></div>
                                <nav id="retype-breadcrumb" class="breadcrumb" aria-label="Breadcrumb">
                                    <ol role="list">
                                        <li class="pr-4">
                                            <a href="../../" class="pr-4">
                                                Home
                                            </a>
                                            <svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" width="18" height="18" viewBox="0 0 24 24" aria-hidden="true">
                                                <path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"/>
                                            </svg>
                                        </li>
                                        <li class="pr-4">
                                            <span class="pr-4">
                                                Getting Started
                                            </span>
                                            <svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" width="18" height="18" viewBox="0 0 24 24" aria-hidden="true">
                                                <path d="M8.72 18.78a.75.75 0 0 1 0-1.06L14.44 12 8.72 6.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018l6.25 6.25a.75.75 0 0 1 0 1.06l-6.25 6.25a.75.75 0 0 1-1.06 0Z"/>
                                            </svg>
                                        </li>
                                        <li class="pr-4">
                                            <a href="../../getting_started/linked_read_data/" class="pr-4">
                                                Linked-â€‹Read Data
                                            </a>
                                        </li>
                                    </ol>
                                </nav>
                
                                <!-- Page content  -->
<doc-anchor-target id="intro-to-linked-read-data" class="break-words">
    <h1>
        <doc-anchor-trigger class="header-anchor-trigger" to="#intro-to-linked-read-data">#</doc-anchor-trigger>
        <span><svg xmlns="http://www.w3.org/2000/svg" class="docs-icon" viewBox="0 0 24 24" role="presentation"><g fill="currentColor"><g><path d="M16.5 19.25a3.25 3.25 0 1 1 6.5 0 3.25 3.25 0 0 1-6.5 0Zm3.25-1.75a1.75 1.75 0 1 0 .001 3.501 1.75 1.75 0 0 0-.001-3.501Z"/><path d="M13.905 1.72a.75.75 0 0 1 0 1.06L12.685 4h4.065a3.75 3.75 0 0 1 3.75 3.75v8.75a.75.75 0 0 1-1.5 0V7.75a2.25 2.25 0 0 0-2.25-2.25h-4.064l1.22 1.22a.75.75 0 0 1-1.061 1.06l-2.5-2.5a.75.75 0 0 1 0-1.06l2.5-2.5a.75.75 0 0 1 1.06 0ZM7.5 4.75a3.25 3.25 0 1 1-6.5 0 3.25 3.25 0 0 1 6.5 0ZM4.25 6.5a1.75 1.75 0 1 0-.001-3.501A1.75 1.75 0 0 0 4.25 6.5Z"/><path d="M10.095 22.28a.75.75 0 0 1 0-1.06l1.22-1.22H7.25a3.75 3.75 0 0 1-3.75-3.75V7.5a.75.75 0 0 1 1.5 0v8.75a2.25 2.25 0 0 0 2.25 2.25h4.064l-1.22-1.22a.748.748 0 0 1 .332-1.265.75.75 0 0 1 .729.205l2.5 2.5a.75.75 0 0 1 0 1.06l-2.5 2.5a.75.75 0 0 1-1.06 0Z"/></g></g></svg> Intro to linked-read data</span>
    </h1>
</doc-anchor-target>
<p>Harpy was originally tailor-made for haplotagging linked read data, but now it works for most linked-read technologies
along with non-linked read data. BRL/LRTK/LongRanger are similar pieces of software for Tellseq, stLFR, and 10X linked-read
data. But, what if you don&#x27;t use linked reads (yet) and want to understand what it actually is? This post walks you through what
linked-read data is and some of the concepts that are unique to it that makes it different than your typical short-read data.</p>
<doc-panel id="you-will-need">
    <template #title><svg xmlns="http://www.w3.org/2000/svg" class="docs-icon" viewBox="0 0 24 24" role="presentation"><g fill="currentColor"><g><path d="M3.5 3.75a.25.25 0 0 1 .25-.25h13.5a.25.25 0 0 1 .25.25v10a.75.75 0 0 0 1.5 0v-10A1.75 1.75 0 0 0 17.25 2H3.75A1.75 1.75 0 0 0 2 3.75v16.5c0 .966.784 1.75 1.75 1.75h7a.75.75 0 0 0 0-1.5h-7a.25.25 0 0 1-.25-.25V3.75Z"/><path d="M6.25 7a.75.75 0 0 0 0 1.5h8.5a.75.75 0 0 0 0-1.5h-8.5Zm-.75 4.75a.75.75 0 0 1 .75-.75h4.5a.75.75 0 0 1 0 1.5h-4.5a.75.75 0 0 1-.75-.75Zm16.28 4.53a.75.75 0 1 0-1.06-1.06l-4.97 4.97-1.97-1.97a.75.75 0 1 0-1.06 1.06l2.5 2.5a.75.75 0 0 0 1.06 0l5.5-5.5Z"/></g></g></svg> You will need</template>
<ul>
<li>a cup of tea, coffee, or water</li>
<li>a can-do attitude</li>
</ul>
</doc-panel>
<doc-anchor-target id="linked-reads">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#linked-reads">#</doc-anchor-trigger>
        <span>Linked reads</span>
    </h2>
</doc-anchor-target>
<p>Let&#x27;s start with the most obvious, what <em>are</em> linked reads, which are sometimes called &quot;synthetic long reads&quot;? They are short read
(e.g. Illumina) data. What makes them different is that they contain an added DNA segment (&quot;barcode&quot;) that lets us associate
sequences as having originated from <strong>a single DNA molecule</strong>. That means if we have 4 sequences that all contain the same added barcode,
then we infer that they must have originated from the same original DNA molecule. Different barcode = different molecule of origin.
If the sequences would map to the same genomic region during sequence alignment, we would know that those sequences with the same
barcode originated from <strong>a single DNA fragment from a single homologous chromosome from a single cell</strong>. That&#x27;s right, <strong>built-in phase information</strong>.
<figure class="content-center">
    <img src="../../static/linked_reads.svg" alt="simplified overview of linked reads" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">simplified overview of linked reads</figcaption>
</figure>
</p>
<doc-anchor-target id="what-do-they-look-like">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#what-do-they-look-like">#</doc-anchor-trigger>
        <span>What do they look like?</span>
    </h3>
</doc-anchor-target>
<p>Linked-read data is sequence data as you would expect it, encoded in a FASTQ file. The first processing step of
linked-read data is demultiplexing to split the raw Illumina-generated batch FASTQ file into samples (if multisample)
and identify/validate the linked-read barcode on every sequence. For 10X data, the barcode would <em>stay</em> inline with
the sequence (to make it LongRanger compatible), but for other varieties (haplotagging, stLFR, etc.) you would also
remove the barcode from the sequence and preserve it using the <code v-pre>BX:Z</code> tag in the sequence headers. The demultiplexing process
is generally similar between non-10X linked-read technologies: a nucleotide barcode gets identified and moved from the sequence
to the read header under the <code v-pre>BX:Z</code> tag. The diagram below preserves the nucleotide barcode under the <code v-pre>OX:Z</code> tag and recodes
it under <code v-pre>BX:Z</code> using the haplotagging &quot;ACBD&quot; segment-aware format, however it would also be valid to just keep the nucleotide
barcode under <code v-pre>BX:Z</code>. Linked-read software is variable in its flexibility towards barcode formatting.
<figure class="content-center">
    <img src="../../static/lr_conversion.png" alt="10X linked read data before and after demultiplexing" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">10X linked read data before and after demultiplexing</figcaption>
</figure>
</p>
<doc-anchor-target id="deconvolution">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#deconvolution">#</doc-anchor-trigger>
        <span>Deconvolution</span>
    </h4>
</doc-anchor-target>
<p>There are approaches to deconvolve linked read data (or &quot;deconvolute&quot;, if you&#x27;re indifferent to the burden of an extra syllable).
In this context, deconvolution is finding out which sequences are sharing barcodes by chance rather than because they originated
from the same DNA molecule. The likelihood of it happening is usually low, but it&#x27;s not impossible. When the algorithm determines
that barcodes are being shared by unrelated molecules, the barcode typically gets suffixed with a hyphenated number (e.g.
<code v-pre>BX:Z:ATACG</code> becomes <code v-pre>BX:Z:ATACG-1</code>). As of this writing, there are only a few pieces of software that can deconvolve linked-read data and
do so with varying degrees of success and computational resource requirements. Similarly, linked-read software is variable in its
flexibility towards accepting the deconvolved-barcode format.</p>
<doc-anchor-target id="linked-read-varieties">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#linked-read-varieties">#</doc-anchor-trigger>
        <span>Linked-read varieties</span>
    </h3>
</doc-anchor-target>
<p>There are a handful of linked-read sample preparation methods (<doc-anchor-trigger to="#linked-read-data-types">read below</doc-anchor-trigger>), but that&#x27;s largely an implementation detail. All of those methods are
laboratory procedures to take genomic DNA and do the necessary modifications to fragment long DNA molecules, tag the resulting fragments with the same
DNA barcode, then add the necessary Illumina adapters. It&#x27;s not unlike the different RAD flavors (e.g. EZrad, ddRAD, 2B-rad)-- they all give you RAD data in the end,
but vary in how you get there in terms of cost and bench time. We obviously subscribe to haplotagging <span class="docs-emoji">&#x1F601;</span>.</p>
<doc-anchor-target id="failsafe">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#failsafe">#</doc-anchor-trigger>
        <span>Failsafe</span>
    </h3>
</doc-anchor-target>
<p>Unlike some other fancy well-touted sample preparation methods (<em>cough cough mate-pair</em>), linked-read data <strong>is</strong> whole genome
sequencing (WGS). What that means is that whether you use the linked-read information or not, the data will always be standard and
viable WGS compatible with whatever you would use WGS for. It&#x27;s WGS, but with a little extra info that goes a long way.</p>
<doc-anchor-target id="linked-read-library-performance">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#linked-read-library-performance">#</doc-anchor-trigger>
        <span>Linked-read library performance</span>
    </h2>
</doc-anchor-target>
<p>Because linked-read data has an extra dimension, there are some additional metrics that are useful to look at to evaluate how &quot;good&quot;
the linked-read library turned out. As a baseline for comparison, think of RAD or WGS data and the kinds of metrics you might look at
for library performance:</p>
<ul>
<li>coverage depth</li>
<li>coverage breadth</li>
<li>PCR/optical duplicates</li>
<li>coverage depth per sample</li>
</ul>
<p>Linked-read library performance also looks at that, but there&#x27;s a few extra parameters that help us assess performance:</p>
<doc-anchor-target id="molecule-coverage">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#molecule-coverage">#</doc-anchor-trigger>
        <span>Molecule Coverage</span>
    </h3>
</doc-anchor-target>
<p>Since linked reads are tagger per DNA molecule, we are interested in understanding coverage on a per-molecule basis. By &quot;molecule&quot;,
we are referring to the original DNA molecule from which barcoded sequences originated from.</p>
<doc-anchor-target id="reads-per-molecule">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#reads-per-molecule">#</doc-anchor-trigger>
        <span>reads per molecule</span>
    </h4>
</doc-anchor-target>
<p>It&#x27;s quite rare that all fragments of a single DNA molecule will get sequenced, so we are interested in getting an average number of
reads (sequences) per original DNA molecule. This is done by getting counts of all the reads with the same barcode, as unique barcodes
correspond with unique DNA molecules. It&#x27;s difficult to say exactly what a good number of reads per molecule would be, but &gt;2 would be
a good starting point (1 would be a singleton, see below). Having 3-6 would be decent, depending on your project goals.</p>
<doc-anchor-target id="percent-molecule-coverage">
    <h4>
        <doc-anchor-trigger class="header-anchor-trigger" to="#percent-molecule-coverage">#</doc-anchor-trigger>
        <span>percent molecule coverage</span>
    </h4>
</doc-anchor-target>
<p>Because only a few fragments from a DNA molecule end up getting sequenced, it would be good to understand how much of a molecule is
represented in seqences (breadth). It&#x27;s impossible to get an accurate calculation for this metric because we have no way of knowing
the size of the original DNA molecule, but what we can do is, after alignment, find the distance between the two furthest reads
sharing a barcode and calculate how many bases between them have been sequenced. It&#x27;s not perfect, but it&#x27;s something. A low number
would indicate large gaps between linked sequences (e.g. few sequences and/or a very large molecule), whereas a high number would
indicate small gaps between linked sequences (e.g. many sequences and/or a very small molecule).
<figure class="content-center">
    <img src="../../static/molecule_coverage.svg" alt="molecule coverage example for 3 linked-reads sharing the same barcode" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">molecule coverage example for 3 linked-reads sharing the same barcode</figcaption>
</figure>
</p>
<doc-anchor-target id="singletons">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#singletons">#</doc-anchor-trigger>
        <span>Singletons</span>
    </h3>
</doc-anchor-target>
<p>Because linked reads need to be, well, <em>linked</em>, we need to know exactly how many of the sequences actually share barcodes. A barcode
that only appears in one paired or unpaired read is a <strong>singleton</strong>, meaning it isn&#x27;t actually linked to any other sequence. Despite
having a linked-read barcode, the absence of other reads with the same barcode means the barcode information for that read (or read
pair) is mostly useless, aka it&#x27;s just a plain-regular short read and can be used as such. Having &lt;60% would be considered decent,
but you would want as few singletons as possible in a linked-read dataset. For perspective, if 90% of your reads were singletons, you
can think of that as &quot;90% of your data aren&#x27;t linked reads&quot;. The opposite of a singleton is when a barcode is shared by more than one
paired or unpaired read, aka a linked read.
<figure class="content-center">
    <img src="../../static/linked_singletons.svg" alt="singletons and linked reads" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">singletons and linked reads</figcaption>
</figure>
</p>
<doc-anchor-target id="linked-coverage">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#linked-coverage">#</doc-anchor-trigger>
        <span>Linked Coverage</span>
    </h2>
</doc-anchor-target>
<p>Because of the &quot;linked&quot; component of linked-read data, we have an additional kind of sequence coverage
to consider, which is <strong>linked coverage</strong>. Since linked-read barcodes preserve long-distance information, we can do an alternative kind of
coverage calculation where you pretend that all the sequences with a shared barcode are one big gapless sequence. If you use your
imagination that way, the name &quot;synthetic long reads&quot; begins to make more sense. Mathematically, the linked coverage (breadth and
depth) should always be higher than the coverage from just the alignments themselves, as gaps between linked reads are counted as
well. If you see that your linked depth is unusually high (e.g. the aligmment depth is 20 and linked depth is &gt;1000), then it&#x27;s
very likely your have barcode clashing that has not been deconvoluted. In other words, reads from different DNA molecules that are
sharing a barcode are mapped very far from each other and inflating the depth values for everything in between.
<figure class="content-center">
    <img src="../../static/linked_depth.svg" alt="linked depth" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">linked depth</figcaption>
</figure>
</p>
<p>In real data, this actually looks very cool. Below is a graph from the per-sample alignment report Harpy generates. It&#x27;s a circos
plot, which is a circular representation of round charts, and each wedge is a chromosome (start to finish), labelled by the &quot;2R&quot;,
&quot;2L&quot;, etc. around the perimeter. The inner circle (grey histogram) is the <strong>sequence alignment depth</strong> (i.e. the standard depth
calculation) and the outer circle (magenta histogram)
is the <strong>linked depth</strong> of those same data in those same 50kbp intervals. It has the most fascinating flower petal pattern, which
is due to the lower likelihood of reads linking at the edges of chromosomes.
<figure class="content-center">
    <img src="../../static/linked_depth_example.png" alt="linked depth in real data" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">linked depth in real data</figcaption>
</figure>
</p>
<doc-anchor-target id="linked-read-data-types">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#linked-read-data-types">#</doc-anchor-trigger>
        <span>Linked-read data types</span>
    </h2>
</doc-anchor-target>
<p>This isn&#x27;t really the place to go into the nitty-gritty of the different linked-read chemistries
(i.e. I don&#x27;t actually know the nuanced details), but it&#x27;s worth describing the obvious differences
of the raw (FASTQ) data. Knowing these details might help you make sense of compatibilties/incompatibilities
for software, or how you can convert between styles.</p>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="compact">
<thead>
<tr>
<th style="text-align: left;">Type</th>
<th style="text-align: left;">Location</th>
<th style="text-align: left;">Format</th>
<th style="text-align: left;">Invalid Encoding</th>
<th style="text-align: left;">Example</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">Standard</td>
<td style="text-align: left;"><code v-pre>BX:Z</code> and <code v-pre>VX:i</code> tags</td>
<td style="text-align: left;">any</td>
<td style="text-align: left;"><code v-pre>VX:i:0</code></td>
<td style="text-align: left;"><code v-pre>BX:Z:31_442_512 VX:i:1</code></td>
</tr>
<tr>
<td style="text-align: left;">Haplotagging</td>
<td style="text-align: left;"><code v-pre>BX:Z</code> tag</td>
<td style="text-align: left;"><code v-pre>ACBD</code></td>
<td style="text-align: left;"><code v-pre>00</code> segment</td>
<td style="text-align: left;"><code v-pre>BX:Z:A04C54B96D11</code></td>
</tr>
<tr>
<td style="text-align: left;">stLFR</td>
<td style="text-align: left;">end of sequence ID</td>
<td style="text-align: left;"><code v-pre>#1_2_3</code></td>
<td style="text-align: left;"><code v-pre>0</code> segment</td>
<td style="text-align: left;"><code v-pre>@A003432423434:1:324#12_432_1</code></td>
</tr>
<tr>
<td style="text-align: left;">TELLseq</td>
<td style="text-align: left;">end of sequence ID</td>
<td style="text-align: left;"><code v-pre>:ATCG</code></td>
<td style="text-align: left;"><code v-pre>N</code></td>
<td style="text-align: left;"><code v-pre>@A003432423434:1:324:TTACCACGAGG</code></td>
</tr>
<tr>
<td style="text-align: left;">10X</td>
<td style="text-align: left;">R1 read</td>
<td style="text-align: left;"><code v-pre>ATCG</code></td>
<td style="text-align: left;"><code v-pre>N</code></td>
<td style="text-align: left;"><code v-pre>AGGTTGGGTAAGATA...</code></td>
</tr>
</tbody>
</table>
</div>
<doc-tabs>
<doc-tab v-cloak id="standard">
<template #title>Standard</template>
<p>We here at Harpy would like to <em>plead</em> for linked-read practitioners to adopt a standardized representation of linked-read barcodes.
We don&#x27;t really care what linked-read technology you use (although we obviously prefer haplotagging), but we do care about
the data formats being consistent. The classic joke in population genetics is that before you write a new piece of software,
you need to first write a new file format. We really don&#x27;t want that here-- inconsistent linked-read data formats make it harder
for developers to write great software that is compatible with all linked-read data types when they need to account for every possible
location a barcode could be in, what a valid vs invalid barcode looks like for that technology, etc. So, we are humbly asking
(nay, begging) for there to be a standard data format (FASTQ and SAM/BAM) that uses two SAM tags to encode barcodes:
<figure class="content-center">
    <img src="../../static/lr_standard.svg" alt="&quot;Standard&quot; linked-read FASTQ format" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">&quot;Standard&quot; linked-read FASTQ format</figcaption>
</figure>
</p>
<ul>
<li><code v-pre>BX:Z</code> tag to record the barcode, the format of which is irrelevant
<ul>
<li>it could be haplotagging <code v-pre>ACBD</code>, stLFR <code v-pre>1_2_3</code>, nucleotides, <em>whatever</em></li>
</ul>
</li>
<li><code v-pre>VX:i</code> tag to record if the corresponding barcode is valid
<ul>
<li><code v-pre>0</code> is invalid</li>
<li><code v-pre>1</code> is valid</li>
</ul>
</li>
</ul>
<p>This format makes it the responsibility of the early data processing software
(as early as demultiplexing) to encode the data in this format for downstream
processing. Our hope is that this htslib-compliant format is adopted to reduce
the fragmentation in the software ecosystem and reduce the need for file format
conversions. It also creates a blueprint for a generic file encoding for any new
linked-read methods that are being developed or can be developed in the future.</p>
<doc-panels>
   <doc-panel id="fastq-example" collapsed>
       <template #title>FASTQ example</template>
<div class="codeblock-wrapper"><doc-codeblock>
<div class="codeblock-title">stLFR style FASTQ header (valid barcode)</div>
<pre translate="no" class="language-none"><code v-pre class="language-none">@A00470:481:HNYFWDRX2:1:2101:29532:1063/1 BX:Z:45_11_361 VX:i:1</code></pre>
</doc-codeblock></div>
<div class="codeblock-wrapper"><doc-codeblock>
<div class="codeblock-title">haplotagging style FASTQ header (invalid barcode)</div>
<pre translate="no" class="language-none"><code v-pre class="language-none">@A00470:481:HNYFWDRX2:1:2101:29532:1063/1 BX:Z:A78C00B14D96 VX:i:0</code></pre>
</doc-codeblock></div>
   </doc-panel>
   <doc-panel id="sambam-example" collapsed>
       <template #title>SAM/BAM example</template>
<div class="codeblock-wrapper"><doc-codeblock>
<div class="codeblock-title">haplotagging style SAM record (valid barcode)</div>
<pre translate="no" class="language-none"><code v-pre class="language-none">A00470:481:HNYFWDRX2:1:2229:29912:29778 99      2R      1222    40      19S80M  =   1500     428     AGATGTGTATAAGAGACAGAGTTATGTCATTTTAAGCGGTCAAAATGGGTGAATTTCCGATTTCAAGTAATAGGCGAACTCAAGATACCTTCTACAGAT  FFFFFFFFFFFFFFFFF:FFFFF:FFFFF:FFFF:FFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:,FFFFFFF:FFFFF  NM:i:0  MD:Z:80 MC:Z:150M       AS:i:80      XS:i:80 RG:Z:sample1    MI:i:1  VX:i:1  BX:Z:A55C67B91D96</code></pre>
</doc-codeblock></div>
<div class="codeblock-wrapper"><doc-codeblock>
<div class="codeblock-title">nucleotide (TELLseq/10X) style SAM record (invalid barcode)</div>
<pre translate="no" class="language-none"><code v-pre class="language-none">A00470:481:HNYFWDRX2:1:2229:29912:29778 99      2R      1222    40      19S80M  =   1500     428     AGATGTGTATAAGAGACAGAGTTATGTCATTTTAAGCGGTCAAAATGGGTGAATTTCCGATTTCAAGTAATAGGCGAACTCAAGATACCTTCTACAGAT  FFFFFFFFFFFFFFFFF:FFFFF:FFFFF:FFFF:FFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:,FFFFFFF:FFFFF  NM:i:0  MD:Z:80 MC:Z:150M       AS:i:80      XS:i:80 RG:Z:sample1    MI:i:1  VX:i:0  BX:Z:TACANNNCACAGAG</code></pre>
</doc-codeblock></div>
   </doc-panel>
</doc-panels>
</doc-tab>
<doc-tab v-cloak id="haplotagging">
<template #title>Haplotagging</template>
<p>Unlike the available chemistries, haplotagging is non-commercial (DIY!!). Haplotagging barcodes are combinatorial and
are made up of four 6bp segments. Two of these segments (&quot;A&quot; and &quot;C&quot;) are the first 12bp of the <span class="math">I1</span> read and
the other two (&quot;B&quot; and &quot;D&quot;) are the first 12bp of the <span class="math">I2</span> read, both of which are provided by Illumina for standard sequencing runs.
Because of this segment design, there are <span class="math">96^4</span> (~84 million) possible barcode combinations (~900,000 per sample).
The barcodes are stored in the sequence header under the <code v-pre>BX:Z</code> SAM tag, recoded in their &quot;<code v-pre>ACBD</code>&quot; format.
<figure class="content-center">
    <img src="../../static/lr_haplotagging.svg" alt="Haplotagging data format" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">Haplotagging data format</figcaption>
</figure>
</p>
<ul>
<li>4 barcode segments
<ul>
<li><code v-pre>A</code> segment is the first 6bp of the <span class="math">I1</span> read</li>
<li><code v-pre>C</code> segment is the next 6bp of the <span class="math">I1</span> read (7-12)</li>
<li><code v-pre>B</code> segment is the first 6bp of the <span class="math">I2</span> read</li>
<li><code v-pre>D</code> segment is the next 6bp of the <span class="math">I2</span> read (7-12)</li>
</ul>
</li>
<li>barcode stored as <code v-pre>BX:Z</code> tag in the read header in <code v-pre>ACBD</code> format
<ul>
<li>e.g. <code v-pre>@A003432423434:1:324 BX:Z:A45C01B84D21</code></li>
</ul>
</li>
</ul>
</doc-tab>
<doc-tab v-cloak id="stlfr">
<template #title>stLFR</template>
<p>Another of the presently available commercial linked-read options. stLFR data uses combinatorial barcodes
made up for three 10bp segments which are at the end of the <span class="math">R2</span> read. Demultiplexing these data results
in the barcode being moved to the sequence ID using a pound (<code v-pre>#</code>) sign between the sequence ID and barcode, with
the barcode recoded in the <code v-pre>1_2_3</code> format, where each segment is an integer.
<figure class="content-center">
    <img src="../../static/lr_stlfr.svg" alt="stLFR FASTQ data format" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">stLFR FASTQ data format</figcaption>
</figure>
</p>
<ul>
<li>depending on the link sequence between segments, will be either the last 54bp or 42bp of the <span class="math">R2</span> read
<ul>
<li>54 base barcode: 10+6+10+18+10</li>
<li>42 base barcode: 10+6+10+6+10</li>
</ul>
</li>
<li>barcode appended to sequence header with <code v-pre>#</code> sign
<ul>
<li>e.g. <code v-pre>@A003432423434:1:324#12_432_1</code></li>
</ul>
</li>
<li>advertised to have a capacity over 3.6 <strong>billion</strong>, with up to 50 <strong>million</strong> per sample (actual results may vary)</li>
</ul>
</doc-tab>
<doc-tab v-cloak id="tellseq">
<template #title>TELLseq</template>
<p>One of the presently available commercial linked-read options. TELLseq data is very similar to 10X, except the
barcode is 18bp long and contained in the <span class="math">I1</span> read that Illumina provides with the standard <span class="math">R1</span> and <span class="math">R2</span> reads. The
barcode gets appended in the read header using a colon (<code v-pre>:</code>).
<figure class="content-center">
    <img src="../../static/lr_tellseq.svg" alt="Haplotagging FASTQ data format" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">Haplotagging FASTQ data format</figcaption>
</figure>
</p>
<ul>
<li>barcode is the first 18bp of the I1 read</li>
<li>barcode is appended to sequence header
<ul>
<li>e.g. <code v-pre>@A00234534562:1:544:AATTATACCACAGCGGTA</code></li>
</ul>
</li>
<li>advertised to have a capacity over 2 <strong>billion</strong> barcodes, but realistically use &lt;24 <strong>million</strong></li>
</ul>
</doc-tab>
<doc-tab v-cloak id="10x">
<template #title>10X</template>
<p>The elder of the bunch and also a discontinued commercial product. 10X-style FASTQ files have the linked-read barcode
as the first 16bp of the forward (<span class="math">R1</span>) read. For these data to be compatible with the 10X LongRanger suite,
the barcode <strong>must</strong> stay in the read. Moving the first 16bp into the read header breaks LongRanger compatibility.
<figure class="content-center">
    <img src="../../static/lr_10x.svg" alt="10X FASTQ data format" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">10X FASTQ data format</figcaption>
</figure>
</p>
<ul>
<li>barcode is the first 16bp of the <span class="math">R1</span> read</li>
<li>barcode stays in the sequence data for LongRanger compatibility</li>
<li>limited to ~4.7 million barcodes</li>
</ul>
</doc-tab>
</doc-tabs>
<doc-anchor-target id="barcode-thresholds">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#barcode-thresholds">#</doc-anchor-trigger>
        <span>Barcode thresholds</span>
    </h2>
</doc-anchor-target>
<p>By the nature of linked read technologies, there will (almost always) be more DNA fragments than unique barcodes for them. As a result,
it&#x27;s common for barcodes to reappear in sequences. Rather than incorrectly assume that all sequences/alignments with the same barcode
originated from the same orignal DNA molecule, linked-read aware programs include a threshold parameter to determine a &quot;cutoff&quot; distance
between alignments with the same barcode. This parameter can be interpreted as &quot;if a barcode appears more than <code v-pre>X</code> base pairs away from the
same barcode (on the same contig), then we&#x27;ll consider them as originating from different molecules.&quot; If this threshold is lower, then
you are being more strict and indicating that alignments sharing barcodes must be closer together to be considered originating from the same
DNA molecule. Conversely, a higher threshold indicates you are being more lax and indicating barcodes can be further away from each other
and still be considered originating from the same DNA molecule. A threshold of 50kb-150kb is considered a decent balance, but you should choose
larger/smaller values if you have evidence to support them.</p>
<p><figure class="content-center">
    <img src="../../static/bc_threshold.png" alt="Molecule origin is determined by the distance between alignments with the same barcode relative to the specified threshold" class="rounded-image-rounded border-image-border border-image-border-width" />
    <figcaption class="caption">Molecule origin is determined by the distance between alignments with the same barcode relative to the specified threshold</figcaption>
</figure>
</p>
<div class="table-wrapper scrollbar overflow-hidden">
<table class="comfortable">
<thead>
<tr>
<th style="text-align: left;">Alignment distance</th>
<th style="text-align: left;">Inferred origin</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">less than threshold</td>
<td style="text-align: left;">same molecule</td>
</tr>
<tr>
<td style="text-align: left;">greater than threshold</td>
<td style="text-align: left;">different molecules</td>
</tr>
</tbody>
</table>
</div>
<div class="flex mb-6">
    <div class="shrink-0 w-1.5 rounded-tl-lg rounded-bl-lg bg-callout-warning"></div>
    <div class="flex w-full py-4 border border-l-0 rounded-tr-lg rounded-br-lg doc-callout bg-callout-base-bg border-callout-base-border" role="alert">
        <div class="flex items-center ml-4 h-7">
            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px text-callout-warning" width="22" height="22" viewBox="0 0 24 24" role="presentation">
                <g fill="currentColor"><g>
                    <path d="M22.48 15.59L14.01 1.45A2.968 2.968 0 0012.16.09c-.78-.19-1.58-.07-2.27.35-.41.25-.76.6-1.01 1.01v.01L.4 15.6c-.83 1.43-.33 3.27 1.1 4.1.45.26.95.4 1.48.4h16.95c.8-.01 1.55-.33 2.11-.9.56-.57.87-1.33.86-2.13a3.04 3.04 0 00-.42-1.48zm-1.87 2.21c-.19.19-.44.3-.69.3H2.99c-.17 0-.34-.05-.49-.13a.992.992 0 01-.37-1.35L10.6 2.48c.08-.14.2-.25.34-.33a.992.992 0 011.37.33l8.46 14.13c.09.15.13.32.13.49 0 .26-.1.51-.29.7z"></path>
                    <path d="M11.45 12.1c.55 0 1-.45 1-1v-4c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .56.45 1 1 1zM11.46 14.1c-.56 0-1 .45-1 1s.45 1 1 1 1-.45 1-1-.45-1-1-1z"></path>
                </g></g>
            </svg>
        </div>
        <div class="pr-5 ml-3 w-full">
            <h5>Potential SV detection obstruction</h5>
<p>This kind of deconvolving method relies on alignment distances, which may worsen
performance in finding structural variants larger than this threshold. For example,
two reads originating from the same DNA molecule may have aligned quite far from each other
due to mapping along the breakpoint of a very large inversion. If the inversion spans 3Mb,
a barcode threshold of 100kb will likely incorrectly deconvolve the two reads and assume
they shared a barcode by chance, which would hurt SV detection performance.</p>
        </div>
    </div>
</div>

                                
                                <!-- Required only on API pages -->
                                <doc-toolbar-member-filter-no-results></doc-toolbar-member-filter-no-results>
                            </div>
                            <footer id="retype-content-footer" class="clear-both">
                                <div class="flex flex-wrap items-center justify-between mt-14">
                                    <a id="retype-edit-this-page" class="my-2.5 inline-flex items-center text-sm whitespace-nowrap text-body-link font-body-link-weight hover:text-body-link-hover hover:underline" href="https://github.com/pdimens/harpy/edit/docs/Getting_Started/linked_read_data.md" target="_blank" rel="noopener">
                                        <svg class="mr-1.5" xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M20 12c-.55 0-1 .45-1 1v7c0 .55-.45 1-1 1H4c-.55 0-1-.45-1-1V6c0-.55.45-1 1-1h7c.55 0 1-.45 1-1s-.45-1-1-1H4C2.35 3 1 4.35 1 6v14c0 1.65 1.35 3 3 3h14c1.65 0 3-1.35 3-3v-7c0-.55-.45-1-1-1z" /><path d="M22.21 1.79c-1.18-1.18-3.24-1.18-4.41 0l-9.5 9.5c-.13.13-.22.29-.26.46l-1 4c-.08.34.01.7.26.95.18.2.44.3.7.3.08 0 .16-.01.24-.03l4-1c.18-.04.34-.13.46-.26l9.5-9.5c1.22-1.22 1.22-3.2.01-4.42zm-1.42 3l-9.3 9.3-2.11.53.53-2.11 9.3-9.3c.42-.42 1.16-.42 1.59 0 .43.43.43 1.15-.01 1.58z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        <span>Edit this page</span>
                                    </a>
                                </div>
                            
                                <nav id="retype-nextprev" class="print:hidden flex mt-14">
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 h-full flex items-center break-normal font-medium text-body-link border border-base-border hover:border-base-border-hover rounded-l-lg transition-colors duration-150 relative hover:z-5" href="../../getting_started/inputformat/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mr-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19 11H7.41l5.29-5.29a.996.996 0 10-1.41-1.41l-7 7a1 1 0 000 1.42l7 7a1.024 1.024 0 001.42-.01.996.996 0 000-1.41L7.41 13H19c.55 0 1-.45 1-1s-.45-1-1-1z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                            <span>
                                                <span class="block text-xs font-normal text-base-text-muted">Previous</span>
                                                <span class="block mt-1">Input Format</span>
                                            </span>
                                        </a>
                                    </div>
                            
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 -mx-px h-full flex items-center justify-end break-normal font-medium text-body-link border border-base-border hover:border-base-border-hover rounded-r-lg transition-colors duration-150 relative hover:z-5" href="../../getting_started/common_options/">
                                            <span>
                                                <span class="block text-xs font-normal text-right text-base-text-muted">Next</span>
                                                <span class="block mt-1">Common Options</span>
                                            </span>
                                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19.92 12.38a1 1 0 00-.22-1.09l-7-7a.996.996 0 10-1.41 1.41l5.3 5.3H5c-.55 0-1 .45-1 1s.45 1 1 1h11.59l-5.29 5.29a.996.996 0 000 1.41c.19.2.44.3.7.3s.51-.1.71-.29l7-7c.09-.09.16-.21.21-.33z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        </a>
                                    </div>
                                </nav>
                            </footer>
                        </main>
                
                        <div id="retype-page-footer" class="print:border-none border-t border-base-border pt-6 mb-8">
                            <footer class="flex flex-wrap items-center justify-between print:justify-center">
                                <div id="retype-footer-links" class="print:hidden">
                                    <ul class="flex flex-wrap items-center text-sm">
                                    </ul>
                                </div>
                                <div id="retype-copyright" class="print:justify-center py-2 text-footer-text font-footer-link-weight text-sm leading-relaxed"><p>Â© Copyright 2025. All rights reserved.</p></div>
                            </footer>
                        </div>
                    </div>
                
                    <!-- Rendered if sidebar right is enabled -->
                    <!-- Sidebar right skeleton-->
                    <div v-cloak class="fixed top-0 bottom-0 right-0 translate-x-full bg-sidebar-right-bg border-sidebar-right-border lg:sticky lg:border-l lg:shrink-0 lg:pt-6 lg:transform-none sm:w-1/2 lg:w-64 lg:z-0 md:w-104 sidebar-right skeleton">
                        <div class="pl-5">
                            <div class="w-32 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                            <div class="w-48 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                            <div class="w-40 h-3 mb-4 bg-skeleton-bg rounded-full loading"></div>
                        </div>
                    </div>
                
                    <!-- User should be able to hide sidebar right -->
                    <doc-sidebar-right v-cloak></doc-sidebar-right>
                </div>

            </div>
        </div>
    
        <doc-search-mobile></doc-search-mobile>
        <doc-back-to-top></doc-back-to-top>
    </div>


    <div id="retype-overlay-target"></div>

    <script data-cfasync="false">window.__DOCS__ = { "title": "Linked-Read Data", level: 2, icon: "file", hasPrism: false, hasMermaid: false, hasMath: true, tocDepth: 23 }</script>
</body>
</html>
