---
title: "Haplotype Phasing Summary"
date: "`r format(Sys.time(), '%d %b, %Y at %H:%M')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    mathjax: NULL
    logo: https://raw.githubusercontent.com/pdimens/harpy/docs/static/logo_report.png
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_report.png"
    navbar:
      - { title : "Docs", icon: "fa-book", href: "https://pdimens.github.io/harpy/", align: right }
      - { title : "Source", icon: "fa-github", href: "https://www.github.com/pdimens/harpy/", align: right }
---

```{r setup}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
logfile <- file(snakemake@log$logfile, open = "wt")
sink(logfile)
sink(logfile, type = "message")
```

```{r package_imports}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr","DT","scales", "plotly", "highcharter")
```

```{r read_input}
#file <- "~/blocks.summary.gz"
file <- snakemake@input[[1]]
df <- read.table(
  file, header = T,
  colClasses = c("factor","factor","integer","integer", "integer")
)
if(nrow(df) < 1){
  cat(paste("No phase blocks were observed in the input file", file, "\n"))
  knitr::knit_exit()
}
df$pos_end <- df$pos_start + df$block_length
df <- df[df$block_length > 0, c(1,2,3,4,6,5)]
levels(df$sample) <- basename(levels(df$sample))
```

```{r process_contigs}
contigs <- group_by(df, contig) %>%
  summarize(size = max(pos_end)) %>%
  arrange(desc(size))
```

# General Stats
## fileheader 
### details desc {.no-title}
<h1> Haplotype Phasing Results </h1>
The data presented here are the results of phasing genotypes into haplotypes using
[HapCut2](https://github.com/vibansal/HapCUT2). The information is derived from `r file`,
which summarized information across all samples using the `.blocks` files HapCut2
generates. The VCF files HapCut2 also generates were not used as they lack
some of the information in the `.blocks` files that were collated in this report.
This page shows general and per-contig information. You may click the plots to
expand them in your browser. Clicking them again will exit from the zoomed pop-up
window. The `Per-Sample Stats` tab in the navigation bar above will show you
statistics relating to haplotypes on the per-sample level. Haplotype blocks with a size
of `0` were removed from the data.

## Haplotyping details {data-height=100}
```{r process_stats}
overall <- df %>% summarise(
    total_snps = sum(n_snp),
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
)
```

### total phased
```{r}
valueBox(scales::comma(overall$total_snps[1]), caption = "Phased SNPs", color = "#81ce9b")
```

### avg snps
```{r}
valueBox(scales::comma(overall$mean_snps[1]), caption = "Avg SNPs/Haplotype", color = "info")
```

### median snps
```{r}
valueBox(scales::comma(overall$median_snps[1]), caption = "Median SNPs/Haplotype", color = "info")
```
### avg blocksize
```{r}
valueBox(scales::comma(overall$mean_blocksize[1]), caption = "Avg Haplotype Length", color = "info")
```

### median blocksize
```{r}
valueBox(scales::comma(overall$median_blocksize[1]), caption = "Median Haplotype Length", color = "info")
```

### longest haplotype
```{r}
valueBox(scales::comma(overall$max_blocksize[1]), caption = "Longest Haplotype (bp)", color = "info")
```

## distribution plots
### haplotype length {.no-title}
<h2> Distribution of Haplotype Lengths </h2>
The plots below show the distribution of haplotype length (in base pairs). 
Phasing typically has a few very large haplotypes, resulting in extreme 
right-tails in these distributions, therefore, the X-axis is **log-scaled lengths**
to collapse the right tail for better readability. 

## dist plot
### the dist plpt {.no-title}
```{r block_len_dist_plot, out.width="100%"}
hs <- hist(
  df$block_length,
  breaks = 500,
  plot = F
)
hs <- data.frame(val = hs$mids, freq = hs$counts)
hchart(hs, "areaspline", hcaes(x = val, y = freq), color = "#7eb495", name = "count") |>
  hc_xAxis(title = list(text = "haplotype length in base pairs (log scale)"), type = "logarithmic") |>
  hc_title(text = "Haplotype block lengths") |>
  hc_yAxis(title = list(text = "count")) |>
  hc_tooltip(crosshairs = TRUE) |>
  hc_exporting(enabled = T, filename = "haplotype.hist",
    buttons = list(contextButton = list(menuItems = c("downloadPNG", "downloadJPEG", "downloadPDF", "downloadSVG")))
  )
```

## Per contig data desc
### table per contig {.no-title}
<h2> Stats Per Contig </h2>
The plot below shows the distribution of haplotype length (in base pairs) for 
each contig, up to 300 contigs. The dotted vertical bar represents the mean haplotype length.

Phasing typically has a few very large haplotypes, resulting in extreme 
right-tails in these distributions, therefore, the haplotype lengths are **log-scaled lengths**
to collapse the right tail for better readability.

```{r dropdown_buttons_func}
dropdown_buttons <- function(COLUMN, cutoff){
  buttonlist <- list()
  categories <- unique(COLUMN)
  idx <- 0
  for(i in categories){
    idx <- idx + 1
    if(idx > cutoff) break
    visibility <- as.list(rep(FALSE, length(categories)))
    visibility[[idx]] <- TRUE
    buttonlist[[idx]] <- list(method = "restyle", label = i, args = list("visible", visibility))
  }
  return( list(list(y = 1, buttons = buttonlist)) )
}
```

## Per contig data
### stats per contig {.no-title}
```{r per_contig_table, paged.print=TRUE, out.width="100%"}
percontig <- df %>% group_by(contig) %>% summarise(
    n_haplo = round(length(n_snp)),
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
  )

DT::datatable(
  percontig,
  rownames = F,
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "phasing_per_contig")), 
    scrollX = TRUE
  ),
  colnames = c("Contig", "Total Haplotypes", "Mean SNPs", "Median SNPs", "Mean Haplotype Length", "Median Haplotype Length", "Largest Haplotype"),
  fillContainer = T
)
```

### per contig plot {.no-title}
```{r contig_ridges, out.width="100%"}
fig <- plot_ly(hoverinfo = "none") %>%
  layout(
    xaxis = list(title = "Log-Scaled Haplotype Length (bp)", fixedrange = TRUE),#, type = "log"),
    yaxis = list(fixedrange = TRUE),
    title = "Distribution of Haplotype Lengths by Contig",
    showlegend = F,
    updatemenus = dropdown_buttons(df$contig, 300)
  ) %>% config(displayModeBar = FALSE)
# Loop over the categories
.idx <- 0
for (cont in unique(df$contig)) {
  .idx <- .idx + 1
  # stop after 300 contigs
  if(.idx > 300) break
  subset_cont <- df$block_length[df$contig == cont]
  fig <- fig %>%
    add_trace(
      x = log(subset_cont),
      type = "violin",
      name = cont,
      side = "positive",
      meanline = list(visible = T),
      visible = "legendonly"
    )
}

fig
```

# Per-Sample Stats
## per sample desc
### stats per sample desc {.no-title}
<h1> Per-Sample Stats </h1>
Haplotype phasing occurs per-sample and the table below provides details on the outcome of haplotype phasing for each sample.

The plots below shows the distribution of haplotype lengths (in base pairs) for each 
sample. While no Y-axis for these counts are provided, this plot is intended to be more of
a visual reference of the relative distributions of haplotype lengths among samples.
The dotted vertical bar represents the mean haplotype length. 

Phasing typically has a few very large haplotypes, resulting in extreme 
right-tails in these distributions, therefore, the haplotype lengths are
**log-scaled lengths** to collapse the right tail for better readability.

## per sample
### stats per sample {.no-title}
```{r table_persample, out.width="100%"}
persample <- df %>% group_by(sample) %>% summarise(
    n_haplo = sum(length(n_snp)),
    mean_snps = round(mean(n_snp), digits = 0),
    median_snps = median(n_snp),
    mean_blocksize = round(mean(block_length), digits = 0),
    median_blocksize = median(block_length),
    max_blocksize = max(block_length)
)

DT::datatable(
  persample,
  rownames = F,
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip', 
    buttons = list(list(extend = "csv",filename = "phasing_per_sample")), 
    scrollX = TRUE
  ),
  colnames = c("Sample", "Haplotypes", "Mean SNPs", "Median SNPs", "Mean Haplotype Length", "Median Haplotype Length", "Largest Haplotype"),
  fillContainer = T
)
```

### the ridgeplot {.no-title}
```{r sample_ridgeplot, out.width="100%"}
fig <- plot_ly(hoverinfo = "none") %>%
  layout(
    xaxis = list(title = "Log-Scaled Haplotype Length (bp)", fixedrange = TRUE),#, type = "log"),
    yaxis = list(fixedrange = TRUE),
    title = "Distribution of Haplotype Lengths by Sample",
    showlegend = F,
    updatemenus = dropdown_buttons(df$sample, 600)
  ) %>% config(displayModeBar = FALSE)
# Loop over the categories
.idx <- 0
for (samp in unique(df$sample)) {
  .idx <- .idx + 1
  if(.idx > 600) break
  subset_sample <- df$block_length[df$sample == samp] 
  fig <- fig %>%
    add_trace(
      x = log(subset_sample),
      type = "violin",
      name = samp,
      side = "positive",
      meanline = list(visible = T),
      visible = "legendonly"
    )
}

fig
```