---
title: "Leviathan SV Calling Summary"
date: "`r format(Sys.time(), '%d %b, %Y at %H:%M')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    mathjax: NULL
    logo: https://raw.githubusercontent.com/pdimens/harpy/docs/static/logo_report.png
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_report.png"
    navbar:
      - { title : "Docs", icon: "fa-book", href: "https://pdimens.github.io/harpy/", align: right }
      - { title : "Source", icon: "fa-github", href: "https://www.github.com/pdimens/harpy/", align: right }
---

```{r setup}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
logfile <- file(snakemake@log$logfile, open = "wt")
sink(logfile)
sink(logfile, type = "message")
```

```{r package_imports}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr", "tidyr", "ggplot2","DT", "viridisLite", "BioCircos", "tools")
```

```{r read_inputs}
statsfiles <- snakemake@input[["statsfiles"]]
faidx <- snakemake@input[["faidx"]]
#statsfiles <- list.files(path = "~/test/", pattern = "*.stats", full.names = T)
#faidx <- "~/testshad.fai"
```

## Intro
### introtext {.no-title}
<h1> Structural Variants Summary </h1>
Below is a series of tables and plots detailing the structural variants
identified by [LEVIATHAN](https://github.com/morispi/LEVIATHAN) [(preprint)](https://doi.org/10.1101/2021.03.25.437002). The variants are given by:

```{r table2}
knitr::kable(
  data.frame(
  "Variant" = c("INV", "DUP", "DEL","BND"),
  "Name" = c("Inversion", "Duplication", "Deletion", "Breakend"),
  "Description" = c(
    "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
    "a type of mutation that involves the production of one or more copies of a gene or region of a chromosome",
    "a type of mutation that involves the loss of one or more nucleotides from a segment of DNA",
    "the endpoint of a structural variant, which can be useful to explain complex variants"
    )
  )
)
```
This report shows overview information.

In the colorful summary boxes are the per-population average number of
variants detected by type.
```{r}
readvariants <- function(x){
  read.table(
    x,
    header = T,
    colClasses = c("factor", "factor", "integer", "integer", "character", "factor", "integer", "integer")
  )
}
```

```{r process_inputs}
emptyfiles <- character()
sv <- data.frame(matrix(ncol=8,nrow=0, dimnames=list(NULL, c("population",	"contig",	"position_start",	"position_end",	"length",	"type",	"n_barcodes",	"n_pairs"))))
for(i in statsfiles){
  .df <- tryCatch(readvariants(i), error = function(e){return(0)})
  if(length(.df)==1){
      emptyfiles <- c(emptyfiles, i)
  } else {
    sv <- rbind(sv, .df)
  }
}
if(length(emptyfiles) == length(statsfiles)){
  cat("There are no variants detected in any of the populations:\n")
  cat(paste0("- ", statsfiles), sep = "\n")
  knitr::knit_exit()
}
if(length(emptyfiles > 0)){
  cat("Some of the populations did not have any variants detected and are omitted from this report:\n")
  cat(paste0("- ", emptyfiles), sep = "\n")
}
```

```{r summary_info}
summinfo <- sv %>%
  group_by(population, type) %>%
  summarise(tot = n()) %>%
  group_by(type) %>%
  summarise(count = sum(tot), avg = round(mean(tot, na.rm = T),2 ), sd = round(sd(tot, na.rm = T), 2)) #%>% 
```

## header info
### npops
```{r}
valueBox(scales::comma(length(levels(sv$population))), caption = "Populations", color = "success")
```

### nbreakends
```{r}
nbnd <- ifelse("BND" %in% summinfo$type, summinfo$type[which(summinfo$type == "BND")], 3)
valueBox(scales::comma(nbnd, accuracy = 0.01), caption = "Breakends (average)", color = "info")
```
### ndeletions
```{r}
ndel <- ifelse("DEL" %in% summinfo$type, summinfo$type[which(summinfo$type == "DEL")], 3)
valueBox(scales::comma(ndel, accuracy = 0.01), caption = "Deletions (average)", color = "info")
```

### nduplications
```{r}
ndup <- ifelse("DUP" %in% summinfo$type, summinfo$type[which(summinfo$type == "DUP")], 3)
valueBox(scales::comma(ndup, accuracy = 0.01), caption = "Duplications (average)", color = "info")
```

### ninversions
```{r}
ninv <- ifelse("INV" %in% summinfo$type, summinfo$type[which(summinfo$type == "INV")], 3)
valueBox(scales::comma(ninv, accuracy = 0.01), caption = "Inversions (average)", color = "info")
```


## Various information {.tabset data-height=450}
### all variants
This table details all the variants LEVIATHAN detected and the quantity/quality of evidence for those detections.
```{r all variants}
DT::datatable(
  sv,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "leviathan_sv")),
    scrollX = TRUE
  ),
  fillContainer = T
  )
```

```{r fix_breakend_length}
sv$length <- as.numeric(gsub("\\.", "1", sv$length))
```
### counts per contig
This table details counts each type of structural variant for every contig for every population.
```{r percontig counts}
grpcounts <- sv %>%
  group_by(population, contig, type) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = population, values_from = count)

DT::datatable(
  grpcounts,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "leviathan_sv_count_pop")),
    scrollX = TRUE
  ),
  fillContainer = T
  )
```

### bp per contig
This table details the lengths (in bp) of structural variants for every contig for every population. Breakends are not counted because they do not have a size.
```{r bp per contig}
grplens <- sv %>% 
  filter(type != "BND") %>%
  group_by(population, contig, type) %>%
  summarise(bp = sum(length)) %>%
  pivot_wider(names_from = population, values_from = bp)

DT::datatable(
  grplens,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "leviathan_sv_bp_pop")),
    scrollX = TRUE
  ),
  fillContainer = T
  )
```

## SV size and location
### legend {.no-title}
<h2>Visualize the Variants</h2>
Below are three circular plots to visualize the distribution of _putative_ structural variants
across (up to) 30 of the largest contigs across the groups you specified in your data.
If you are unfamiliar with this kind of visualization, it's a circular version of a linear genome.
Each "wedge" is a different contig, from position 0 to  the end of the contig, and is labelled by the contig name.
Each internal (grey) ring is a plot of observed structural variants for that contig for a different population. 

Very small variants may be difficult to see or will not appear on the plot. Deletions are displayed as points as they tend to be small
and clump together. Breakends were not plotted as LEVIATHAN tends to find many more of them than other variants and their presence usually
isn't as useful to visualize as putative inversions, deletions, and duplications. You may hover your cursor over variants to view their
positions, pan by clicking and dragging, and zoom using scroll (mouse or touchpad). In case you become unable to scroll up from the plot
due to these interactive features, place your cursor over the navigation bar of this report and you will be able to scroll the report instead of
zooming on the plot.
```{r plots_setup}
fa.sizes <- read.table(faidx, header = F)[,1:2] %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")

# make sure that the contigs with SV are the ones being plotted, not the others
fa.sizes <- fa.sizes[fa.sizes$contig %in% unique(sv$contig), ]

# limit the data to only the 30 of the largest present contigs
fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]

# filter out variants that aren't on one of these contigs
sv <- sv[sv$contig %in% fa.sizes$contig, ]
sv$contig <- factor(sv$contig)

populations <- levels(sv$population)
color.palette <- turbo(length(populations))
names(color.palette) <- populations

genomeChr <- fa.sizes$size
names(genomeChr) <- fa.sizes$contig
genomeChr <- as.list(genomeChr)
radius <- 0.9
radius_increments <- round(.6 / length(populations), 3)
padding <- radius_increments * 0.2
```

## legendrow {data-height=200}
### Population legend
```{r circoslegend, fig.width = 14, fig.height=1}
pops <- data.frame(x = 1, population = populations)
ggplot(pops, aes(y = population, x = 1, fill = population)) +
  geom_tile(color = "white", lwd = 1.5,linetype = 1) +
  scale_fill_manual(values=color.palette, drop=FALSE) +
  theme_minimal() +
  scale_y_discrete(position = "right", guide = guide_axis(n.dodge=2)) +
  theme(
    panel.grid = element_blank(),
    legend.position = "none",
    axis.text.y=element_blank(),
    axis.ticks.y=element_blank(),
    axis.title = element_blank()
  ) +
  coord_flip() +
  labs(title = "Population")
```

## plotrow {.tabset data-height=900}
### Inversions
```{r circos_inversions, out.width= "80%", out.height="900px"}
radius <- 0.9
tracks <- BioCircosTracklist()
variant_type <- "INV"
# Add one track for each population
for( pop in populations ){
  variant <- sv[(sv$type == variant_type) & (sv$population == pop),]
  if(nrow(variant) > 0){
    tracks <- tracks + BioCircosArcTrack(
      paste0(variant_type, pop),
      as.character(variant$contig),
      variant$position_start,
      variant$position_end,
      opacities = 0.7,
      values = variant$n_barcodes,
      minRadius = radius - (radius_increments - padding),
      maxRadius = radius - padding,
      colors = color.palette[pop],
      labels = paste0("Population: ", pop, "</br>Type: Inversion"),
    )
  }
  # loop to create backgrounds
  tracks <- tracks + BioCircosBackgroundTrack(
    paste0(pop,"_background"),
    minRadius = radius - radius_increments, maxRadius = radius,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius - radius_increments
}
BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  SNPMouseOverCircleSize = 2,
  SNPMouseOverTooltipsHtml03 = "<br/>Length: ",
  width = "100%",
  height = "1000px",
  elementId = paste0(variant_type, "_circosplot") 
)
```


### Duplications {data-height=900}
```{r circos_duplications, out.width= "80%", out.height="900px"}
# Add one track for each population
radius <- 0.9
tracks <- BioCircosTracklist()
variant_type <- "DUP"
# Add one track for each population
for( pop in populations ){
  variant <- sv[(sv$type == variant_type) & (sv$population == pop),]
  if(nrow(variant) > 0){
    tracks <- tracks + BioCircosArcTrack(
      paste0(variant_type, pop),
      as.character(variant$contig),
      variant$position_start,
      variant$position_end,
      opacities = 0.7,
      values = variant$n_barcodes,
      minRadius = radius - (radius_increments - padding),
      maxRadius = radius - padding,
      colors = color.palette[pop],
      labels = paste0("Population: ", pop, "</br>Type: Duplication"),
    )
  }
  # loop to create backgrounds
  tracks <- tracks + BioCircosBackgroundTrack(
    paste0(pop,"_background"),
    minRadius = radius - radius_increments, maxRadius = radius,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius - radius_increments
}

BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  width = "100%",
  height = "1000px",
  elementId = paste0(variant_type, "_circosplot") 
)
```

### Deletions {data-height=900}
```{r circos_deletions, out.width= "80%", out.height="900px"}
# Add one track for each population
radius <- 0.9
tracks <- BioCircosTracklist()
variant_type <- "DEL"
# Add one track for each population
for( pop in populations ){
  variant <- sv[(sv$type == variant_type) & (sv$population == pop),]
  if(nrow(variant) > 0){
    tracks <- tracks + BioCircosSNPTrack(
      paste0(variant_type, pop),
      as.character(variant$contig),
      variant$position_start,
      values = abs(variant$position_end - variant$position_start),
      minRadius = radius - (radius_increments - padding),
      maxRadius = radius - padding,
      size = 2,
      shape = "rect",
      opacities = 0.7,
      colors = color.palette[pop],
      labels = paste0("Population: ", pop, "</br>Type: Deletion"),
  )
  }
  # loop to create backgrounds
  tracks <- tracks + BioCircosBackgroundTrack(
    paste0(pop,"_background"),
    minRadius = radius - radius_increments, maxRadius = radius,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius - radius_increments
}

BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  SNPMouseOverCircleSize = 2,
  SNPMouseOverTooltipsHtml03 = "<br/>Length: ",
  width = "100%",
  height = "1000px",
  elementId = paste0(variant_type, "_circosplot") 
)
```