---
title: "NAIBR SV Calling Summary"
date: "`r format(Sys.time(), '%d %b, %Y at %H:%M')`"
output:
  flexdashboard::flex_dashboard:
    theme: lumen
    orientation: rows
    vertical_layout: scroll
    horizontal_layout: fill
    mathjax: NULL
    logo: https://raw.githubusercontent.com/pdimens/harpy/docs/static/logo_report.png
    favicon: "https://raw.githubusercontent.com/pdimens/harpy/docs/static/favicon_report.png"
    navbar:
      - { title : "Docs", icon: "fa-book", href: "https://pdimens.github.io/harpy/", align: right }
      - { title : "Source", icon: "fa-github", href: "https://www.github.com/pdimens/harpy/", align: right }
---

```{r setup}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
logfile <- file(snakemake@log$logfile, open = "wt")
sink(logfile)
sink(logfile, type = "message")
```

```{r package_imports}
using<-function(...) {
    libs<-unlist(list(...))
    req<-unlist(lapply(libs,require,character.only=TRUE))
    need<-libs[req==FALSE]
    if(length(need)>0){ 
        install.packages(need, repos = "https://cloud.r-project.org/")
        lapply(need,require,character.only=TRUE)
    }
}
using("flexdashboard","dplyr", "ggplot2", "tidyr","DT","BioCircos", "viridisLite", "tools")
```

```{r read_inputs}
infiles <- snakemake@input[["bedpe"]]
fai <- snakemake@input[["fai"]]
#infiles <- list.files(path = "~/test", pattern = "*bedpe", full.names = T)
#fai <- "~/testshad.fai"
```

## Intro
### introtext {.no-title}
<h1> Structural Variants Summary </h1>
Below is a series of tables and plots detailing the structural variants
identified by [NAIBR](https://github.com/pontushojer/NAIBR) [(publication)](https://doi.org/10.1093/bioinformatics/btx712).
The variants are given by:

```{r table2, results='asis'}
knitr::kable(
  data.frame(
    "Name" = c("Inversion", "Duplication", "Deletion"),
    "Description" = c(
      "a segment that broke off and reattached within the same chromosome, but in reverse orientation",
      "a type of mutation involving the production of one or more copies of a gene or region of a chromosome",
      "a type of mutation involving the loss of one or more nucleotides from a segment of DNA"
      )
    )
)
```
This report shows overview information.

In the colorful summary boxes are the per-population average number of
variants detected by type.

```{r readvariants_func}
readvariants <- function(x){
  y <- read.table(
    x,
    header = T,
    colClasses = c("factor", "integer", "factor", "integer", "integer", "integer", "character", "character", "numeric", "factor", "factor")
  )
  y$Population <- gsub(".bedpe", "", basename(x))
  y$Population <- as.factor(y$Population)
  y <- y[, c(12,1:11)]
  return(y)
}

cleanbedpe <- function(x){
  y <- x[x$Chr1 == x$Chr2, c(-4, -11)]
  names(y)[2] <- "Chr1"
  y$Score <- round(y$Score, digits = 3)
  y$Length <- abs(y$Break2 - y$Break1)
  return(y)
}

chimeric <- function(x){
  y <- x[x$Chr1!=x$Chr2, -11]
  y$Score <- round(y$Score, digits = 3)
  return(y)
}
```

```{r process_inputs}
emptyfiles <- character()
sv <- data.frame(matrix(ncol=12,nrow=0, dimnames=list(NULL, c("Population", "Chr1", "Break1", "Chr2", "Break2", "SplitMolecules", "DiscordantReads", "Orientation", "Haplotype", "Score", "PassFilter", "SV"))))
for(i in infiles){
  .df <- tryCatch(readvariants(i), error = function(e){return(0)})
  if(nrow(.df)==0){
      emptyfiles <- c(emptyfiles, i)
  } else {
    sv <- rbind(sv, .df)
  }
}
if(length(emptyfiles) == length(infiles)){
  cat("There are no variants detected in any of the populations:\n")
  cat(paste0("- ", infiles), sep = "\n")
  knitr::knit_exit()
}
if(length(emptyfiles > 0)){
  cat("Some of the populations did not have any variants detected and are omitted from this report:\n")
  cat(paste0("- ", emptyfiles), sep = "\n")
}
```

```{r filter_variants}
sv_clean <- cleanbedpe(sv)
```

```{r summary_info}
summinfo <- sv %>%
  group_by(Population, SV) %>%
  summarise(tot = n()) %>%
  group_by(SV) %>%
  summarise(count = sum(tot), avg = round(mean(tot, na.rm = T),2 ), sd = round(sd(tot, na.rm = T), 2)) #%>% 
```

## header info
### npops
```{r}
valueBox(scales::comma(length(levels(sv$Population))), caption = "Populations", color = "success")
```

### ndeletions
```{r}
variant <- "deletion"
ndel <- ifelse(variant %in% summinfo$SV, summinfo$avg[which(summinfo$SV == variant)], 0)
valueBox(scales::comma(ndel, accuracy = 0.1), caption = "Deletions (average)", color = "info")
```

### nduplications
```{r}
variant <- "duplication"
ndup <- ifelse(variant %in% summinfo$SV, summinfo$avg[which(summinfo$SV == variant)], 0)
valueBox(scales::comma(ndup, accuracy = 0.1), caption = "Duplications (average)", color = "info")
```

### ninversions
```{r}
variant <- "inversion"
ninv <- ifelse(variant %in% summinfo$SV, summinfo$avg[which(summinfo$SV == variant)], 0)
valueBox(scales::comma(ninv, accuracy = 0.1), caption = "Inversions (average)", color = "info")
```

## per contig descs
### by count {.no-title}
<h3> Variants by count </h3>
The table below details counts each type of structural variant for every contig for every population.

### by length {.no-title}
<h3> Variants by length </h3>
The table below details the lengths (in bp) of structural variants for every contig for every population.

## Per contig counts
### Variants by count {.no-title}
```{r sv_per_cont_table}
grpcounts <- sv_clean %>%
  group_by(Population, Chr1, SV) %>%
  summarise(count = n()) %>%
  pivot_wider(names_from = Population, values_from = count)

DT::datatable(
  grpcounts,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "naibr_sv_count_pop")),
    scrollX = TRUE
  ),
  fillContainer = T
)
```

### Variants by base-pair span {.no-title}
```{r variant_len_table}
grplens <- sv_clean %>%
  group_by(Population, Chr1, SV) %>%
  summarise(bp = sum(Length)) %>%
  pivot_wider(names_from = Population, values_from = bp)

DT::datatable(
  grplens,
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "naibr_sv_count_bp")),
    scrollX = TRUE
  ),
  fillContainer = T
)
```

## Various Stats
### interpreting output {.no-title}
<h2> Interpreting Output </h2>
NAIBR outputs a tab-delimited file with named columns (along with a VCF
and a reformatted bedpe file with extra information). The columns of the
bedpe file are deciphered as such:

```{r explanation_table}
knitr::kable(
  data.frame(
    "Column" = names(sv),
    "Descrption" = c(
      "Name of the population",
      "Name of the contig where the variant starts",
      "Base-pair position in `Chr1` of the start of the variant",
      "Name of the contig where the variant ends",
      "Base-pair position in `Chr2` of the end of the variant",
      "Number of split molecules supporting variant",
      "Number of discordant reads supporting variant",
      "Orientation of variant relative to reference",
      "The haplotype of the variant",
      "log-likelihood score for variant",
      "`PASS` if passes internal NAIBR filter threshold, otherwise `FAIL`",
      "The type of variant (inversion, deletion, or duplication)"
    )
  ), col.names = c("Column Name", "Description")
)
```
## variants anc chimeras desc
### all variants desc {.no-title}
<h3> All non-chimeric variants </h3>
The table below details the variants detected by NAIBR that appear on a single contig/chromosome and passed the programs internal filtering.

### chimeras desc {.no-title}
<h3> Chimeric variants </h3>
The table below shows any structural variants whose breakpoints span multiple contigs and may require further assessment. These variants are not shown in the plots below.

## Variants and Chimeras
### all variants {.no-title}
```{r all_sv_table}
DT::datatable(
  sv_clean,
  rownames = F,
  filter = "top",
  extensions = 'Buttons', 
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "naibr_sv")),
    scrollX = TRUE
  ),
  fillContainer = T
)
```
### chimeric variants {.no-title}
```{r chimeric_table}
DT::datatable(
  chimeric(sv),
  rownames = F,
  filter = "top",
  extensions = 'Buttons',
  options = list(
    dom = 'Brtip',
    buttons = list(list(extend = "csv",filename = "naibr_sv_chimeric")),
    scrollX = TRUE
  )
)
```

## SV size and location
### legend {.no-title}
<h2>Visualize the Variants</h2>
Below are three circular plots to visualize the distribution of _putative_ structural variants
across (up to) 30 of the largest contigs across the groups you specified in your data.
If you are unfamiliar with this kind of visualization, it's a circular version of a linear genome.
Each "wedge" is a different contig, from position 0 to  the end of the contig, and is labelled by the contig name.
Each internal (grey) ring is a plot of observed structural variants for that contig for a different population. 

Very small variants may be difficult to see or will not appear on the plot. Deletions are displayed as points as they tend to be small
and clump together. You may hover your cursor over variants to view their
positions, pan by clicking and dragging, and zoom using scroll (mouse or touchpad). In case you become unable to scroll up from the plot
due to these interactive features, place your cursor over the navigation bar of this report and you will be able to scroll the report instead of
zooming on the plot.
```{r plots_setup}
fa.sizes <- read.table(fai, header = F) %>% arrange(desc(2))
colnames(fa.sizes) <- c("contig", "size")
plot_contigs <- snakemake@params$contigs
if (all(plot_contigs == "default")){
  # make sure that the contigs with SV are the ones being plotted, not the others
  fa.sizes <- fa.sizes[fa.sizes$contig %in% unique(sv$Chr1), 1:2]
  # limit the data to only the 30 of the largest present contigs
  fa.sizes <- fa.sizes[1:min(nrow(fa.sizes), 30), ]
} else {
  fa.sizes <- fa.sizes[fa.sizes$contig %in% plot_contigs, 1:2]
}

# filter out variants that aren't on one of these contigs
sv_clean <- sv_clean[sv_clean$Chr1 %in% fa.sizes$contig, ]
sv_clean$Chr1 <- factor(sv_clean$Chr1)

populations <- levels(sv_clean$Population)
color.palette <- turbo(length(populations))
names(color.palette) <- populations

genomeChr <- fa.sizes$size
names(genomeChr) <- fa.sizes$contig
genomeChr <- as.list(genomeChr)
radius <- 0.9
radius_increments <- round(.6 / length(populations), 3)
padding <- radius_increments * 0.2
```

## legendrow {data-height=200}
### Population legend
```{r circoslegend, fig.width = 14, fig.height=1}
pops <- data.frame(x = 1, population = populations)
ggplot(pops, aes(y = population, x = 1, fill = population)) +
  geom_tile(color = "white", lwd = 1.5,linetype = 1) +
  scale_fill_manual(values=color.palette, drop=FALSE) +
  theme_minimal() +
  scale_y_discrete(position = "right", guide = guide_axis(n.dodge=2)) +
  theme(
    panel.grid = element_blank(),
    legend.position = "none",
    axis.text.y=element_blank(),
    axis.ticks.y=element_blank(),
    axis.title = element_blank()
  ) +
  coord_flip() +
  labs(title = "Population")
```

## plotrow {.tabset data-height=900}
### Inversions
```{r circos_inversions, out.width= "80%", out.height="900px"}
radius <- 0.9
tracks <- BioCircosTracklist()
variant_type <- "inversion"
# Add one track for each population
for( pop in populations ){
  variant <- sv_clean[(sv_clean$SV == variant_type) & (sv_clean$Population == pop),]
  if(nrow(variant) > 0){
    tracks <- tracks + BioCircosArcTrack(
      paste0(variant_type, pop),
      as.character(variant$Chr1),
      variant$Break1,
      variant$Break2,
      opacities = 0.7,
      values = variant$SplitMolecules,
      minRadius = radius - (radius_increments - padding),
      maxRadius = radius - padding,
      colors = color.palette[pop],
      labels = paste0("Population: ", pop, "</br>Type: Inversion"),
    )
  }
  # loop to create backgrounds
  tracks <- tracks + BioCircosBackgroundTrack(
    paste0(pop,"_background"),
    minRadius = radius - radius_increments, maxRadius = radius,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius - radius_increments
}
BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  SNPMouseOverCircleSize = 2,
  SNPMouseOverTooltipsHtml03 = "<br/>Length: ",
  width = "100%",
  height = "1000px",
  elementId = paste0(variant_type, "_circosplot") 
)
```


### Duplications {data-height=900}
```{r circos_duplications, out.width= "80%", out.height="900px"}
# Add one track for each population
radius <- 0.9
tracks <- BioCircosTracklist()
variant_type <- "duplication"
# Add one track for each population
for( pop in populations ){
  variant <- sv_clean[(sv_clean$SV == variant_type) & (sv_clean$Population == pop),]
  if(nrow(variant) > 0){
    tracks <- tracks + BioCircosArcTrack(
      paste0(variant_type, pop),
      as.character(variant$Chr1),
      variant$Break1,
      variant$Break2,
      opacities = 0.7,
      values = variant$SplitMolecules,
      minRadius = radius - (radius_increments - padding),
      maxRadius = radius - padding,
      colors = color.palette[pop],
      labels = paste0("Population: ", pop, "</br>Type: Duplication"),
    )
  }
  # loop to create backgrounds
  tracks <- tracks + BioCircosBackgroundTrack(
    paste0(pop,"_background"),
    minRadius = radius - radius_increments, maxRadius = radius,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius - radius_increments
}

BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  width = "100%",
  height = "1000px",
  elementId = paste0(variant_type, "_circosplot") 
)
```

### Deletions {data-height=900}
```{r circos_deletions, out.width= "80%", out.height="900px"}
# Add one track for each population
radius <- 0.9
tracks <- BioCircosTracklist()
variant_type <- "deletion"
# Add one track for each population
for( pop in populations ){
  variant <- sv_clean[(sv_clean$SV == variant_type) & (sv_clean$Population == pop),]
  if(nrow(variant) > 0){
    tracks <- tracks + BioCircosSNPTrack(
      paste0(variant_type, pop),
      as.character(variant$Chr1),
      variant$Break1,
      values = abs(variant$Break2 - variant$Break1),
      minRadius = radius - (radius_increments - padding),
      maxRadius = radius - padding,
      size = 2,
      shape = "rect",
      opacities = 0.7,
      colors = color.palette[pop],
      labels = paste0("Population: ", pop, "</br>Type: Deletion"),
  )
  }
  # loop to create backgrounds
  tracks <- tracks + BioCircosBackgroundTrack(
    paste0(pop,"_background"),
    minRadius = radius - radius_increments, maxRadius = radius,
    fillColors = "#F9F9F9",
    borderColors = "white",
    borderSize = 1
  )
  radius <- radius - radius_increments
}

BioCircos(
  tracks,
  displayGenomeBorder = F,
  genome = genomeChr,
  chrPad = 0.03,
  genomeTicksDisplay = F,
  genomeLabelDy = 3,
  genomeLabelTextSize = 7,
  SNPMouseOverCircleSize = 2,
  SNPMouseOverTooltipsHtml03 = "<br/>Length: ",
  width = "100%",
  height = "1000px",
  elementId = paste0(variant_type, "_circosplot") 
)
```
